<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 12 Answers to practical exercises | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences</title>
  <meta name="description" content="Chapter 12 Answers to practical exercises | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 12 Answers to practical exercises | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 12 Answers to practical exercises | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  
  
  

<meta name="author" content="Natalie Cooper" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="diversification-rate-analyses.html"/>
<link rel="next" href="common-mistakes-using-pcms-in-r.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">PCM Primer</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#datasets-and-scripts"><i class="fa fa-check"></i>Datasets and scripts</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#how-to-use-these-materials"><i class="fa fa-check"></i>How to use these materials</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#boxes"><i class="fa fa-check"></i>Boxes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#example-datasets"><i class="fa fa-check"></i>Example datasets</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#frog-eye-size-evolution"><i class="fa fa-check"></i>Frog eye size evolution</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#natricine-snake-head-shape-evolution"><i class="fa fa-check"></i>Natricine snake head shape evolution</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#diversification-in-dragonflies"><i class="fa fa-check"></i>Diversification in dragonflies</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#citing-r-and-r-packages"><i class="fa fa-check"></i>Citing R and R packages</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#acknowledgements"><i class="fa fa-check"></i>Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><i class="fa fa-check"></i><b>1</b> What you need to be able to do in R before you start</a>
<ul>
<li class="chapter" data-level="1.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-r-and-rstudio"><i class="fa fa-check"></i><b>1.1</b> Installing R (and RStudio)</a></li>
<li class="chapter" data-level="1.2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#downloading-the-data"><i class="fa fa-check"></i><b>1.2</b> Downloading the data</a></li>
<li class="chapter" data-level="1.3" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#r-projects"><i class="fa fa-check"></i><b>1.3</b> R Projects</a></li>
<li class="chapter" data-level="1.4" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#using-a-script"><i class="fa fa-check"></i><b>1.4</b> Using a script</a></li>
<li class="chapter" data-level="1.5" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-and-loading-extra-packages-in-r"><i class="fa fa-check"></i><b>1.5</b> Installing and loading extra packages in R</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#loading-packages-into-r"><i class="fa fa-check"></i><b>1.5.1</b> Loading packages into R</a></li>
<li class="chapter" data-level="1.5.2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#ggtree-a-special-case-for-package-installation"><i class="fa fa-check"></i><b>1.5.2</b> <code>ggtree</code>, a special case for package installation</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#loading-and-viewing-your-data-in-r"><i class="fa fa-check"></i><b>1.6</b> Loading and viewing your data in R</a></li>
<li class="chapter" data-level="1.7" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#errors-and-warnings"><i class="fa fa-check"></i><b>1.7</b> Errors and warnings</a>
<ul>
<li class="chapter" data-level="1.7.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#removing-the"><i class="fa fa-check"></i><b>1.7.1</b> Removing the <code>+</code></a></li>
</ul></li>
<li class="chapter" data-level="1.8" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#summary"><i class="fa fa-check"></i><b>1.8</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html"><i class="fa fa-check"></i><b>2</b> Data wrangling comparative data in R</a>
<ul>
<li class="chapter" data-level="" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#before-you-start"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="2.1" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#chameleon-data"><i class="fa fa-check"></i><b>2.1</b> Chameleon data</a></li>
<li class="chapter" data-level="2.2" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#summary-1"><i class="fa fa-check"></i><b>2.2</b> Summary</a></li>
<li class="chapter" data-level="2.3" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#practical-exercise"><i class="fa fa-check"></i><b>2.3</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html"><i class="fa fa-check"></i><b>3</b> Understanding, plotting and manipulating phylogenies in R</a>
<ul>
<li class="chapter" data-level="" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#before-you-start-1"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="3.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#what-do-phylogeny-files-look-like"><i class="fa fa-check"></i><b>3.1</b> What do phylogeny files look like?</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#newick-trees"><i class="fa fa-check"></i><b>3.1.1</b> Newick trees</a></li>
<li class="chapter" data-level="3.1.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#nexus-trees"><i class="fa fa-check"></i><b>3.1.2</b> NEXUS trees</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#how-are-phylogenies-formatted-in-r"><i class="fa fa-check"></i><b>3.2</b> How are phylogenies formatted in R?</a></li>
<li class="chapter" data-level="3.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#plotting-phylogenies"><i class="fa fa-check"></i><b>3.3</b> Plotting phylogenies</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#basic-phylogeny-plotting-with-ape"><i class="fa fa-check"></i><b>3.3.1</b> Basic phylogeny plotting with APE</a></li>
<li class="chapter" data-level="3.3.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#basic-phylogeny-plotting-with-ggtree"><i class="fa fa-check"></i><b>3.3.2</b> Basic phylogeny plotting with <code>ggtree</code></a></li>
<li class="chapter" data-level="3.3.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#back-to-ggtree"><i class="fa fa-check"></i><b>3.3.3</b> Back to <code>ggtree</code></a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#phylogeny-manipulation"><i class="fa fa-check"></i><b>3.4</b> Phylogeny manipulation</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#removing-polytomies"><i class="fa fa-check"></i><b>3.4.1</b> Removing polytomies</a></li>
<li class="chapter" data-level="3.4.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#rooting-your-phylogeny"><i class="fa fa-check"></i><b>3.4.2</b> Rooting your phylogeny</a></li>
<li class="chapter" data-level="3.4.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#manipulating-the-species-in-your-phlyogeny"><i class="fa fa-check"></i><b>3.4.3</b> Manipulating the species in your phlyogeny</a></li>
<li class="chapter" data-level="3.4.4" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#writing-phylogenies-to-file"><i class="fa fa-check"></i><b>3.4.4</b> Writing phylogenies to file</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#summary-2"><i class="fa fa-check"></i><b>3.5</b> Summary</a></li>
<li class="chapter" data-level="3.6" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#practical-exercise-1"><i class="fa fa-check"></i><b>3.6</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html"><i class="fa fa-check"></i><b>4</b> Preparing your tree and data for PCMs in R</a>
<ul>
<li class="chapter" data-level="" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#before-you-start-2"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="4.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#preparation"><i class="fa fa-check"></i><b>4.1</b> Preparation</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#load-the-required-packages"><i class="fa fa-check"></i><b>4.1.1</b> Load the required packages</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-and-checking-your-phylogeny-in-r"><i class="fa fa-check"></i><b>4.2</b> Reading and checking your phylogeny in R</a></li>
<li class="chapter" data-level="4.3" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>4.3</b> Reading the data into R</a></li>
<li class="chapter" data-level="4.4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-your-data-to-your-phylogeny"><i class="fa fa-check"></i><b>4.4</b> Matching your data to your phylogeny</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#species-names-with-spaces"><i class="fa fa-check"></i><b>4.4.1</b> Species names with spaces</a></li>
<li class="chapter" data-level="4.4.2" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#mismatches-between-species-in-your-data-and-phylogeny"><i class="fa fa-check"></i><b>4.4.2</b> Mismatches between species in your data and phylogeny</a></li>
<li class="chapter" data-level="4.4.3" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-the-tree-and-the-data"><i class="fa fa-check"></i><b>4.4.3</b> Matching the tree and the data</a></li>
<li class="chapter" data-level="4.4.4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#nas-in-your-data"><i class="fa fa-check"></i><b>4.4.4</b> NAs in your data</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#quick-template-code"><i class="fa fa-check"></i><b>4.5</b> Quick template code</a></li>
<li class="chapter" data-level="4.6" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#summary-3"><i class="fa fa-check"></i><b>4.6</b> Summary</a></li>
<li class="chapter" data-level="4.7" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#practical-exercise-2"><i class="fa fa-check"></i><b>4.7</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html"><i class="fa fa-check"></i><b>5</b> Phylogenetic Signal in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#before-you-start-3"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="5.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#preparation-1"><i class="fa fa-check"></i><b>5.1</b> Preparation</a></li>
<li class="chapter" data-level="5.2" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#estimating-phylogenetic-signal-for-continuous-variables"><i class="fa fa-check"></i><b>5.2</b> Estimating phylogenetic signal for continuous variables</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#pagels-lambda"><i class="fa fa-check"></i><b>5.2.1</b> Pagel’s <span class="math inline">\(\lambda\)</span></a></li>
<li class="chapter" data-level="5.2.2" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#blombergs-k-blomberg-et-al-2003"><i class="fa fa-check"></i><b>5.2.2</b> Blomberg’s <em>K</em> (Blomberg et al 2003)</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#estimating-phylogenetic-signal-for-non-continuous-variables"><i class="fa fa-check"></i><b>5.3</b> Estimating phylogenetic signal for non-continuous variables</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#d"><i class="fa fa-check"></i><b>5.3.1</b> <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#summary-4"><i class="fa fa-check"></i><b>5.4</b> Summary</a></li>
<li class="chapter" data-level="5.5" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#practical-exercises"><i class="fa fa-check"></i><b>5.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html"><i class="fa fa-check"></i><b>6</b> Phylogenetic Generalised Least Squares (PGLS) in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#before-you-start-4"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="6.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#preparation-2"><i class="fa fa-check"></i><b>6.1</b> Preparation</a></li>
<li class="chapter" data-level="6.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#phylogenetic-generalized-least-squares-models-pgls"><i class="fa fa-check"></i><b>6.2</b> Phylogenetic generalized least squares models (PGLS)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#fitting-pgls-models-in-r"><i class="fa fa-check"></i><b>6.2.1</b> Fitting PGLS models in R</a></li>
<li class="chapter" data-level="6.2.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#model-diagnostics-for-pgls-models"><i class="fa fa-check"></i><b>6.2.2</b> Model diagnostics for PGLS models</a></li>
<li class="chapter" data-level="6.2.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#likelihood-profiles-for-lambda-in-pgls-models"><i class="fa fa-check"></i><b>6.2.3</b> Likelihood profiles for <span class="math inline">\(\lambda\)</span> in PGLS models</a></li>
<li class="chapter" data-level="6.2.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#a-pgls-example-with-more-than-one-predictor-variable"><i class="fa fa-check"></i><b>6.2.4</b> A PGLS example with more than one predictor variable</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#common-issues-when-running-pgls-models-in-r"><i class="fa fa-check"></i><b>6.3</b> Common issues when running PGLS models in R</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#optim-errors"><i class="fa fa-check"></i><b>6.3.1</b> Optim errors</a></li>
<li class="chapter" data-level="6.3.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#big-problems-with-small-datasets"><i class="fa fa-check"></i><b>6.3.2</b> Big problems with small datasets</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#summary-5"><i class="fa fa-check"></i><b>6.4</b> Summary</a></li>
<li class="chapter" data-level="6.5" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#practical-exercises-1"><i class="fa fa-check"></i><b>6.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html"><i class="fa fa-check"></i><b>7</b> Phylogenetic generalised linear mixed models in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#before-you-start-5"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="7.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#why-fit-phylogenetic-generalised-linear-mixed-models"><i class="fa fa-check"></i><b>7.1</b> Why fit phylogenetic generalised linear mixed models?</a></li>
<li class="chapter" data-level="7.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#a-really-quick-reminder-of-how-bayesian-analyses-work-with-apologies-to-bayesian-scholars-for-the-simplification"><i class="fa fa-check"></i><b>7.2</b> A really quick reminder of how Bayesian analyses work, with apologies to Bayesian scholars for the simplification!</a></li>
<li class="chapter" data-level="7.3" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#preparation-3"><i class="fa fa-check"></i><b>7.3</b> Preparation</a></li>
<li class="chapter" data-level="7.4" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmcglmm-specific-preparation"><i class="fa fa-check"></i><b>7.4</b> MCMCglmm specific preparation</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#inverse-phylogenetic-variance-covariance-matrix"><i class="fa fa-check"></i><b>7.4.1</b> Inverse phylogenetic variance-covariance matrix</a></li>
<li class="chapter" data-level="7.4.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#priors"><i class="fa fa-check"></i><b>7.4.2</b> Priors</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#other-arguments-in-mcmcglmm"><i class="fa fa-check"></i><b>7.5</b> Other arguments in <code>MCMCglmm</code></a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#nitt"><i class="fa fa-check"></i><b>7.5.1</b> nitt</a></li>
<li class="chapter" data-level="7.5.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#burnin"><i class="fa fa-check"></i><b>7.5.2</b> burnin</a></li>
<li class="chapter" data-level="7.5.3" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#thin"><i class="fa fa-check"></i><b>7.5.3</b> thin</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#fitting-the-mcmcglmm-model"><i class="fa fa-check"></i><b>7.6</b> Fitting the MCMCglmm model</a></li>
<li class="chapter" data-level="7.7" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#did-the-model-converge"><i class="fa fa-check"></i><b>7.7</b> Did the model converge?</a></li>
<li class="chapter" data-level="7.8" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmc-outputs"><i class="fa fa-check"></i><b>7.8</b> MCMC outputs</a></li>
<li class="chapter" data-level="7.9" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#extracting-lambda-for-mcmcglmm-models"><i class="fa fa-check"></i><b>7.9</b> Extracting lambda for MCMCglmm models</a></li>
<li class="chapter" data-level="7.10" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#checking-the-effects-of-priors"><i class="fa fa-check"></i><b>7.10</b> Checking the effects of priors</a></li>
<li class="chapter" data-level="7.11" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#an-example-using-a-pglmm"><i class="fa fa-check"></i><b>7.11</b> An example using a PGLMM</a></li>
<li class="chapter" data-level="7.12" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#adding-extra-fixed-and-random-effects"><i class="fa fa-check"></i><b>7.12</b> Adding extra fixed and random effects</a></li>
<li class="chapter" data-level="7.13" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#summary-6"><i class="fa fa-check"></i><b>7.13</b> Summary</a></li>
<li class="chapter" data-level="7.14" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#practical-exercises-2"><i class="fa fa-check"></i><b>7.14</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html"><i class="fa fa-check"></i><b>8</b> Macroevolutionary models for continuous traits</a>
<ul>
<li class="chapter" data-level="" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#before-you-start-6"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="8.1" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#preparation-4"><i class="fa fa-check"></i><b>8.1</b> Preparation</a></li>
<li class="chapter" data-level="8.2" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#models-of-evolution-for-continuous-traits"><i class="fa fa-check"></i><b>8.2</b> Models of evolution for continuous traits</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#look-at-the-data"><i class="fa fa-check"></i><b>8.2.1</b> Look at the data</a></li>
<li class="chapter" data-level="8.2.2" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#fitting-the-brownian-motion-and-ou-models-of-evolution-using-fitcontinuous"><i class="fa fa-check"></i><b>8.2.2</b> Fitting the Brownian motion and OU models of evolution using <code>fitContinuous</code></a></li>
<li class="chapter" data-level="8.2.3" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#comparing-models-of-evolution-using-aic"><i class="fa fa-check"></i><b>8.2.3</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#summary-7"><i class="fa fa-check"></i><b>8.3</b> Summary</a></li>
<li class="chapter" data-level="8.4" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#practical-exercises-3"><i class="fa fa-check"></i><b>8.4</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html"><i class="fa fa-check"></i><b>9</b> Macroevolutionary models for discrete traits</a>
<ul>
<li class="chapter" data-level="" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#before-you-start-7"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="9.1" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#preparation-5"><i class="fa fa-check"></i><b>9.1</b> Preparation</a></li>
<li class="chapter" data-level="9.2" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#models-of-evolution-for-discrete-traits"><i class="fa fa-check"></i><b>9.2</b> Models of evolution for discrete traits</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#fitting-the-er-sym-and-ard-models-of-evolution-using-fitdiscrete"><i class="fa fa-check"></i><b>9.2.1</b> Fitting the ER, SYM and ARD models of evolution using <code>fitDiscrete</code></a></li>
<li class="chapter" data-level="9.2.2" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#comparing-models-of-evolution-using-aic-1"><i class="fa fa-check"></i><b>9.2.2</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#ancestral-state-estimations"><i class="fa fa-check"></i><b>9.3</b> Ancestral state estimations</a></li>
<li class="chapter" data-level="9.4" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#summary-8"><i class="fa fa-check"></i><b>9.4</b> Summary</a></li>
<li class="chapter" data-level="9.5" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#practical-exercises-4"><i class="fa fa-check"></i><b>9.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html"><i class="fa fa-check"></i><b>10</b> Beyond simple macroevolutionary models</a>
<ul>
<li class="chapter" data-level="" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#before-you-start-8"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="10.1" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#preparation-6"><i class="fa fa-check"></i><b>10.1</b> Preparation</a></li>
<li class="chapter" data-level="10.2" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#more-complex-models-of-trait-evolution"><i class="fa fa-check"></i><b>10.2</b> More complex models of trait evolution</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#preparing-the-tree-and-data-for-ouwie"><i class="fa fa-check"></i><b>10.2.1</b> Preparing the tree and data for OUwie</a></li>
<li class="chapter" data-level="10.2.2" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#running-models-using-ouwie"><i class="fa fa-check"></i><b>10.2.2</b> Running models using OUwie</a></li>
<li class="chapter" data-level="10.2.3" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#single-rate-bm-models"><i class="fa fa-check"></i><b>10.2.3</b> Single rate BM models</a></li>
<li class="chapter" data-level="10.2.4" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#single-optimum-ou-models"><i class="fa fa-check"></i><b>10.2.4</b> Single optimum OU models</a></li>
<li class="chapter" data-level="10.2.5" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-bm-bms-models"><i class="fa fa-check"></i><b>10.2.5</b> Multi-rate BM (BMS) models</a></li>
<li class="chapter" data-level="10.2.6" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-optima-ou-oum-models"><i class="fa fa-check"></i><b>10.2.6</b> Multi-optima OU (OUM) models</a></li>
<li class="chapter" data-level="10.2.7" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-multi-optima-ou-motion-models-oumv"><i class="fa fa-check"></i><b>10.2.7</b> Multi-rate multi-optima OU motion models (OUMV)</a></li>
<li class="chapter" data-level="10.2.8" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-alpha-multi-optima-ou-motion-models-ouma"><i class="fa fa-check"></i><b>10.2.8</b> Multi-alpha multi-optima OU motion models (OUMA)</a></li>
<li class="chapter" data-level="10.2.9" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-multi-optima-ou-motion-models-oumva"><i class="fa fa-check"></i><b>10.2.9</b> Multi-rate multi-optima OU motion models (OUMVA)</a></li>
<li class="chapter" data-level="10.2.10" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#comparing-models-of-evolution-using-aic-2"><i class="fa fa-check"></i><b>10.2.10</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#summary-9"><i class="fa fa-check"></i><b>10.3</b> Summary</a></li>
<li class="chapter" data-level="10.4" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#practical-exercises-5"><i class="fa fa-check"></i><b>10.4</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html"><i class="fa fa-check"></i><b>11</b> Diversification rate analyses</a>
<ul>
<li class="chapter" data-level="" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#before-you-start-9"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="11.1" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#preparation-7"><i class="fa fa-check"></i><b>11.1</b> Preparation</a></li>
<li class="chapter" data-level="11.2" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#lineage-through-time-plots-and-the-gamma-statistic"><i class="fa fa-check"></i><b>11.2</b> Lineage through time plots and the <span class="math inline">\(\gamma\)</span> statistic</a></li>
<li class="chapter" data-level="11.3" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#bamm-bayesian-analysis-of-macroevolutionary-mixtures"><i class="fa fa-check"></i><b>11.3</b> BAMM: Bayesian Analysis of Macroevolutionary Mixtures</a></li>
<li class="chapter" data-level="11.4" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#running-bamm"><i class="fa fa-check"></i><b>11.4</b> Running BAMM</a></li>
<li class="chapter" data-level="11.5" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#looking-at-the-output-of-bamm-using-bammtools-in-r"><i class="fa fa-check"></i><b>11.5</b> Looking at the output of BAMM using BAMMtools in R</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#assessing-mcmc-convergence"><i class="fa fa-check"></i><b>11.5.1</b> Assessing MCMC convergence</a></li>
<li class="chapter" data-level="11.5.2" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#analysis-of-rate-shifts-in-the-dragonfly-phylogeny"><i class="fa fa-check"></i><b>11.5.2</b> Analysis of rate shifts in the dragonfly phylogeny</a></li>
<li class="chapter" data-level="11.5.3" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#mean-phylorate-plot"><i class="fa fa-check"></i><b>11.5.3</b> Mean phylorate plot</a></li>
<li class="chapter" data-level="11.5.4" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#best-single-distinct-shift-configuration"><i class="fa fa-check"></i><b>11.5.4</b> Best single distinct shift configuration</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#credible-sets-of-shifts"><i class="fa fa-check"></i><b>11.6</b> Credible sets of shifts</a></li>
<li class="chapter" data-level="11.7" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#clade-specific-evolutionary-rates"><i class="fa fa-check"></i><b>11.7</b> Clade specific evolutionary rates</a></li>
<li class="chapter" data-level="11.8" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#rate-through-time-analysis"><i class="fa fa-check"></i><b>11.8</b> Rate-through-time analysis</a></li>
<li class="chapter" data-level="11.9" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#macroevolutionary-cohort-analysis"><i class="fa fa-check"></i><b>11.9</b> Macroevolutionary cohort analysis</a></li>
<li class="chapter" data-level="11.10" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#assumptions-and-issues-with-bamm"><i class="fa fa-check"></i><b>11.10</b> Assumptions and issues with BAMM</a></li>
<li class="chapter" data-level="11.11" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#summary-10"><i class="fa fa-check"></i><b>11.11</b> Summary</a></li>
<li class="chapter" data-level="11.12" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#practical-exercises-6"><i class="fa fa-check"></i><b>11.12</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html"><i class="fa fa-check"></i><b>12</b> Answers to practical exercises</a>
<ul>
<li class="chapter" data-level="12.1" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#need-to-know-r"><i class="fa fa-check"></i><b>12.1</b> Need to know R</a></li>
<li class="chapter" data-level="12.2" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#data-wrangling"><i class="fa fa-check"></i><b>12.2</b> Data wrangling</a></li>
<li class="chapter" data-level="12.3" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#phylogenies"><i class="fa fa-check"></i><b>12.3</b> Phylogenies</a></li>
<li class="chapter" data-level="12.4" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#preparation-8"><i class="fa fa-check"></i><b>12.4</b> Preparation</a></li>
<li class="chapter" data-level="12.5" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#phylogenetic-signal"><i class="fa fa-check"></i><b>12.5</b> Phylogenetic signal</a></li>
<li class="chapter" data-level="12.6" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#pgls"><i class="fa fa-check"></i><b>12.6</b> PGLS</a></li>
<li class="chapter" data-level="12.7" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#mcmcglmm"><i class="fa fa-check"></i><b>12.7</b> MCMCglmm</a></li>
<li class="chapter" data-level="12.8" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#models-of-evolution-with-continuous-traits"><i class="fa fa-check"></i><b>12.8</b> Models of evolution with continuous traits</a></li>
<li class="chapter" data-level="12.9" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#models-of-evolution-with-discrete-traits"><i class="fa fa-check"></i><b>12.9</b> Models of evolution with discrete traits</a></li>
<li class="chapter" data-level="12.10" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#beyond-simple-macroevolutionary-models-of-trait-evolution"><i class="fa fa-check"></i><b>12.10</b> Beyond simple macroevolutionary models of trait evolution</a></li>
<li class="chapter" data-level="12.11" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#diversification"><i class="fa fa-check"></i><b>12.11</b> Diversification</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html"><i class="fa fa-check"></i><b>13</b> Common mistakes using PCMs in R</a>
<ul>
<li class="chapter" data-level="13.1" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#good-practice-to-prevent-errors"><i class="fa fa-check"></i><b>13.1</b> Good practice to prevent errors</a></li>
<li class="chapter" data-level="13.2" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#common-errors"><i class="fa fa-check"></i><b>13.2</b> Common errors</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#standard-r-issues"><i class="fa fa-check"></i><b>13.2.1</b> Standard R issues</a></li>
<li class="chapter" data-level="13.2.2" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#issues-with-the-tree-or-data"><i class="fa fa-check"></i><b>13.2.2</b> Issues with the tree or data</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#what-to-do-if-you-get-an-error-message"><i class="fa fa-check"></i><b>13.3</b> What to do if you get an Error message</a></li>
<li class="chapter" data-level="13.4" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#what-to-do-if-you-get-a-warning-message"><i class="fa fa-check"></i><b>13.4</b> What to do if you get a Warning message</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="answers-to-practical-exercises" class="section level1 hasAnchor" number="12">
<h1><span class="header-section-number">Chapter 12</span> Answers to practical exercises<a href="answers-to-practical-exercises.html#answers-to-practical-exercises" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Below are all the questions and answers to the practical exercises at the end of each chapter.</p>
<p>Note that you will need to make sure R knows where to find your data. My data are all in a <code>data</code> folder within the R Project I’m working with. Assuming you have not moved the data from where you downloaded everything, you will have a separate R Project and data folder for each of the chapters of this online book.</p>
<div id="need-to-know-r" class="section level2 hasAnchor" number="12.1">
<h2><span class="header-section-number">12.1</span> Need to know R<a href="answers-to-practical-exercises.html#need-to-know-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>No exercises</p>
</div>
<div id="data-wrangling" class="section level2 hasAnchor" number="12.2">
<h2><span class="header-section-number">12.2</span> Data wrangling<a href="answers-to-practical-exercises.html#data-wrangling" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Using the <code>frog-eyes-raw-data.csv</code> data and R, create a new dataframe that contains only the columns <code>genus_species</code>, <code>Family</code>, and two new columns containing the <em>species mean values</em> for <code>rootmass</code> and <code>eyemean</code>. Note that we will talk more about this dataset in later exercises, but it comes from <span class="citation">K. N. Thomas et al. (<a href="#ref-thomas2020eye" role="doc-biblioref">2020</a>)</span>.</p>
<p>There are many ways of doing this, but here’s how I would do this:</p>
<div class="sourceCode" id="cb552"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb552-1"><a href="answers-to-practical-exercises.html#cb552-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load packages</span></span>
<span id="cb552-2"><a href="answers-to-practical-exercises.html#cb552-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb552-3"><a href="answers-to-practical-exercises.html#cb552-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the data</span></span>
<span id="cb552-4"><a href="answers-to-practical-exercises.html#cb552-4" aria-hidden="true" tabindex="-1"></a>frog <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/frog-eyes-raw-data.csv&quot;</span>)</span>
<span id="cb552-5"><a href="answers-to-practical-exercises.html#cb552-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at it</span></span>
<span id="cb552-6"><a href="answers-to-practical-exercises.html#cb552-6" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(frog)</span></code></pre></div>
<pre><code>## Rows: 640
## Columns: 21
## $ genus_species    &lt;chr&gt; &quot;Acris_gryllus&quot;, &quot;Acris_gryllus&quot;, &quot;Acris_gryllus&quot;, &quot;Adenomus_kelaar…
## $ Order            &lt;chr&gt; &quot;Anura&quot;, &quot;Anura&quot;, &quot;Anura&quot;, &quot;Anura&quot;, &quot;Anura&quot;, &quot;Anura&quot;, &quot;Anura&quot;, &quot;Anu…
## $ Family           &lt;chr&gt; &quot;Hylidae&quot;, &quot;Hylidae&quot;, &quot;Hylidae&quot;, &quot;Bufonidae&quot;, &quot;Bufonidae&quot;, &quot;Hyperol…
## $ Genus            &lt;chr&gt; &quot;Acris&quot;, &quot;Acris&quot;, &quot;Acris&quot;, &quot;Adenomus&quot;, &quot;Adenomus&quot;, &quot;Afrixalus&quot;, &quot;Af…
## $ Species          &lt;chr&gt; &quot;gryllus&quot;, &quot;gryllus&quot;, &quot;gryllus&quot;, &quot;kelaartii&quot;, &quot;kelaartii&quot;, &quot;dorsali…
## $ tiplabel         &lt;chr&gt; &quot;Acris_gryllus_Hylidae&quot;, &quot;Acris_gryllus_Hylidae&quot;, &quot;Acris_gryllus_Hy…
## $ SVL_mm           &lt;dbl&gt; 22.6, 20.7, 16.7, 41.0, 41.7, 25.8, 24.7, 25.4, 48.9, 49.1, 62.5, 6…
## $ Mass_g           &lt;dbl&gt; 1.1, 0.9, 0.6, 6.2, 6.5, 1.5, 1.3, 1.3, 7.0, 7.0, 14.0, 12.0, 30.1,…
## $ rootmass         &lt;dbl&gt; 1.0322801, 0.9654894, 0.8434327, 1.8370906, 1.8662556, 1.1447142, 1…
## $ ED_right_mm      &lt;dbl&gt; 3.3, 3.1, 2.9, NA, 4.7, 3.7, 3.6, 3.5, 7.4, 6.3, 8.3, 7.7, 8.7, 8.1…
## $ ED_left_mm       &lt;dbl&gt; 2.3, 2.8, 2.5, 4.6, 4.7, 3.8, 3.5, 3.8, 7.2, 6.8, 7.8, 7.5, 8.3, 8.…
## $ eyemean          &lt;dbl&gt; 2.80, 2.95, 2.70, 4.60, 4.70, 3.75, 3.55, 3.65, 7.30, 6.55, 8.05, 7…
## $ CD_right_mm      &lt;dbl&gt; 3.1, 2.3, 2.0, NA, 3.8, 2.9, 2.8, 2.7, 5.7, 4.8, 5.8, 5.3, 6.7, 7.3…
## $ CD_left_mm       &lt;dbl&gt; 2.5, 2.2, 2.3, 3.9, 4.2, 3.0, 2.7, 2.8, 5.3, 5.2, 5.9, 5.3, 6.6, 7.…
## $ cormean          &lt;dbl&gt; 2.80, 2.25, 2.15, 3.90, 4.00, 2.95, 2.75, 2.75, 5.50, 5.00, 5.85, 5…
## $ Adult_habitat    &lt;chr&gt; &quot;Ground-dwelling&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwelling&quot;, &quot;Subfossor…
## $ Activity_period  &lt;chr&gt; &quot;Both&quot;, &quot;Both&quot;, &quot;Both&quot;, &quot;Nocturnal&quot;, &quot;Nocturnal&quot;, &quot;Nocturnal&quot;, &quot;Noc…
## $ Mating_habitat   &lt;chr&gt; &quot;Lentic water&quot;, &quot;Lentic water&quot;, &quot;Lentic water&quot;, &quot;Lotic water&quot;, &quot;Lot…
## $ Life_history     &lt;chr&gt; &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;…
## $ Larval_habitat   &lt;chr&gt; &quot;Lentic water&quot;, &quot;Lentic water&quot;, &quot;Lentic water&quot;, NA, NA, &quot;Lentic wat…
## $ Sex_dichromatism &lt;chr&gt; &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent…</code></pre>
<div class="sourceCode" id="cb554"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb554-1"><a href="answers-to-practical-exercises.html#cb554-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a new dataframe called frog2</span></span>
<span id="cb554-2"><a href="answers-to-practical-exercises.html#cb554-2" aria-hidden="true" tabindex="-1"></a>frog2 <span class="ot">&lt;-</span></span>
<span id="cb554-3"><a href="answers-to-practical-exercises.html#cb554-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Start with the frog dataframe</span></span>
<span id="cb554-4"><a href="answers-to-practical-exercises.html#cb554-4" aria-hidden="true" tabindex="-1"></a>  frog <span class="sc">%&gt;%</span></span>
<span id="cb554-5"><a href="answers-to-practical-exercises.html#cb554-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Select only the columns of interest</span></span>
<span id="cb554-6"><a href="answers-to-practical-exercises.html#cb554-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(genus_species, Family, rootmass, eyemean) <span class="sc">%&gt;%</span></span>
<span id="cb554-7"><a href="answers-to-practical-exercises.html#cb554-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Group by species name, and Family (or summarise will try and average this too!)</span></span>
<span id="cb554-8"><a href="answers-to-practical-exercises.html#cb554-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(genus_species, Family) <span class="sc">%&gt;%</span></span>
<span id="cb554-9"><a href="answers-to-practical-exercises.html#cb554-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Summarise to get the mean values</span></span>
<span id="cb554-10"><a href="answers-to-practical-exercises.html#cb554-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">mean_rootmass =</span> <span class="fu">mean</span>(rootmass),</span>
<span id="cb554-11"><a href="answers-to-practical-exercises.html#cb554-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">mean_eyemean =</span> <span class="fu">mean</span>(eyemean))</span>
<span id="cb554-12"><a href="answers-to-practical-exercises.html#cb554-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at frog2</span></span>
<span id="cb554-13"><a href="answers-to-practical-exercises.html#cb554-13" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(frog2)</span></code></pre></div>
<pre><code>## Rows: 220
## Columns: 4
## Groups: genus_species [220]
## $ genus_species &lt;chr&gt; &quot;Acris_gryllus&quot;, &quot;Adenomus_kelaartii&quot;, &quot;Afrixalus_dorsalis&quot;, &quot;Agalychn…
## $ Family        &lt;chr&gt; &quot;Hylidae&quot;, &quot;Bufonidae&quot;, &quot;Hyperoliidae&quot;, &quot;Hylidae&quot;, &quot;Phyllomedusidae&quot;, …
## $ mean_rootmass &lt;dbl&gt; 0.9470674, 1.8516731, 1.1091667, 2.1313583, 2.7840487, 1.9917795, 1.25…
## $ mean_eyemean  &lt;dbl&gt; 2.816667, 4.650000, 3.650000, 7.375000, 8.700000, 6.166667, 4.116667, …</code></pre>
<div class="sourceCode" id="cb556"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb556-1"><a href="answers-to-practical-exercises.html#cb556-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Write to file</span></span>
<span id="cb556-2"><a href="answers-to-practical-exercises.html#cb556-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This is commented out as I don&#39;t want to save this</span></span>
<span id="cb556-3"><a href="answers-to-practical-exercises.html#cb556-3" aria-hidden="true" tabindex="-1"></a><span class="co"># write_csv(frog2, path = &quot;data/clean-frog-data.csv&quot;)</span></span></code></pre></div>
</div>
<div id="phylogenies" class="section level2 hasAnchor" number="12.3">
<h2><span class="header-section-number">12.3</span> Phylogenies<a href="answers-to-practical-exercises.html#phylogenies" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Read in the <code>frog-tree.nex</code> phylogeny from the folder. This comes from <span class="citation">Feng et al. (<a href="#ref-feng2017phylogenomics" role="doc-biblioref">2017</a>)</span>. Then do the following:</p>
<p>Read the tree into R.</p>
<div class="sourceCode" id="cb557"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb557-1"><a href="answers-to-practical-exercises.html#cb557-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load packages</span></span>
<span id="cb557-2"><a href="answers-to-practical-exercises.html#cb557-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb557-3"><a href="answers-to-practical-exercises.html#cb557-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb557-4"><a href="answers-to-practical-exercises.html#cb557-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb557-5"><a href="answers-to-practical-exercises.html#cb557-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb557-6"><a href="answers-to-practical-exercises.html#cb557-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the tree</span></span>
<span id="cb557-7"><a href="answers-to-practical-exercises.html#cb557-7" aria-hidden="true" tabindex="-1"></a>frogtree <span class="ot">&lt;-</span> <span class="fu">read.nexus</span>(<span class="st">&quot;data/frog-tree.nex&quot;</span>)</span>
<span id="cb557-8"><a href="answers-to-practical-exercises.html#cb557-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb557-9"><a href="answers-to-practical-exercises.html#cb557-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at it</span></span>
<span id="cb557-10"><a href="answers-to-practical-exercises.html#cb557-10" aria-hidden="true" tabindex="-1"></a>frogtree</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 214 tips and 213 internal nodes.
## 
## Tip labels:
##   Ascaphus_truei_Ascaphidae, Leiopelma_hochstetteri_Leiopelmatidae, Alytes_obstetricans_Alytidae, Discoglossus_pictus_Alytidae, Barbourula_busuangensis_Bombinatoridae, Bombina_orientalis_Bombinatoridae, ...
## 
## Rooted; includes branch lengths.</code></pre>
<ol style="list-style-type: decimal">
<li>Use R functions to determine:</li>
</ol>
<ol style="list-style-type: lower-alpha">
<li>how many species are in the tree?</li>
</ol>
<div class="sourceCode" id="cb559"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb559-1"><a href="answers-to-practical-exercises.html#cb559-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the tree structure</span></span>
<span id="cb559-2"><a href="answers-to-practical-exercises.html#cb559-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(frogtree)</span>
<span id="cb559-3"><a href="answers-to-practical-exercises.html#cb559-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Or the whole tree</span></span>
<span id="cb559-4"><a href="answers-to-practical-exercises.html#cb559-4" aria-hidden="true" tabindex="-1"></a>frogtree</span></code></pre></div>
<p><em>There are 214 tips</em></p>
<ol start="2" style="list-style-type: lower-alpha">
<li>is the tree fully resolved?</li>
</ol>
<div class="sourceCode" id="cb560"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb560-1"><a href="answers-to-practical-exercises.html#cb560-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is binary/fully resolved</span></span>
<span id="cb560-2"><a href="answers-to-practical-exercises.html#cb560-2" aria-hidden="true" tabindex="-1"></a><span class="fu">is.binary</span>(frogtree)</span></code></pre></div>
<p><em>Yes</em></p>
<ol start="3" style="list-style-type: lower-alpha">
<li>is the tree rooted?</li>
</ol>
<div class="sourceCode" id="cb561"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb561-1"><a href="answers-to-practical-exercises.html#cb561-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is rooted</span></span>
<span id="cb561-2"><a href="answers-to-practical-exercises.html#cb561-2" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(frogtree)</span></code></pre></div>
<p><em>Yes</em></p>
<ol start="2" style="list-style-type: decimal">
<li>Use <code>plot.phylo</code> to plot the tree.
<ol style="list-style-type: lower-alpha">
<li>Can you change the size of the tip labels?</li>
<li>Can you make a fan-shaped plot?</li>
<li>Can you change the colour of the tips and branches?</li>
</ol></li>
</ol>
<div class="sourceCode" id="cb562"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb562-1"><a href="answers-to-practical-exercises.html#cb562-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the tree</span></span>
<span id="cb562-2"><a href="answers-to-practical-exercises.html#cb562-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(frogtree, </span>
<span id="cb562-3"><a href="answers-to-practical-exercises.html#cb562-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">no.margin =</span> <span class="cn">TRUE</span>, <span class="co"># to make sure it fills the whole space</span></span>
<span id="cb562-4"><a href="answers-to-practical-exercises.html#cb562-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">type =</span> <span class="st">&quot;fan&quot;</span>, <span class="co"># to make a fan tree</span></span>
<span id="cb562-5"><a href="answers-to-practical-exercises.html#cb562-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">cex =</span> <span class="fl">0.6</span>, <span class="co"># to reduce the tip label size (default = 1)</span></span>
<span id="cb562-6"><a href="answers-to-practical-exercises.html#cb562-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">tip.color =</span> <span class="st">&quot;deeppink&quot;</span>, <span class="co"># to change the tip colour</span></span>
<span id="cb562-7"><a href="answers-to-practical-exercises.html#cb562-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">edge.color =</span> <span class="st">&quot;springgreen&quot;</span>) <span class="co"># to change the branch colour</span></span>
<span id="cb562-8"><a href="answers-to-practical-exercises.html#cb562-8" aria-hidden="true" tabindex="-1"></a><span class="co"># yuk!</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Save the tree to file as “mysuperdoopertree.nex”</li>
</ol>
<div class="sourceCode" id="cb563"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb563-1"><a href="answers-to-practical-exercises.html#cb563-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save the tree</span></span>
<span id="cb563-2"><a href="answers-to-practical-exercises.html#cb563-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.nexus</span>(frogtree, <span class="at">file =</span> <span class="st">&quot;data/mysuperdoopertree.nex&quot;</span>)</span></code></pre></div>
</div>
<div id="preparation-8" class="section level2 hasAnchor" number="12.4">
<h2><span class="header-section-number">12.4</span> Preparation<a href="answers-to-practical-exercises.html#preparation-8" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Read in the tree and data then prepare them for a PCM analysis.</p>
<div class="sourceCode" id="cb564"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb564-1"><a href="answers-to-practical-exercises.html#cb564-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load packages</span></span>
<span id="cb564-2"><a href="answers-to-practical-exercises.html#cb564-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb564-3"><a href="answers-to-practical-exercises.html#cb564-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb564-4"><a href="answers-to-practical-exercises.html#cb564-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb564-5"><a href="answers-to-practical-exercises.html#cb564-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phytools)</span>
<span id="cb564-6"><a href="answers-to-practical-exercises.html#cb564-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb564-7"><a href="answers-to-practical-exercises.html#cb564-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the tree</span></span>
<span id="cb564-8"><a href="answers-to-practical-exercises.html#cb564-8" aria-hidden="true" tabindex="-1"></a>primatetree <span class="ot">&lt;-</span> <span class="fu">read.nexus</span>(<span class="st">&quot;data/primate-tree.nex&quot;</span>)</span>
<span id="cb564-9"><a href="answers-to-practical-exercises.html#cb564-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the tree summary</span></span>
<span id="cb564-10"><a href="answers-to-practical-exercises.html#cb564-10" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(primatetree)</span></code></pre></div>
<pre><code>## List of 4
##  $ edge       : int [1:446, 1:2] 227 228 229 230 231 232 233 234 234 235 ...
##  $ edge.length: num [1:446] 4.95 17.69 19.65 8.12 4.82 ...
##  $ Nnode      : int 221
##  $ tip.label  : chr [1:226] &quot;Allenopithecus_nigroviridis&quot; &quot;Cercopithecus_ascanius&quot; &quot;Cercopithecus_cephus&quot; &quot;Cercopithecus_cephus_cephus&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;cladewise&quot;</code></pre>
<div class="sourceCode" id="cb566"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb566-1"><a href="answers-to-practical-exercises.html#cb566-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the tree as a circular/fan phylogeny with small labels</span></span>
<span id="cb566-2"><a href="answers-to-practical-exercises.html#cb566-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(primatetree, <span class="at">cex =</span> <span class="fl">0.2</span>, <span class="at">typ =</span> <span class="st">&quot;fan&quot;</span>, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-359-1.png" width="672" /></p>
<div class="sourceCode" id="cb567"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb567-1"><a href="answers-to-practical-exercises.html#cb567-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is binary</span></span>
<span id="cb567-2"><a href="answers-to-practical-exercises.html#cb567-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE</span></span>
<span id="cb567-3"><a href="answers-to-practical-exercises.html#cb567-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.binary</span>(primatetree) </span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb569"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb569-1"><a href="answers-to-practical-exercises.html#cb569-1" aria-hidden="true" tabindex="-1"></a><span class="co"># It is FALSE so make the tree binary:</span></span>
<span id="cb569-2"><a href="answers-to-practical-exercises.html#cb569-2" aria-hidden="true" tabindex="-1"></a>primatetree <span class="ot">&lt;-</span> <span class="fu">multi2di</span>(primatetree)</span>
<span id="cb569-3"><a href="answers-to-practical-exercises.html#cb569-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb569-4"><a href="answers-to-practical-exercises.html#cb569-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is binary now...</span></span>
<span id="cb569-5"><a href="answers-to-practical-exercises.html#cb569-5" aria-hidden="true" tabindex="-1"></a><span class="fu">is.binary</span>(primatetree) </span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb571"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb571-1"><a href="answers-to-practical-exercises.html#cb571-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is rooted</span></span>
<span id="cb571-2"><a href="answers-to-practical-exercises.html#cb571-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE</span></span>
<span id="cb571-3"><a href="answers-to-practical-exercises.html#cb571-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(primatetree)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb573"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb573-1"><a href="answers-to-practical-exercises.html#cb573-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is ultrametric</span></span>
<span id="cb573-2"><a href="answers-to-practical-exercises.html#cb573-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE, but it is FALSE</span></span>
<span id="cb573-3"><a href="answers-to-practical-exercises.html#cb573-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.ultrametric</span>(primatetree)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb575"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb575-1"><a href="answers-to-practical-exercises.html#cb575-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Force ultrametric</span></span>
<span id="cb575-2"><a href="answers-to-practical-exercises.html#cb575-2" aria-hidden="true" tabindex="-1"></a>primatetree <span class="ot">&lt;-</span> <span class="fu">force.ultrametric</span>(primatetree)</span></code></pre></div>
<pre><code>## ***************************************************************
## *                          Note:                              *
## *    force.ultrametric does not include a formal method to    *
## *    ultrametricize a tree &amp; should only be used to coerce    *
## *   a phylogeny that fails is.ultramtric due to rounding --   *
## *    not as a substitute for formal rate-smoothing methods.   *
## ***************************************************************</code></pre>
<div class="sourceCode" id="cb577"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb577-1"><a href="answers-to-practical-exercises.html#cb577-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the data</span></span>
<span id="cb577-2"><a href="answers-to-practical-exercises.html#cb577-2" aria-hidden="true" tabindex="-1"></a>primatedata <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/primate-data.csv&quot;</span>)</span></code></pre></div>
<pre><code>## Rows: 77 Columns: 9
## ── Column specification ──────────────────────────────────────────────────────────────────────
## Delimiter: &quot;,&quot;
## chr (3): Order, Family, Binomial
## dbl (6): AdultBodyMass_g, GestationLen_d, HomeRange_km2, MaxLongevity_m, SocialGroupSize, ...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div class="sourceCode" id="cb579"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb579-1"><a href="answers-to-practical-exercises.html#cb579-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the data</span></span>
<span id="cb579-2"><a href="answers-to-practical-exercises.html#cb579-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(primatedata)</span></code></pre></div>
<pre><code>## Rows: 77
## Columns: 9
## $ Order           &lt;chr&gt; &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primate…
## $ Family          &lt;chr&gt; &quot;Atelidae&quot;, &quot;Atelidae&quot;, &quot;Atelidae&quot;, &quot;Pitheciidae&quot;, &quot;Cebidae&quot;, &quot;Cebid…
## $ Binomial        &lt;chr&gt; &quot;Ateles belzebuth&quot;, &quot;Ateles geoffroyi&quot;, &quot;Ateles paniscus&quot;, &quot;Calliceb…
## $ AdultBodyMass_g &lt;dbl&gt; 6692.42, 7582.40, 8697.25, 958.13, 558.00, 290.21, 123.94, 7077.66, …
## $ GestationLen_d  &lt;dbl&gt; 138.20, 226.37, 228.18, 164.00, 153.99, 144.00, 134.44, 174.43, 148.…
## $ HomeRange_km2   &lt;dbl&gt; 2.28000, 0.73000, 1.36000, 0.02000, 0.32000, 0.02000, 0.00212, 0.510…
## $ MaxLongevity_m  &lt;dbl&gt; 336.0, 327.6, 453.6, 303.6, 214.8, 201.6, 181.2, 252.0, 339.6, 276.0…
## $ SocialGroupSize &lt;dbl&gt; 14.50, 42.00, 20.00, 2.95, 6.85, 8.55, 6.00, 20.35, 26.30, 11.00, 16…
## $ SocialStatus    &lt;dbl&gt; 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1,…</code></pre>
<div class="sourceCode" id="cb581"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb581-1"><a href="answers-to-practical-exercises.html#cb581-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace spaces with underscores in species names</span></span>
<span id="cb581-2"><a href="answers-to-practical-exercises.html#cb581-2" aria-hidden="true" tabindex="-1"></a>primatedata <span class="ot">&lt;-</span> </span>
<span id="cb581-3"><a href="answers-to-practical-exercises.html#cb581-3" aria-hidden="true" tabindex="-1"></a>  primatedata <span class="sc">%&gt;%</span></span>
<span id="cb581-4"><a href="answers-to-practical-exercises.html#cb581-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Binomial =</span> <span class="fu">str_replace</span>(<span class="at">string =</span> Binomial, <span class="at">pattern =</span> <span class="st">&quot; &quot;</span>, <span class="at">replacement =</span> <span class="st">&quot;_&quot;</span>))</span>
<span id="cb581-5"><a href="answers-to-practical-exercises.html#cb581-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Check it worked</span></span>
<span id="cb581-6"><a href="answers-to-practical-exercises.html#cb581-6" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(primatedata)</span></code></pre></div>
<pre><code>## Rows: 77
## Columns: 9
## $ Order           &lt;chr&gt; &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primate…
## $ Family          &lt;chr&gt; &quot;Atelidae&quot;, &quot;Atelidae&quot;, &quot;Atelidae&quot;, &quot;Pitheciidae&quot;, &quot;Cebidae&quot;, &quot;Cebid…
## $ Binomial        &lt;chr&gt; &quot;Ateles_belzebuth&quot;, &quot;Ateles_geoffroyi&quot;, &quot;Ateles_paniscus&quot;, &quot;Calliceb…
## $ AdultBodyMass_g &lt;dbl&gt; 6692.42, 7582.40, 8697.25, 958.13, 558.00, 290.21, 123.94, 7077.66, …
## $ GestationLen_d  &lt;dbl&gt; 138.20, 226.37, 228.18, 164.00, 153.99, 144.00, 134.44, 174.43, 148.…
## $ HomeRange_km2   &lt;dbl&gt; 2.28000, 0.73000, 1.36000, 0.02000, 0.32000, 0.02000, 0.00212, 0.510…
## $ MaxLongevity_m  &lt;dbl&gt; 336.0, 327.6, 453.6, 303.6, 214.8, 201.6, 181.2, 252.0, 339.6, 276.0…
## $ SocialGroupSize &lt;dbl&gt; 14.50, 42.00, 20.00, 2.95, 6.85, 8.55, 6.00, 20.35, 26.30, 11.00, 16…
## $ SocialStatus    &lt;dbl&gt; 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1,…</code></pre>
<div class="sourceCode" id="cb583"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb583-1"><a href="answers-to-practical-exercises.html#cb583-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the names match in the data and the tree</span></span>
<span id="cb583-2"><a href="answers-to-practical-exercises.html#cb583-2" aria-hidden="true" tabindex="-1"></a>check <span class="ot">&lt;-</span> <span class="fu">name.check</span>(<span class="at">phy =</span> primatetree, <span class="at">data =</span> primatedata, </span>
<span id="cb583-3"><a href="answers-to-practical-exercises.html#cb583-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">data.names =</span> primatedata<span class="sc">$</span>Binomial)</span>
<span id="cb583-4"><a href="answers-to-practical-exercises.html#cb583-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at check</span></span>
<span id="cb583-5"><a href="answers-to-practical-exercises.html#cb583-5" aria-hidden="true" tabindex="-1"></a>check</span></code></pre></div>
<pre><code>## $tree_not_data
##   [1] &quot;Allenopithecus_nigroviridis&quot;                  
##   [2] &quot;Allocebus_trichotis&quot;                          
##   [3] &quot;Alouatta_caraya&quot;                              
##   [4] &quot;Alouatta_sara&quot;                                
##   [5] &quot;Aotus_azarae&quot;                                 
##   [6] &quot;Aotus_azarae_infulatus&quot;                       
##   [7] &quot;Aotus_lemurinus_griseimembra&quot;                 
##   [8] &quot;Aotus_nancymaae&quot;                              
##   [9] &quot;Arctocebus_aureus&quot;                            
##  [10] &quot;Arctocebus_calabarensis&quot;                      
##  [11] &quot;Ateles_fusciceps&quot;                             
##  [12] &quot;Ateles_geoffroyi_ornatus&quot;                     
##  [13] &quot;Ateles_geoffroyi_vellerosus&quot;                  
##  [14] &quot;Avahi_laniger&quot;                                
##  [15] &quot;Avahi_occidentalis&quot;                           
##  [16] &quot;Brachyteles_arachnoides&quot;                      
##  [17] &quot;Bunopithecus_hoolock&quot;                         
##  [18] &quot;Callicebus_donacophilus&quot;                      
##  [19] &quot;Callithrix_(Mico)_emiliae&quot;                    
##  [20] &quot;Callithrix_argentata&quot;                         
##  [21] &quot;Callithrix_aurita&quot;                            
##  [22] &quot;Callithrix_geoffroyi&quot;                         
##  [23] &quot;Callithrix_humeralifera&quot;                      
##  [24] &quot;Callithrix_kuhli&quot;                             
##  [25] &quot;Callithrix_penicillata&quot;                       
##  [26] &quot;Cercocebus_agilis&quot;                            
##  [27] &quot;Cercocebus_atys&quot;                              
##  [28] &quot;Cercocebus_torquatus&quot;                         
##  [29] &quot;Cercopithecus_cephus_cephus&quot;                  
##  [30] &quot;Cercopithecus_cephus_ngottoensis&quot;             
##  [31] &quot;Cercopithecus_diana&quot;                          
##  [32] &quot;Cercopithecus_erythrogaster_erythrogaster&quot;    
##  [33] &quot;Cercopithecus_erythrotis&quot;                     
##  [34] &quot;Cercopithecus_hamlyni&quot;                        
##  [35] &quot;Cercopithecus_lhoesti&quot;                        
##  [36] &quot;Cercopithecus_lowei&quot;                          
##  [37] &quot;Cercopithecus_mona&quot;                           
##  [38] &quot;Cercopithecus_petaurista&quot;                     
##  [39] &quot;Cercopithecus_preussi&quot;                        
##  [40] &quot;Cercopithecus_solatus&quot;                        
##  [41] &quot;Cercopithecus_wolfi&quot;                          
##  [42] &quot;Cheirogaleus_crossleyi&quot;                       
##  [43] &quot;Chlorocebus_aethiops&quot;                         
##  [44] &quot;Chlorocebus_pygerythrus&quot;                      
##  [45] &quot;Chlorocebus_sabaeus&quot;                          
##  [46] &quot;Chlorocebus_tantalus&quot;                         
##  [47] &quot;Colobus_angolensis&quot;                           
##  [48] &quot;Eulemur_albifrons&quot;                            
##  [49] &quot;Eulemur_albocollaris&quot;                         
##  [50] &quot;Eulemur_collaris&quot;                             
##  [51] &quot;Eulemur_macaco_flavifrons&quot;                    
##  [52] &quot;Eulemur_macaco_macaco&quot;                        
##  [53] &quot;Eulemur_rubriventer&quot;                          
##  [54] &quot;Eulemur_rufus&quot;                                
##  [55] &quot;Eulemur_sanfordi&quot;                             
##  [56] &quot;Euoticus_elegantulus&quot;                         
##  [57] &quot;Galago_gallarum&quot;                              
##  [58] &quot;Galago_zanzibaricus&quot;                          
##  [59] &quot;Gorilla_gorilla_gorilla&quot;                      
##  [60] &quot;Hapalemur_alaotrensis&quot;                        
##  [61] &quot;Hapalemur_aureus&quot;                             
##  [62] &quot;Hapalemur_griseus_griseus&quot;                    
##  [63] &quot;Hapalemur_griseus_meridionalis&quot;               
##  [64] &quot;Hapalemur_occidentalis&quot;                       
##  [65] &quot;Homo_sapiens&quot;                                 
##  [66] &quot;Hylobates_agilis&quot;                             
##  [67] &quot;Hylobates_klossii&quot;                            
##  [68] &quot;Hylobates_moloch&quot;                             
##  [69] &quot;Hylobates_muelleri&quot;                           
##  [70] &quot;Indri_indri&quot;                                  
##  [71] &quot;Leontopithecus_chrysomelas&quot;                   
##  [72] &quot;Leontopithecus_chrysopygus&quot;                   
##  [73] &quot;Lepilemur_aeeclis&quot;                            
##  [74] &quot;Lepilemur_ankaranensis&quot;                       
##  [75] &quot;Lepilemur_dorsalis&quot;                           
##  [76] &quot;Lepilemur_edwardsi&quot;                           
##  [77] &quot;Lepilemur_microdon&quot;                           
##  [78] &quot;Lepilemur_mitsinjoensis&quot;                      
##  [79] &quot;Lepilemur_randrianasoli&quot;                      
##  [80] &quot;Lepilemur_ruficaudatus&quot;                       
##  [81] &quot;Lepilemur_sahamalazensis&quot;                     
##  [82] &quot;Lepilemur_seali&quot;                              
##  [83] &quot;Lepilemur_septentrionalis&quot;                    
##  [84] &quot;Lophocebus_aterrimus&quot;                         
##  [85] &quot;Loris_lydekkerianus_grandis&quot;                  
##  [86] &quot;Loris_lydekkerianus_malabaricus&quot;              
##  [87] &quot;Macaca_arctoides&quot;                             
##  [88] &quot;Macaca_assamensis&quot;                            
##  [89] &quot;Macaca_cyclopis&quot;                              
##  [90] &quot;Macaca_hecki&quot;                                 
##  [91] &quot;Macaca_leonina&quot;                               
##  [92] &quot;Macaca_maura&quot;                                 
##  [93] &quot;Macaca_nigra&quot;                                 
##  [94] &quot;Macaca_nigrescens&quot;                            
##  [95] &quot;Macaca_ochreata&quot;                              
##  [96] &quot;Macaca_ochreata_brunnescens&quot;                  
##  [97] &quot;Macaca_pagensis&quot;                              
##  [98] &quot;Macaca_siberu&quot;                                
##  [99] &quot;Macaca_thibetana&quot;                             
## [100] &quot;Macaca_tonkeana&quot;                              
## [101] &quot;Mandrillus_leucophaeus&quot;                       
## [102] &quot;Microcebus_berthae&quot;                           
## [103] &quot;Microcebus_bongolavensis&quot;                     
## [104] &quot;Microcebus_danfossi&quot;                          
## [105] &quot;Microcebus_griseorufus&quot;                       
## [106] &quot;Microcebus_jollyae&quot;                           
## [107] &quot;Microcebus_lehilahytsara&quot;                     
## [108] &quot;Microcebus_lokobensis&quot;                        
## [109] &quot;Microcebus_mittermeieri&quot;                      
## [110] &quot;Microcebus_myoxinus&quot;                          
## [111] &quot;Microcebus_ravelobensis&quot;                      
## [112] &quot;Microcebus_sambiranensis&quot;                     
## [113] &quot;Microcebus_simmonsi&quot;                          
## [114] &quot;Microcebus_tavaratra&quot;                         
## [115] &quot;Nomascus_gabriellae&quot;                          
## [116] &quot;Nomascus_leucogenys&quot;                          
## [117] &quot;Nycticebus_coucang&quot;                           
## [118] &quot;Nycticebus_pygmaeus&quot;                          
## [119] &quot;Pan_troglodytes_schweinfurthii&quot;               
## [120] &quot;Pan_troglodytes_troglodytes&quot;                  
## [121] &quot;Pan_troglodytes_verus&quot;                        
## [122] &quot;Papio_papio&quot;                                  
## [123] &quot;Piliocolobus_badius&quot;                          
## [124] &quot;Pongo_abelii&quot;                                 
## [125] &quot;Pongo_pygmaeus_pygmaeus&quot;                      
## [126] &quot;Presbytis_melalophos&quot;                         
## [127] &quot;Propithecus_coquereli&quot;                        
## [128] &quot;Propithecus_diadema&quot;                          
## [129] &quot;Propithecus_edwardsi&quot;                         
## [130] &quot;Propithecus_tattersalli&quot;                      
## [131] &quot;Pygathrix_nemaeus&quot;                            
## [132] &quot;Rhinopithecus_avunculus&quot;                      
## [133] &quot;Rhinopithecus_bieti&quot;                          
## [134] &quot;Rhinopithecus_brelichi&quot;                       
## [135] &quot;Rhinopithecus_roxellana&quot;                      
## [136] &quot;Rungwecebus_kipunji&quot;                          
## [137] &quot;Saguinus_geoffroyi&quot;                           
## [138] &quot;Saguinus_imperator&quot;                           
## [139] &quot;Saimiri_boliviensis_boliviensis&quot;              
## [140] &quot;Saimiri_oerstedii&quot;                            
## [141] &quot;Trachypithecus_(Trachypithecus)_auratus&quot;      
## [142] &quot;Trachypithecus_(Trachypithecus)_poliocephalus&quot;
## [143] &quot;Trachypithecus_cristatus&quot;                     
## [144] &quot;Trachypithecus_francoisi&quot;                     
## [145] &quot;Trachypithecus_johnii&quot;                        
## [146] &quot;Trachypithecus_phayrei&quot;                       
## [147] &quot;Trachypithecus_pileatus&quot;                      
## [148] &quot;Varecia_rubra&quot;                                
## [149] &quot;Varecia_variegata_variegata&quot;                  
## 
## $data_not_tree
## character(0)</code></pre>
<div class="sourceCode" id="cb585"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb585-1"><a href="answers-to-practical-exercises.html#cb585-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 149 species are in the tree but not the data</span></span>
<span id="cb585-2"><a href="answers-to-practical-exercises.html#cb585-2" aria-hidden="true" tabindex="-1"></a><span class="co"># 0 species are in the data but not the tree</span></span>
<span id="cb585-3"><a href="answers-to-practical-exercises.html#cb585-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb585-4"><a href="answers-to-practical-exercises.html#cb585-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove species missing from the data</span></span>
<span id="cb585-5"><a href="answers-to-practical-exercises.html#cb585-5" aria-hidden="true" tabindex="-1"></a>mytree <span class="ot">&lt;-</span> <span class="fu">drop.tip</span>(primatetree, check<span class="sc">$</span>tree_not_data)</span>
<span id="cb585-6"><a href="answers-to-practical-exercises.html#cb585-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb585-7"><a href="answers-to-practical-exercises.html#cb585-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove species missing from the tree</span></span>
<span id="cb585-8"><a href="answers-to-practical-exercises.html#cb585-8" aria-hidden="true" tabindex="-1"></a>matches <span class="ot">&lt;-</span> <span class="fu">match</span>(primatedata<span class="sc">$</span>Binomial, check<span class="sc">$</span>data_not_tree, <span class="at">nomatch =</span> <span class="dv">0</span>)</span>
<span id="cb585-9"><a href="answers-to-practical-exercises.html#cb585-9" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">subset</span>(primatedata, matches <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb585-10"><a href="answers-to-practical-exercises.html#cb585-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb585-11"><a href="answers-to-practical-exercises.html#cb585-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the tree summary</span></span>
<span id="cb585-12"><a href="answers-to-practical-exercises.html#cb585-12" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(mytree)</span></code></pre></div>
<pre><code>## List of 4
##  $ edge       : int [1:152, 1:2] 78 79 80 81 82 83 84 85 86 87 ...
##  $ edge.length: num [1:152] 4.95 17.69 19.65 8.12 4.82 ...
##  $ Nnode      : int 76
##  $ tip.label  : chr [1:77] &quot;Cercopithecus_ascanius&quot; &quot;Cercopithecus_cephus&quot; &quot;Cercopithecus_mitis&quot; &quot;Cercopithecus_neglectus&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;cladewise&quot;
##  - attr(*, &quot;RSS&quot;)= num 2.08e-08</code></pre>
<div class="sourceCode" id="cb587"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb587-1"><a href="answers-to-practical-exercises.html#cb587-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the data</span></span>
<span id="cb587-2"><a href="answers-to-practical-exercises.html#cb587-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(mydata)</span></code></pre></div>
<pre><code>## Rows: 77
## Columns: 9
## $ Order           &lt;chr&gt; &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primates&quot;, &quot;Primate…
## $ Family          &lt;chr&gt; &quot;Atelidae&quot;, &quot;Atelidae&quot;, &quot;Atelidae&quot;, &quot;Pitheciidae&quot;, &quot;Cebidae&quot;, &quot;Cebid…
## $ Binomial        &lt;chr&gt; &quot;Ateles_belzebuth&quot;, &quot;Ateles_geoffroyi&quot;, &quot;Ateles_paniscus&quot;, &quot;Calliceb…
## $ AdultBodyMass_g &lt;dbl&gt; 6692.42, 7582.40, 8697.25, 958.13, 558.00, 290.21, 123.94, 7077.66, …
## $ GestationLen_d  &lt;dbl&gt; 138.20, 226.37, 228.18, 164.00, 153.99, 144.00, 134.44, 174.43, 148.…
## $ HomeRange_km2   &lt;dbl&gt; 2.28000, 0.73000, 1.36000, 0.02000, 0.32000, 0.02000, 0.00212, 0.510…
## $ MaxLongevity_m  &lt;dbl&gt; 336.0, 327.6, 453.6, 303.6, 214.8, 201.6, 181.2, 252.0, 339.6, 276.0…
## $ SocialGroupSize &lt;dbl&gt; 14.50, 42.00, 20.00, 2.95, 6.85, 8.55, 6.00, 20.35, 26.30, 11.00, 16…
## $ SocialStatus    &lt;dbl&gt; 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1,…</code></pre>
<div class="sourceCode" id="cb589"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb589-1"><a href="answers-to-practical-exercises.html#cb589-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to a dataframe</span></span>
<span id="cb589-2"><a href="answers-to-practical-exercises.html#cb589-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(mydata)</span>
<span id="cb589-3"><a href="answers-to-practical-exercises.html#cb589-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check this is now a data frame</span></span>
<span id="cb589-4"><a href="answers-to-practical-exercises.html#cb589-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mydata)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb591"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb591-1"><a href="answers-to-practical-exercises.html#cb591-1" aria-hidden="true" tabindex="-1"></a><span class="do">## OPTIONAL</span></span>
<span id="cb591-2"><a href="answers-to-practical-exercises.html#cb591-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Write the cleaned data to a new file</span></span>
<span id="cb591-3"><a href="answers-to-practical-exercises.html#cb591-3" aria-hidden="true" tabindex="-1"></a><span class="co"># write_csv(mydata, path = &quot;data/clean-primate-data.csv&quot;)</span></span>
<span id="cb591-4"><a href="answers-to-practical-exercises.html#cb591-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb591-5"><a href="answers-to-practical-exercises.html#cb591-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Write the cleaned tree to a new file</span></span>
<span id="cb591-6"><a href="answers-to-practical-exercises.html#cb591-6" aria-hidden="true" tabindex="-1"></a><span class="co"># write.nexus(mytree, file = &quot;data/clean-primate-tree.nex&quot;)</span></span></code></pre></div>
</div>
<div id="phylogenetic-signal" class="section level2 hasAnchor" number="12.5">
<h2><span class="header-section-number">12.5</span> Phylogenetic signal<a href="answers-to-practical-exercises.html#phylogenetic-signal" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Read in the tree and data, then prepare them for a PCM analysis (you may have already done this in the previous exercise which should save you some time). Then answer the following questions:</p>
<div class="sourceCode" id="cb592"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb592-1"><a href="answers-to-practical-exercises.html#cb592-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the packages</span></span>
<span id="cb592-2"><a href="answers-to-practical-exercises.html#cb592-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb592-3"><a href="answers-to-practical-exercises.html#cb592-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb592-4"><a href="answers-to-practical-exercises.html#cb592-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caper)</span>
<span id="cb592-5"><a href="answers-to-practical-exercises.html#cb592-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb592-6"><a href="answers-to-practical-exercises.html#cb592-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phytools)</span>
<span id="cb592-7"><a href="answers-to-practical-exercises.html#cb592-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb592-8"><a href="answers-to-practical-exercises.html#cb592-8" aria-hidden="true" tabindex="-1"></a><span class="do">### Prepare the tree and the data as shown in the Preparation answer above </span><span class="al">###</span></span>
<span id="cb592-9"><a href="answers-to-practical-exercises.html#cb592-9" aria-hidden="true" tabindex="-1"></a><span class="do">### This should mean you now have mydata and mytree ready to work with </span><span class="al">###</span></span></code></pre></div>
<ol style="list-style-type: decimal">
<li>What is <span class="math inline">\(\lambda\)</span> for log gestation length?</li>
</ol>
<div class="sourceCode" id="cb593"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb593-1"><a href="answers-to-practical-exercises.html#cb593-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First create logGestation containing just log gestation length values</span></span>
<span id="cb593-2"><a href="answers-to-practical-exercises.html#cb593-2" aria-hidden="true" tabindex="-1"></a>logGestation <span class="ot">&lt;-</span> <span class="fu">log</span>(<span class="fu">pull</span>(mydata, GestationLen_d))</span>
<span id="cb593-3"><a href="answers-to-practical-exercises.html#cb593-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Give log Eye names = species names at the tips of the phylogeny</span></span>
<span id="cb593-4"><a href="answers-to-practical-exercises.html#cb593-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(logGestation) <span class="ot">&lt;-</span> mydata<span class="sc">$</span>Binomial</span>
<span id="cb593-5"><a href="answers-to-practical-exercises.html#cb593-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the first few rows</span></span>
<span id="cb593-6"><a href="answers-to-practical-exercises.html#cb593-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(logGestation)</span></code></pre></div>
<pre><code>##   Ateles_belzebuth   Ateles_geoffroyi    Ateles_paniscus  Callicebus_moloch 
##           4.928702           5.422171           5.430135           5.099866 
##  Callimico_goeldii Callithrix_jacchus 
##           5.036888           4.969813</code></pre>
<div class="sourceCode" id="cb595"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb595-1"><a href="answers-to-practical-exercises.html#cb595-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate lambda</span></span>
<span id="cb595-2"><a href="answers-to-practical-exercises.html#cb595-2" aria-hidden="true" tabindex="-1"></a>lambdaGestation <span class="ot">&lt;-</span> <span class="fu">phylosig</span>(mytree, logGestation, <span class="at">method =</span> <span class="st">&quot;lambda&quot;</span>, <span class="at">test =</span> <span class="cn">TRUE</span>)</span>
<span id="cb595-3"><a href="answers-to-practical-exercises.html#cb595-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the results</span></span>
<span id="cb595-4"><a href="answers-to-practical-exercises.html#cb595-4" aria-hidden="true" tabindex="-1"></a>lambdaGestation</span></code></pre></div>
<pre><code>## 
## Phylogenetic signal lambda : 0.948326 
## logL(lambda) : 34.9204 
## LR(lambda=0) : 89.026 
## P-value (based on LR test) : 3.89649e-21</code></pre>
<p><em>0.948</em></p>
<ol start="2" style="list-style-type: decimal">
<li>What is <em>K</em> for log gestation length?</li>
</ol>
<div class="sourceCode" id="cb597"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb597-1"><a href="answers-to-practical-exercises.html#cb597-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate Blomberg’s *K*</span></span>
<span id="cb597-2"><a href="answers-to-practical-exercises.html#cb597-2" aria-hidden="true" tabindex="-1"></a>KGestation <span class="ot">&lt;-</span> <span class="fu">phylosig</span>(mytree, logGestation, <span class="at">method =</span> <span class="st">&quot;K&quot;</span>, <span class="at">test =</span> <span class="cn">TRUE</span>, <span class="at">nsim =</span> <span class="dv">1000</span>)</span>
<span id="cb597-3"><a href="answers-to-practical-exercises.html#cb597-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the results</span></span>
<span id="cb597-4"><a href="answers-to-practical-exercises.html#cb597-4" aria-hidden="true" tabindex="-1"></a>KGestation</span></code></pre></div>
<pre><code>## 
## Phylogenetic signal K : 0.775777 
## P-value (based on 1000 randomizations) : 0.001</code></pre>
<p><em>0.776</em></p>
<ol start="3" style="list-style-type: decimal">
<li>What is D for social status?</li>
</ol>
<div class="sourceCode" id="cb599"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb599-1"><a href="answers-to-practical-exercises.html#cb599-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare the data for use in caper</span></span>
<span id="cb599-2"><a href="answers-to-practical-exercises.html#cb599-2" aria-hidden="true" tabindex="-1"></a>primate <span class="ot">&lt;-</span> <span class="fu">comparative.data</span>(<span class="at">phy =</span> mytree, <span class="at">data =</span> mydata, </span>
<span id="cb599-3"><a href="answers-to-practical-exercises.html#cb599-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">names.col =</span> Binomial, <span class="at">vcv =</span> <span class="cn">TRUE</span>, </span>
<span id="cb599-4"><a href="answers-to-practical-exercises.html#cb599-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">na.omit =</span> <span class="cn">FALSE</span>, <span class="at">warn.dropped =</span> <span class="cn">TRUE</span>)</span>
<span id="cb599-5"><a href="answers-to-practical-exercises.html#cb599-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb599-6"><a href="answers-to-practical-exercises.html#cb599-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Force the phylogeny to have polytomies again nrather than zero length branches</span></span>
<span id="cb599-7"><a href="answers-to-practical-exercises.html#cb599-7" aria-hidden="true" tabindex="-1"></a>primate<span class="sc">$</span>phy <span class="ot">&lt;-</span> <span class="fu">di2multi</span>(primate<span class="sc">$</span>phy)</span>
<span id="cb599-8"><a href="answers-to-practical-exercises.html#cb599-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb599-9"><a href="answers-to-practical-exercises.html#cb599-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate D</span></span>
<span id="cb599-10"><a href="answers-to-practical-exercises.html#cb599-10" aria-hidden="true" tabindex="-1"></a>Dsocial <span class="ot">&lt;-</span> <span class="fu">phylo.d</span>(<span class="at">data =</span> primate, <span class="at">names.col =</span> tiplabel, <span class="at">binvar =</span> SocialStatus, </span>
<span id="cb599-11"><a href="answers-to-practical-exercises.html#cb599-11" aria-hidden="true" tabindex="-1"></a>                  <span class="at">permut =</span> <span class="dv">1000</span>)</span>
<span id="cb599-12"><a href="answers-to-practical-exercises.html#cb599-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb599-13"><a href="answers-to-practical-exercises.html#cb599-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the output</span></span>
<span id="cb599-14"><a href="answers-to-practical-exercises.html#cb599-14" aria-hidden="true" tabindex="-1"></a>Dsocial</span></code></pre></div>
<p><em>About -0.70. This is a randomisation test so may vary slightly</em></p>
</div>
<div id="pgls" class="section level2 hasAnchor" number="12.6">
<h2><span class="header-section-number">12.6</span> PGLS<a href="answers-to-practical-exercises.html#pgls" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Let’s investigate the relationship between gestation length in Primates and their body size. Theory leads us to believe that bigger mammals have longer gestation lengths (compare mice to elephants), but is this true in Primates?</p>
<p>Read in the tree and data, then prepare them for a PCM analysis (you may have already done this in a previous exercise which should save you some time). Fit a PGLS model to investigate the relationship between log gestation length (<code>y = log(GestationLen_d)</code>) and log body size (<code>x = log(AdultBodyMass_g)</code>) in Primates. Don’t forget to look at the model diagnostics!</p>
<div class="sourceCode" id="cb600"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb600-1"><a href="answers-to-practical-exercises.html#cb600-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the packages</span></span>
<span id="cb600-2"><a href="answers-to-practical-exercises.html#cb600-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb600-3"><a href="answers-to-practical-exercises.html#cb600-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb600-4"><a href="answers-to-practical-exercises.html#cb600-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caper)</span>
<span id="cb600-5"><a href="answers-to-practical-exercises.html#cb600-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb600-6"><a href="answers-to-practical-exercises.html#cb600-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb600-7"><a href="answers-to-practical-exercises.html#cb600-7" aria-hidden="true" tabindex="-1"></a><span class="do">### Prepare the tree and the data as shown in the Preparation answer above </span><span class="al">###</span></span>
<span id="cb600-8"><a href="answers-to-practical-exercises.html#cb600-8" aria-hidden="true" tabindex="-1"></a><span class="do">### This should mean you now have mydata and mytree ready to work with </span><span class="al">###</span></span>
<span id="cb600-9"><a href="answers-to-practical-exercises.html#cb600-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb600-10"><a href="answers-to-practical-exercises.html#cb600-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare the data for use in caper</span></span>
<span id="cb600-11"><a href="answers-to-practical-exercises.html#cb600-11" aria-hidden="true" tabindex="-1"></a>primate <span class="ot">&lt;-</span> <span class="fu">comparative.data</span>(<span class="at">phy =</span> mytree, <span class="at">data =</span> mydata, </span>
<span id="cb600-12"><a href="answers-to-practical-exercises.html#cb600-12" aria-hidden="true" tabindex="-1"></a>                            <span class="at">names.col =</span> Binomial, <span class="at">vcv =</span> <span class="cn">TRUE</span>, </span>
<span id="cb600-13"><a href="answers-to-practical-exercises.html#cb600-13" aria-hidden="true" tabindex="-1"></a>                            <span class="at">na.omit =</span> <span class="cn">FALSE</span>, <span class="at">warn.dropped =</span> <span class="cn">TRUE</span>)</span>
<span id="cb600-14"><a href="answers-to-practical-exercises.html#cb600-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb600-15"><a href="answers-to-practical-exercises.html#cb600-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the list of dropped species is what you expected</span></span>
<span id="cb600-16"><a href="answers-to-practical-exercises.html#cb600-16" aria-hidden="true" tabindex="-1"></a>primate<span class="sc">$</span>dropped<span class="sc">$</span>tips</span>
<span id="cb600-17"><a href="answers-to-practical-exercises.html#cb600-17" aria-hidden="true" tabindex="-1"></a>primate<span class="sc">$</span>dropped<span class="sc">$</span>unmatched.rows</span>
<span id="cb600-18"><a href="answers-to-practical-exercises.html#cb600-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb600-19"><a href="answers-to-practical-exercises.html#cb600-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit a PGLS model</span></span>
<span id="cb600-20"><a href="answers-to-practical-exercises.html#cb600-20" aria-hidden="true" tabindex="-1"></a>model.pgls <span class="ot">&lt;-</span> <span class="fu">pgls</span>(<span class="fu">log</span>(GestationLen_d) <span class="sc">~</span> <span class="fu">log</span>(AdultBodyMass_g), </span>
<span id="cb600-21"><a href="answers-to-practical-exercises.html#cb600-21" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> primate, <span class="at">lambda =</span> <span class="st">&quot;ML&quot;</span>, <span class="at">bounds =</span> <span class="fu">list</span>(<span class="at">lambda =</span> <span class="fu">c</span>(<span class="fl">0.01</span>, <span class="dv">1</span>)))</span>
<span id="cb600-22"><a href="answers-to-practical-exercises.html#cb600-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb600-23"><a href="answers-to-practical-exercises.html#cb600-23" aria-hidden="true" tabindex="-1"></a><span class="do">### Model diagnostics</span></span>
<span id="cb600-24"><a href="answers-to-practical-exercises.html#cb600-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a plotting window with 4 panes as there are four plots</span></span>
<span id="cb600-25"><a href="answers-to-practical-exercises.html#cb600-25" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb600-26"><a href="answers-to-practical-exercises.html#cb600-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Now plot the model diagnostics</span></span>
<span id="cb600-27"><a href="answers-to-practical-exercises.html#cb600-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(model.pgls)</span>
<span id="cb600-28"><a href="answers-to-practical-exercises.html#cb600-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Return the plot window to one pane for later plotting</span></span>
<span id="cb600-29"><a href="answers-to-practical-exercises.html#cb600-29" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb600-30"><a href="answers-to-practical-exercises.html#cb600-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb600-31"><a href="answers-to-practical-exercises.html#cb600-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the model outputs</span></span>
<span id="cb600-32"><a href="answers-to-practical-exercises.html#cb600-32" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model.pgls)</span>
<span id="cb600-33"><a href="answers-to-practical-exercises.html#cb600-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb600-34"><a href="answers-to-practical-exercises.html#cb600-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the results</span></span>
<span id="cb600-35"><a href="answers-to-practical-exercises.html#cb600-35" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mydata, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log</span>(AdultBodyMass_g), </span>
<span id="cb600-36"><a href="answers-to-practical-exercises.html#cb600-36" aria-hidden="true" tabindex="-1"></a>                     <span class="at">y =</span> <span class="fu">log</span>(GestationLen_d))) <span class="sc">+</span></span>
<span id="cb600-37"><a href="answers-to-practical-exercises.html#cb600-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb600-38"><a href="answers-to-practical-exercises.html#cb600-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">slope =</span> <span class="fu">coefficients</span>(model.pgls)[<span class="dv">2</span>], </span>
<span id="cb600-39"><a href="answers-to-practical-exercises.html#cb600-39" aria-hidden="true" tabindex="-1"></a>              <span class="at">intercept =</span> <span class="fu">coefficients</span>(model.pgls)[<span class="dv">1</span>]) <span class="sc">+</span></span>
<span id="cb600-40"><a href="answers-to-practical-exercises.html#cb600-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p>Then answer the following questions.</p>
<ol style="list-style-type: decimal">
<li>What is the maximum likelihood estimate of <span class="math inline">\(\lambda\)</span> in the model?</li>
</ol>
<div class="sourceCode" id="cb601"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb601-1"><a href="answers-to-practical-exercises.html#cb601-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the model outputs. Lambda is near the middle</span></span>
<span id="cb601-2"><a href="answers-to-practical-exercises.html#cb601-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model.pgls)</span></code></pre></div>
<p><em>0.892</em></p>
<ol start="2" style="list-style-type: decimal">
<li>What are the confidence intervals on the maximum likelihood estimate of <span class="math inline">\(\lambda\)</span>. How would you interpret these?</li>
</ol>
<div class="sourceCode" id="cb602"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb602-1"><a href="answers-to-practical-exercises.html#cb602-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the confidence intervals on lambda</span></span>
<span id="cb602-2"><a href="answers-to-practical-exercises.html#cb602-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pgls.confint</span>(model.pgls, <span class="st">&quot;lambda&quot;</span>)<span class="sc">$</span>ci.val</span></code></pre></div>
<p><em>0.753434, 0.966543. We are 95% certain that <span class="math inline">\(\lambda\)</span> is between these two values.</em></p>
<ol start="3" style="list-style-type: decimal">
<li>Plot the <span class="math inline">\(\lambda\)</span> profile for the the maximum likelihood estimate of <span class="math inline">\(\lambda\)</span>. What does it look like?</li>
</ol>
<div class="sourceCode" id="cb603"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb603-1"><a href="answers-to-practical-exercises.html#cb603-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a likelihood profile of the lambda estimate</span></span>
<span id="cb603-2"><a href="answers-to-practical-exercises.html#cb603-2" aria-hidden="true" tabindex="-1"></a>lambda.profile <span class="ot">&lt;-</span> <span class="fu">pgls.profile</span>(model.pgls, <span class="st">&quot;lambda&quot;</span>)</span>
<span id="cb603-3"><a href="answers-to-practical-exercises.html#cb603-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the likelihood profile</span></span>
<span id="cb603-4"><a href="answers-to-practical-exercises.html#cb603-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(lambda.profile)</span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>Is there a significant relationship between log gestation length and log body size? What is the slope of this relationship?</li>
</ol>
<div class="sourceCode" id="cb604"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb604-1"><a href="answers-to-practical-exercises.html#cb604-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the model outputs. </span></span>
<span id="cb604-2"><a href="answers-to-practical-exercises.html#cb604-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model.pgls)</span></code></pre></div>
<p><em>Yes. p &lt; 0.001. Slope = 0.105 <span class="math inline">\(\pm\)</span> 0.019</em></p>
</div>
<div id="mcmcglmm" class="section level2 hasAnchor" number="12.7">
<h2><span class="header-section-number">12.7</span> MCMCglmm<a href="answers-to-practical-exercises.html#mcmcglmm" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Let’s repeat the analyses we did in the PGLS chapter but using <code>MCMCglmm</code>. We will investigate the relationship between gestation length in Primates and their body size.</p>
<p>Read in the tree and data, then prepare them for a PCM analysis (you may have already done this in a previous exercise which should save you some time). Fit a MCMCglmm model to investigate the relationship between log gestation length (<code>y = log(GestationLen_d)</code>) and log body size (<code>x = log(AdultBodyMass_g)</code>) in Primates. Don’t forget to check that the model has converged, and to look at the model diagnostics and for autocorrelation.</p>
<div class="sourceCode" id="cb605"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb605-1"><a href="answers-to-practical-exercises.html#cb605-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the packages</span></span>
<span id="cb605-2"><a href="answers-to-practical-exercises.html#cb605-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb605-3"><a href="answers-to-practical-exercises.html#cb605-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb605-4"><a href="answers-to-practical-exercises.html#cb605-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb605-5"><a href="answers-to-practical-exercises.html#cb605-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MCMCglmm)</span>
<span id="cb605-6"><a href="answers-to-practical-exercises.html#cb605-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(coda)</span>
<span id="cb605-7"><a href="answers-to-practical-exercises.html#cb605-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-8"><a href="answers-to-practical-exercises.html#cb605-8" aria-hidden="true" tabindex="-1"></a><span class="do">## MCMCglmm specific preparation</span></span>
<span id="cb605-9"><a href="answers-to-practical-exercises.html#cb605-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-10"><a href="answers-to-practical-exercises.html#cb605-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove zero length branches and replace with polytomies</span></span>
<span id="cb605-11"><a href="answers-to-practical-exercises.html#cb605-11" aria-hidden="true" tabindex="-1"></a>mytree2 <span class="ot">&lt;-</span> <span class="fu">di2multi</span>(mytree)</span>
<span id="cb605-12"><a href="answers-to-practical-exercises.html#cb605-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-13"><a href="answers-to-practical-exercises.html#cb605-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove node labels </span></span>
<span id="cb605-14"><a href="answers-to-practical-exercises.html#cb605-14" aria-hidden="true" tabindex="-1"></a>mytree2<span class="sc">$</span>node.label <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb605-15"><a href="answers-to-practical-exercises.html#cb605-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-16"><a href="answers-to-practical-exercises.html#cb605-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the inverse vcv matrix for the phylogeny</span></span>
<span id="cb605-17"><a href="answers-to-practical-exercises.html#cb605-17" aria-hidden="true" tabindex="-1"></a>inv.phylo <span class="ot">&lt;-</span> <span class="fu">inverseA</span>(mytree2, <span class="at">nodes =</span> <span class="st">&quot;TIPS&quot;</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>Ainv</span>
<span id="cb605-18"><a href="answers-to-practical-exercises.html#cb605-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-19"><a href="answers-to-practical-exercises.html#cb605-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up priors for MCMCglmm</span></span>
<span id="cb605-20"><a href="answers-to-practical-exercises.html#cb605-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Inverse Wishart with V = 1 and nu = 0.02</span></span>
<span id="cb605-21"><a href="answers-to-practical-exercises.html#cb605-21" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. fairly uninformative priors</span></span>
<span id="cb605-22"><a href="answers-to-practical-exercises.html#cb605-22" aria-hidden="true" tabindex="-1"></a>prior <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">G =</span> <span class="fu">list</span>(<span class="at">G1 =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">1</span>, <span class="at">nu =</span> <span class="fl">0.02</span>)),</span>
<span id="cb605-23"><a href="answers-to-practical-exercises.html#cb605-23" aria-hidden="true" tabindex="-1"></a>              <span class="at">R =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">1</span>, <span class="at">nu =</span> <span class="fl">0.02</span>))</span>
<span id="cb605-24"><a href="answers-to-practical-exercises.html#cb605-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-25"><a href="answers-to-practical-exercises.html#cb605-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Set number of iterations</span></span>
<span id="cb605-26"><a href="answers-to-practical-exercises.html#cb605-26" aria-hidden="true" tabindex="-1"></a>nitt <span class="ot">&lt;-</span> <span class="dv">1000000</span></span>
<span id="cb605-27"><a href="answers-to-practical-exercises.html#cb605-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-28"><a href="answers-to-practical-exercises.html#cb605-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Set burnin</span></span>
<span id="cb605-29"><a href="answers-to-practical-exercises.html#cb605-29" aria-hidden="true" tabindex="-1"></a>burnin <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb605-30"><a href="answers-to-practical-exercises.html#cb605-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-31"><a href="answers-to-practical-exercises.html#cb605-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Sampling interval</span></span>
<span id="cb605-32"><a href="answers-to-practical-exercises.html#cb605-32" aria-hidden="true" tabindex="-1"></a>thin <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb605-33"><a href="answers-to-practical-exercises.html#cb605-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-34"><a href="answers-to-practical-exercises.html#cb605-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit MCMCglmm model</span></span>
<span id="cb605-35"><a href="answers-to-practical-exercises.html#cb605-35" aria-hidden="true" tabindex="-1"></a>model_mcmcglmm <span class="ot">&lt;-</span> <span class="fu">MCMCglmm</span>(<span class="fu">log</span>(GestationLen_d) <span class="sc">~</span> <span class="fu">log</span>(AdultBodyMass_g), </span>
<span id="cb605-36"><a href="answers-to-practical-exercises.html#cb605-36" aria-hidden="true" tabindex="-1"></a>                           <span class="at">data =</span> mydata, </span>
<span id="cb605-37"><a href="answers-to-practical-exercises.html#cb605-37" aria-hidden="true" tabindex="-1"></a>                           <span class="at">random =</span> <span class="sc">~</span> Binomial,</span>
<span id="cb605-38"><a href="answers-to-practical-exercises.html#cb605-38" aria-hidden="true" tabindex="-1"></a>                           <span class="at">ginverse =</span> <span class="fu">list</span>(<span class="at">Binomial =</span> inv.phylo), </span>
<span id="cb605-39"><a href="answers-to-practical-exercises.html#cb605-39" aria-hidden="true" tabindex="-1"></a>                           <span class="at">prior =</span> prior,</span>
<span id="cb605-40"><a href="answers-to-practical-exercises.html#cb605-40" aria-hidden="true" tabindex="-1"></a>                           <span class="at">nitt =</span> nitt, <span class="at">thin =</span> thin, <span class="at">burnin =</span> burnin,</span>
<span id="cb605-41"><a href="answers-to-practical-exercises.html#cb605-41" aria-hidden="true" tabindex="-1"></a>                           <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb605-42"><a href="answers-to-practical-exercises.html#cb605-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb605-43"><a href="answers-to-practical-exercises.html#cb605-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Save the model</span></span>
<span id="cb605-44"><a href="answers-to-practical-exercises.html#cb605-44" aria-hidden="true" tabindex="-1"></a><span class="co"># saveRDS(model_mcmcglmm, file = &quot;data/model_mcmcglmm_primate.rda&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb606"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb606-1"><a href="answers-to-practical-exercises.html#cb606-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the saved model output run in advance to save time</span></span>
<span id="cb606-2"><a href="answers-to-practical-exercises.html#cb606-2" aria-hidden="true" tabindex="-1"></a>model_mcmcglmm <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/model_mcmcglmm_primate.rda&quot;</span>)</span>
<span id="cb606-3"><a href="answers-to-practical-exercises.html#cb606-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb606-4"><a href="answers-to-practical-exercises.html#cb606-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Did the model converge?</span></span>
<span id="cb606-5"><a href="answers-to-practical-exercises.html#cb606-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot model diagnostics for MCMCglmm</span></span>
<span id="cb606-6"><a href="answers-to-practical-exercises.html#cb606-6" aria-hidden="true" tabindex="-1"></a><span class="co"># For fixed effects</span></span>
<span id="cb606-7"><a href="answers-to-practical-exercises.html#cb606-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(model_mcmcglmm<span class="sc">$</span>Sol)  </span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-370-1.png" width="672" /></p>
<div class="sourceCode" id="cb607"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb607-1"><a href="answers-to-practical-exercises.html#cb607-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot model diagnostics for MCMCglmm</span></span>
<span id="cb607-2"><a href="answers-to-practical-exercises.html#cb607-2" aria-hidden="true" tabindex="-1"></a><span class="co"># For random effects</span></span>
<span id="cb607-3"><a href="answers-to-practical-exercises.html#cb607-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(model_mcmcglmm<span class="sc">$</span>VCV)  </span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-370-2.png" width="672" /></p>
<div class="sourceCode" id="cb608"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb608-1"><a href="answers-to-practical-exercises.html#cb608-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate ESS</span></span>
<span id="cb608-2"><a href="answers-to-practical-exercises.html#cb608-2" aria-hidden="true" tabindex="-1"></a><span class="fu">effectiveSize</span>(model_mcmcglmm<span class="sc">$</span>Sol[, <span class="dv">1</span><span class="sc">:</span>model_mcmcglmm<span class="sc">$</span>Fixed<span class="sc">$</span>nfl, </span>
<span id="cb608-3"><a href="answers-to-practical-exercises.html#cb608-3" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">drop =</span> <span class="cn">FALSE</span>])[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## [1] 1998</code></pre>
<div class="sourceCode" id="cb610"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb610-1"><a href="answers-to-practical-exercises.html#cb610-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look for autocorrelation</span></span>
<span id="cb610-2"><a href="answers-to-practical-exercises.html#cb610-2" aria-hidden="true" tabindex="-1"></a><span class="fu">autocorr</span>(model_mcmcglmm<span class="sc">$</span>VCV)</span></code></pre></div>
<pre><code>## , , Binomial
## 
##                Binomial        units
## Lag 0      1.0000000000 -0.301495441
## Lag 500    0.0008200833 -0.004268382
## Lag 2500   0.0052744916  0.038062203
## Lag 5000  -0.0446744855  0.014890222
## Lag 25000  0.0104991347 -0.008261472
## 
## , , units
## 
##               Binomial        units
## Lag 0     -0.301495441  1.000000000
## Lag 500    0.012393397  0.002583728
## Lag 2500   0.001343481 -0.015740264
## Lag 5000   0.039810907 -0.009084549
## Lag 25000  0.029927932 -0.018984469</code></pre>
<ol style="list-style-type: decimal">
<li>What is <span class="math inline">\(\lambda\)</span> in the model?</li>
</ol>
<div class="sourceCode" id="cb612"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb612-1"><a href="answers-to-practical-exercises.html#cb612-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get lambda</span></span>
<span id="cb612-2"><a href="answers-to-practical-exercises.html#cb612-2" aria-hidden="true" tabindex="-1"></a>lambda <span class="ot">&lt;-</span> model_mcmcglmm<span class="sc">$</span>VCV[,<span class="st">&#39;Binomial&#39;</span>]<span class="sc">/</span></span>
<span id="cb612-3"><a href="answers-to-practical-exercises.html#cb612-3" aria-hidden="true" tabindex="-1"></a>          (model_mcmcglmm<span class="sc">$</span>VCV[,<span class="st">&#39;Binomial&#39;</span>] <span class="sc">+</span> model_mcmcglmm<span class="sc">$</span>VCV[,<span class="st">&#39;units&#39;</span>])</span>
<span id="cb612-4"><a href="answers-to-practical-exercises.html#cb612-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb612-5"><a href="answers-to-practical-exercises.html#cb612-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Posterior mean of lambda</span></span>
<span id="cb612-6"><a href="answers-to-practical-exercises.html#cb612-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(lambda)</span></code></pre></div>
<pre><code>## [1] 0.8699022</code></pre>
<div class="sourceCode" id="cb614"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb614-1"><a href="answers-to-practical-exercises.html#cb614-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Posterior mode of lambda</span></span>
<span id="cb614-2"><a href="answers-to-practical-exercises.html#cb614-2" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior.mode</span>(lambda)</span></code></pre></div>
<pre><code>##      var1 
## 0.8925655</code></pre>
<p><em>Posterior mean lambda is 0.870, the posterior mode is 0.893</em></p>
<ol start="2" style="list-style-type: decimal">
<li>What are the 95% credible intervals (HPD) around <span class="math inline">\(\lambda\)</span>?</li>
</ol>
<div class="sourceCode" id="cb616"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb616-1"><a href="answers-to-practical-exercises.html#cb616-1" aria-hidden="true" tabindex="-1"></a><span class="co"># HPD interval</span></span>
<span id="cb616-2"><a href="answers-to-practical-exercises.html#cb616-2" aria-hidden="true" tabindex="-1"></a><span class="fu">HPDinterval</span>(lambda)</span></code></pre></div>
<pre><code>##          lower     upper
## var1 0.7615502 0.9613453
## attr(,&quot;Probability&quot;)
## [1] 0.9499499</code></pre>
<p><em>HPD interval of lambda is 0.762 to 0.961</em></p>
<ol start="3" style="list-style-type: decimal">
<li>Plot the <span class="math inline">\(\lambda\)</span> profile for the the maximum likelihood estimate of <span class="math inline">\(\lambda\)</span>.</li>
</ol>
<div class="sourceCode" id="cb618"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb618-1"><a href="answers-to-practical-exercises.html#cb618-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the posterior for lambda</span></span>
<span id="cb618-2"><a href="answers-to-practical-exercises.html#cb618-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(lambda), <span class="fu">aes</span>(<span class="at">x =</span> lambda))<span class="sc">+</span></span>
<span id="cb618-3"><a href="answers-to-practical-exercises.html#cb618-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb618-4"><a href="answers-to-practical-exercises.html#cb618-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">posterior.mode</span>(lambda), <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb618-5"><a href="answers-to-practical-exercises.html#cb618-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">HPDinterval</span>(lambda)[[<span class="dv">1</span>]], </span>
<span id="cb618-6"><a href="answers-to-practical-exercises.html#cb618-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb618-7"><a href="answers-to-practical-exercises.html#cb618-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">HPDinterval</span>(lambda)[[<span class="dv">2</span>]], </span>
<span id="cb618-8"><a href="answers-to-practical-exercises.html#cb618-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb618-9"><a href="answers-to-practical-exercises.html#cb618-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<pre><code>## Don&#39;t know how to automatically pick scale for object of type &lt;mcmc&gt;. Defaulting to
## continuous.</code></pre>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-373-1.png" width="672" /></p>
<ol start="4" style="list-style-type: decimal">
<li>Is there a relationship between log gestation length and log body size? What is the slope of this relationship?</li>
</ol>
<div class="sourceCode" id="cb620"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb620-1"><a href="answers-to-practical-exercises.html#cb620-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Results</span></span>
<span id="cb620-2"><a href="answers-to-practical-exercises.html#cb620-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at summary of fixed effect results</span></span>
<span id="cb620-3"><a href="answers-to-practical-exercises.html#cb620-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model_mcmcglmm)<span class="sc">$</span>solutions</span></code></pre></div>
<pre><code>##                      post.mean   l-95% CI  u-95% CI eff.samp        pMCMC
## (Intercept)          4.2876174 3.97118491 4.6374322     1998 0.0005005005
## log(AdultBodyMass_g) 0.1050013 0.06455035 0.1460903     1998 0.0005005005</code></pre>
<div class="sourceCode" id="cb622"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb622-1"><a href="answers-to-practical-exercises.html#cb622-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at G-structure and R-structure posterior modes</span></span>
<span id="cb622-2"><a href="answers-to-practical-exercises.html#cb622-2" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior.mode</span>(model_mcmcglmm<span class="sc">$</span>VCV)</span></code></pre></div>
<pre><code>##    Binomial       units 
## 0.040772793 0.005794703</code></pre>
<div class="sourceCode" id="cb624"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb624-1"><a href="answers-to-practical-exercises.html#cb624-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at G-structure and R-structure 95% credible intervals</span></span>
<span id="cb624-2"><a href="answers-to-practical-exercises.html#cb624-2" aria-hidden="true" tabindex="-1"></a><span class="fu">HPDinterval</span>(model_mcmcglmm<span class="sc">$</span>VCV)</span></code></pre></div>
<pre><code>##                lower      upper
## Binomial 0.021550421 0.06927166
## units    0.002813069 0.01043645
## attr(,&quot;Probability&quot;)
## [1] 0.9499499</code></pre>
<p><em>Yes! The slope is 0.105 (0.065-0.146). This is close to what we got using PGLS.</em></p>
</div>
<div id="models-of-evolution-with-continuous-traits" class="section level2 hasAnchor" number="12.8">
<h2><span class="header-section-number">12.8</span> Models of evolution with continuous traits<a href="answers-to-practical-exercises.html#models-of-evolution-with-continuous-traits" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Read in the tree and data, then prepare them for a PCM analysis (you may have already done this in a previous exercise which should save you some time). Fit Brownian and OU models to log adult body mass for Primates (<code>AdultBodyMass_g</code>). Then answer the following questions.</p>
<div class="sourceCode" id="cb626"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb626-1"><a href="answers-to-practical-exercises.html#cb626-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the packages</span></span>
<span id="cb626-2"><a href="answers-to-practical-exercises.html#cb626-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb626-3"><a href="answers-to-practical-exercises.html#cb626-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb626-4"><a href="answers-to-practical-exercises.html#cb626-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caper)</span>
<span id="cb626-5"><a href="answers-to-practical-exercises.html#cb626-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb626-6"><a href="answers-to-practical-exercises.html#cb626-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb626-7"><a href="answers-to-practical-exercises.html#cb626-7" aria-hidden="true" tabindex="-1"></a><span class="do">### Prepare the tree and the data as shown in the Preparation answer above </span><span class="al">###</span></span>
<span id="cb626-8"><a href="answers-to-practical-exercises.html#cb626-8" aria-hidden="true" tabindex="-1"></a><span class="do">### This should mean you now have mydata and mytree ready to work with </span><span class="al">###</span></span>
<span id="cb626-9"><a href="answers-to-practical-exercises.html#cb626-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb626-10"><a href="answers-to-practical-exercises.html#cb626-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder the data so the tree and data are in the same order</span></span>
<span id="cb626-11"><a href="answers-to-practical-exercises.html#cb626-11" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> mydata[<span class="fu">match</span>(mytree<span class="sc">$</span>tip.label, mydata<span class="sc">$</span>Binomial), ]</span>
<span id="cb626-12"><a href="answers-to-practical-exercises.html#cb626-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb626-13"><a href="answers-to-practical-exercises.html#cb626-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Create logBodyMass containing just log body mass values</span></span>
<span id="cb626-14"><a href="answers-to-practical-exercises.html#cb626-14" aria-hidden="true" tabindex="-1"></a>logBodyMass <span class="ot">&lt;-</span> <span class="fu">log</span>(<span class="fu">pull</span>(mydata, AdultBodyMass_g))</span>
<span id="cb626-15"><a href="answers-to-practical-exercises.html#cb626-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Give log body mass names = species names at the tips of the phylogeny</span></span>
<span id="cb626-16"><a href="answers-to-practical-exercises.html#cb626-16" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(logBodyMass) <span class="ot">&lt;-</span> mydata<span class="sc">$</span>Binomial</span>
<span id="cb626-17"><a href="answers-to-practical-exercises.html#cb626-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the first few rows</span></span>
<span id="cb626-18"><a href="answers-to-practical-exercises.html#cb626-18" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(logBodyMass)</span></code></pre></div>
<pre><code>##  Cercopithecus_ascanius    Cercopithecus_cephus     Cercopithecus_mitis 
##                8.171950                8.144644                8.525417 
## Cercopithecus_neglectus Cercopithecus_nictitans      Erythrocebus_patas 
##                8.580078                8.567299                8.982975</code></pre>
<div class="sourceCode" id="cb628"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb628-1"><a href="answers-to-practical-exercises.html#cb628-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the Brownian model</span></span>
<span id="cb628-2"><a href="answers-to-practical-exercises.html#cb628-2" aria-hidden="true" tabindex="-1"></a>BM <span class="ot">&lt;-</span> <span class="fu">fitContinuous</span>(mytree, logBodyMass, <span class="at">model =</span> <span class="fu">c</span>(<span class="st">&quot;BM&quot;</span>))</span>
<span id="cb628-3"><a href="answers-to-practical-exercises.html#cb628-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the OU model</span></span>
<span id="cb628-4"><a href="answers-to-practical-exercises.html#cb628-4" aria-hidden="true" tabindex="-1"></a>OU <span class="ot">&lt;-</span> <span class="fu">fitContinuous</span>(mytree, logBodyMass, <span class="at">model =</span> <span class="fu">c</span>(<span class="st">&quot;OU&quot;</span>))</span></code></pre></div>
<ol style="list-style-type: decimal">
<li>Which model fits best?</li>
</ol>
<div class="sourceCode" id="cb629"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb629-1"><a href="answers-to-practical-exercises.html#cb629-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Could look at AIC weights...</span></span>
<span id="cb629-2"><a href="answers-to-practical-exercises.html#cb629-2" aria-hidden="true" tabindex="-1"></a>aic.scores <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">c</span>(BM<span class="sc">$</span>opt<span class="sc">$</span>aic, OU<span class="sc">$</span>opt<span class="sc">$</span>aic), <span class="fu">c</span>(<span class="st">&quot;BM&quot;</span>,<span class="st">&quot;OU&quot;</span>))</span>
<span id="cb629-3"><a href="answers-to-practical-exercises.html#cb629-3" aria-hidden="true" tabindex="-1"></a><span class="fu">aicw</span>(aic.scores)</span></code></pre></div>
<pre><code>##         fit delta         w
## BM 160.1921     0 0.7310586
## OU 162.1921     2 0.2689414</code></pre>
<div class="sourceCode" id="cb631"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb631-1"><a href="answers-to-practical-exercises.html#cb631-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or just the raw AIC values</span></span>
<span id="cb631-2"><a href="answers-to-practical-exercises.html#cb631-2" aria-hidden="true" tabindex="-1"></a>BM<span class="sc">$</span>opt<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 160.1921</code></pre>
<div class="sourceCode" id="cb633"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb633-1"><a href="answers-to-practical-exercises.html#cb633-1" aria-hidden="true" tabindex="-1"></a>OU<span class="sc">$</span>opt<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 162.1921</code></pre>
<p><em>BM but only by a small amount</em></p>
<ol start="2" style="list-style-type: decimal">
<li>What is the rate of evolution for the Brownian model?</li>
</ol>
<div class="sourceCode" id="cb635"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb635-1"><a href="answers-to-practical-exercises.html#cb635-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the whole output</span></span>
<span id="cb635-2"><a href="answers-to-practical-exercises.html#cb635-2" aria-hidden="true" tabindex="-1"></a>BM</span></code></pre></div>
<pre><code>## GEIGER-fitted comparative model of continuous data
##  fitted &#39;BM&#39; model parameters:
##  sigsq = 0.028655
##  z0 = 6.773956
## 
##  model summary:
##  log-likelihood = -78.096042
##  AIC = 160.192084
##  AICc = 160.354246
##  free parameters = 2
## 
## Convergence diagnostics:
##  optimization iterations = 100
##  failed iterations = 0
##  number of iterations with same best fit = 100
##  frequency of best fit = 1.00
## 
##  object summary:
##  &#39;lik&#39; -- likelihood function
##  &#39;bnd&#39; -- bounds for likelihood search
##  &#39;res&#39; -- optimization iteration summary
##  &#39;opt&#39; -- maximum likelihood parameter estimates</code></pre>
<div class="sourceCode" id="cb637"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb637-1"><a href="answers-to-practical-exercises.html#cb637-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or just the one parameter</span></span>
<span id="cb637-2"><a href="answers-to-practical-exercises.html#cb637-2" aria-hidden="true" tabindex="-1"></a>BM<span class="sc">$</span>opt<span class="sc">$</span>sigsq</span></code></pre></div>
<pre><code>## [1] 0.0286548</code></pre>
<p><em>0.028655</em></p>
<ol start="3" style="list-style-type: decimal">
<li>What is the strength of pull towards the long-term mean in the OU model? What does this value suggest?</li>
</ol>
<div class="sourceCode" id="cb639"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb639-1"><a href="answers-to-practical-exercises.html#cb639-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the whole output</span></span>
<span id="cb639-2"><a href="answers-to-practical-exercises.html#cb639-2" aria-hidden="true" tabindex="-1"></a>OU</span></code></pre></div>
<pre><code>## GEIGER-fitted comparative model of continuous data
##  fitted &#39;OU&#39; model parameters:
##  alpha = 0.000000
##  sigsq = 0.028655
##  z0 = 6.773956
## 
##  model summary:
##  log-likelihood = -78.096042
##  AIC = 162.192084
##  AICc = 162.520851
##  free parameters = 3
## 
## Convergence diagnostics:
##  optimization iterations = 100
##  failed iterations = 0
##  number of iterations with same best fit = 50
##  frequency of best fit = 0.50
## 
##  object summary:
##  &#39;lik&#39; -- likelihood function
##  &#39;bnd&#39; -- bounds for likelihood search
##  &#39;res&#39; -- optimization iteration summary
##  &#39;opt&#39; -- maximum likelihood parameter estimates</code></pre>
<div class="sourceCode" id="cb641"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb641-1"><a href="answers-to-practical-exercises.html#cb641-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or just the one parameter</span></span>
<span id="cb641-2"><a href="answers-to-practical-exercises.html#cb641-2" aria-hidden="true" tabindex="-1"></a>OU<span class="sc">$</span>opt<span class="sc">$</span>alpha</span></code></pre></div>
<pre><code>## [1] 2.571618e-42</code></pre>
<p><em>&lt;0.0001, i.e. very close to zero which makes it very close to a Brownian model</em></p>
</div>
<div id="models-of-evolution-with-discrete-traits" class="section level2 hasAnchor" number="12.9">
<h2><span class="header-section-number">12.9</span> Models of evolution with discrete traits<a href="answers-to-practical-exercises.html#models-of-evolution-with-discrete-traits" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Read in the tree and data, then prepare them for a PCM analysis (you may have already done this in a previous exercise which should save you some time). Fit equal rates, symmetric rates and all rates different models to the social status variable for Primates (<code>SocialStatus</code>). Note that social status has two states: non-social (1) or social (2).</p>
<div class="sourceCode" id="cb643"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb643-1"><a href="answers-to-practical-exercises.html#cb643-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the packages</span></span>
<span id="cb643-2"><a href="answers-to-practical-exercises.html#cb643-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb643-3"><a href="answers-to-practical-exercises.html#cb643-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb643-4"><a href="answers-to-practical-exercises.html#cb643-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caper)</span>
<span id="cb643-5"><a href="answers-to-practical-exercises.html#cb643-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb643-6"><a href="answers-to-practical-exercises.html#cb643-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phytools)</span>
<span id="cb643-7"><a href="answers-to-practical-exercises.html#cb643-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb643-8"><a href="answers-to-practical-exercises.html#cb643-8" aria-hidden="true" tabindex="-1"></a><span class="do">### Prepare the tree and the data as shown in the Preparation answer above </span><span class="al">###</span></span>
<span id="cb643-9"><a href="answers-to-practical-exercises.html#cb643-9" aria-hidden="true" tabindex="-1"></a><span class="do">### This should mean you now have mydata and mytree ready to work with </span><span class="al">###</span></span>
<span id="cb643-10"><a href="answers-to-practical-exercises.html#cb643-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb643-11"><a href="answers-to-practical-exercises.html#cb643-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder the data so the tree and data are in the same order</span></span>
<span id="cb643-12"><a href="answers-to-practical-exercises.html#cb643-12" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> mydata[<span class="fu">match</span>(mytree<span class="sc">$</span>tip.label, mydata<span class="sc">$</span>Binomial), ]</span>
<span id="cb643-13"><a href="answers-to-practical-exercises.html#cb643-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb643-14"><a href="answers-to-practical-exercises.html#cb643-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Create Social containing just social status values</span></span>
<span id="cb643-15"><a href="answers-to-practical-exercises.html#cb643-15" aria-hidden="true" tabindex="-1"></a>Social <span class="ot">&lt;-</span> <span class="fu">pull</span>(mydata, SocialStatus)</span>
<span id="cb643-16"><a href="answers-to-practical-exercises.html#cb643-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Give social status names = species names at the tips of the phylogeny</span></span>
<span id="cb643-17"><a href="answers-to-practical-exercises.html#cb643-17" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(Social) <span class="ot">&lt;-</span> mydata<span class="sc">$</span>Binomial</span>
<span id="cb643-18"><a href="answers-to-practical-exercises.html#cb643-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the first few rows</span></span>
<span id="cb643-19"><a href="answers-to-practical-exercises.html#cb643-19" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(Social)</span></code></pre></div>
<pre><code>##  Cercopithecus_ascanius    Cercopithecus_cephus     Cercopithecus_mitis 
##                       2                       2                       2 
## Cercopithecus_neglectus Cercopithecus_nictitans      Erythrocebus_patas 
##                       2                       2                       2</code></pre>
<div class="sourceCode" id="cb645"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb645-1"><a href="answers-to-practical-exercises.html#cb645-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the ER model</span></span>
<span id="cb645-2"><a href="answers-to-practical-exercises.html#cb645-2" aria-hidden="true" tabindex="-1"></a>equal <span class="ot">&lt;-</span> <span class="fu">fitDiscrete</span>(mytree, Social, <span class="at">model =</span> <span class="st">&quot;ER&quot;</span>)</span>
<span id="cb645-3"><a href="answers-to-practical-exercises.html#cb645-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the ARD model</span></span>
<span id="cb645-4"><a href="answers-to-practical-exercises.html#cb645-4" aria-hidden="true" tabindex="-1"></a>ard <span class="ot">&lt;-</span> <span class="fu">fitDiscrete</span>(mytree, Social, <span class="at">model =</span> <span class="st">&quot;ARD&quot;</span>)</span></code></pre></div>
<p>Then answer the following questions.</p>
<ol style="list-style-type: decimal">
<li>Which model fits best?</li>
</ol>
<div class="sourceCode" id="cb646"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb646-1"><a href="answers-to-practical-exercises.html#cb646-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Using AIC weights</span></span>
<span id="cb646-2"><a href="answers-to-practical-exercises.html#cb646-2" aria-hidden="true" tabindex="-1"></a>aic.scores <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">c</span>(equal<span class="sc">$</span>opt<span class="sc">$</span>aic, ard<span class="sc">$</span>opt<span class="sc">$</span>aic), </span>
<span id="cb646-3"><a href="answers-to-practical-exercises.html#cb646-3" aria-hidden="true" tabindex="-1"></a>                       <span class="fu">c</span>(<span class="st">&quot;equal&quot;</span>, <span class="st">&quot;different&quot;</span>))</span>
<span id="cb646-4"><a href="answers-to-practical-exercises.html#cb646-4" aria-hidden="true" tabindex="-1"></a><span class="fu">aicw</span>(aic.scores)</span></code></pre></div>
<pre><code>##                fit    delta          w
## equal     38.98307 5.313651 0.06556956
## different 33.66942 0.000000 0.93443044</code></pre>
<div class="sourceCode" id="cb648"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb648-1"><a href="answers-to-practical-exercises.html#cb648-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or just the raw AIC values</span></span>
<span id="cb648-2"><a href="answers-to-practical-exercises.html#cb648-2" aria-hidden="true" tabindex="-1"></a>equal<span class="sc">$</span>opt<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 38.98307</code></pre>
<div class="sourceCode" id="cb650"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb650-1"><a href="answers-to-practical-exercises.html#cb650-1" aria-hidden="true" tabindex="-1"></a>ard<span class="sc">$</span>opt<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 33.66942</code></pre>
<p><em>All rates different</em></p>
<ol start="2" style="list-style-type: decimal">
<li>What is the transition rate from non-social (1) to social (2), and vice versa, in the all rates different model? What does this mean biologically?</li>
</ol>
<div class="sourceCode" id="cb652"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb652-1"><a href="answers-to-practical-exercises.html#cb652-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Could look at the model output</span></span>
<span id="cb652-2"><a href="answers-to-practical-exercises.html#cb652-2" aria-hidden="true" tabindex="-1"></a>ard</span></code></pre></div>
<pre><code>## GEIGER-fitted comparative model of discrete data
##  fitted Q matrix:
##                   1             2
##     1 -9.264914e-03  9.264914e-03
##     2  1.476682e-17 -1.476682e-17
## 
##  model summary:
##  log-likelihood = -14.834711
##  AIC = 33.669423
##  AICc = 33.831585
##  free parameters = 2
## 
## Convergence diagnostics:
##  optimization iterations = 100
##  failed iterations = 0
##  number of iterations with same best fit = 46
##  frequency of best fit = 0.46
## 
##  object summary:
##  &#39;lik&#39; -- likelihood function
##  &#39;bnd&#39; -- bounds for likelihood search
##  &#39;res&#39; -- optimization iteration summary
##  &#39;opt&#39; -- maximum likelihood parameter estimates</code></pre>
<div class="sourceCode" id="cb654"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb654-1"><a href="answers-to-practical-exercises.html#cb654-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or plot the Q matrix as network</span></span>
<span id="cb654-2"><a href="answers-to-practical-exercises.html#cb654-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.gfit</span>(ard)</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-381-1.png" width="672" /></p>
<p><em>1 to 2 is ~0.009, 2 to 1 is very close to zero. Transitions from social (2) to non-social (1) behaviour are extremely rare.</em></p>
<ol start="3" style="list-style-type: decimal">
<li>Plot ancestral state estimates of social status (from the ARD model) on the phylogeny. Is the root node estimated as being non-social or social?</li>
</ol>
<div class="sourceCode" id="cb655"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb655-1"><a href="answers-to-practical-exercises.html#cb655-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ancestral state estimation (ignore the warnings)</span></span>
<span id="cb655-2"><a href="answers-to-practical-exercises.html#cb655-2" aria-hidden="true" tabindex="-1"></a>ancestral_social <span class="ot">&lt;-</span> <span class="fu">ace</span>(Social, mytree, <span class="at">type =</span> <span class="st">&quot;discrete&quot;</span>, <span class="at">model =</span> <span class="st">&quot;ARD&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in sqrt(diag(solve(h))): NaNs produced</code></pre>
<div class="sourceCode" id="cb657"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb657-1"><a href="answers-to-practical-exercises.html#cb657-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Choose colours</span></span>
<span id="cb657-2"><a href="answers-to-practical-exercises.html#cb657-2" aria-hidden="true" tabindex="-1"></a>social_colours <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;deeppink&quot;</span>, <span class="st">&quot;goldenrod&quot;</span>)</span>
<span id="cb657-3"><a href="answers-to-practical-exercises.html#cb657-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb657-4"><a href="answers-to-practical-exercises.html#cb657-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the tree with colours at the tips and nodes to represent socail status</span></span>
<span id="cb657-5"><a href="answers-to-practical-exercises.html#cb657-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(mytree, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">adj =</span> <span class="fu">c</span>(<span class="fl">0.2</span>), <span class="at">type =</span> <span class="st">&quot;fan&quot;</span>, </span>
<span id="cb657-6"><a href="answers-to-practical-exercises.html#cb657-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">no.margin =</span> <span class="cn">TRUE</span>, <span class="at">show.tip.label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb657-7"><a href="answers-to-practical-exercises.html#cb657-7" aria-hidden="true" tabindex="-1"></a><span class="fu">tiplabels</span>(<span class="at">pch =</span> <span class="dv">16</span>, <span class="at">col =</span> social_colours[mydata<span class="sc">$</span>SocialStatus])</span>
<span id="cb657-8"><a href="answers-to-practical-exercises.html#cb657-8" aria-hidden="true" tabindex="-1"></a><span class="fu">nodelabels</span>(<span class="at">pie =</span> ancestral_social<span class="sc">$</span>lik.anc, <span class="at">piecol =</span> social_colours, <span class="at">cex =</span> <span class="fl">0.5</span>)</span>
<span id="cb657-9"><a href="answers-to-practical-exercises.html#cb657-9" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomleft&quot;</span>, <span class="at">pch =</span> <span class="dv">15</span>, <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb657-10"><a href="answers-to-practical-exercises.html#cb657-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">&quot;Non Social&quot;</span>, <span class="st">&quot;Social&quot;</span>), </span>
<span id="cb657-11"><a href="answers-to-practical-exercises.html#cb657-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> social_colours)</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-382-1.png" width="672" /></p>
<p><em>Root is non-social</em></p>
<ol start="4" style="list-style-type: decimal">
<li>Why didn’t I ask you to fit the symmetric model?</li>
</ol>
<p><em>Because with only two states the symmetric model is equivalent to the equal rates model</em></p>
</div>
<div id="beyond-simple-macroevolutionary-models-of-trait-evolution" class="section level2 hasAnchor" number="12.10">
<h2><span class="header-section-number">12.10</span> Beyond simple macroevolutionary models of trait evolution<a href="answers-to-practical-exercises.html#beyond-simple-macroevolutionary-models-of-trait-evolution" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Read in the tree and data, then prepare them for a PCM analysis (you may have already done this in a previous exercise which should save you some time).</p>
<p>Using log body size (<code>AdultBodyMass_g</code>) as your continuous trait, and social status (<code>SocialStatus</code>) to define the selective regimes, fit (A) a multi-rate Brownian motion model and (B) an OU model with trait optima or long-term mean values for different evolutionary regimes but the same <span class="math inline">\(\sigma^2\)</span> and <span class="math inline">\(\alpha\)</span>. Note that social status has two states: non-social (1) or social (2). Assume it is evolving according to an all rates different model.</p>
<div class="sourceCode" id="cb658"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb658-1"><a href="answers-to-practical-exercises.html#cb658-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the packages</span></span>
<span id="cb658-2"><a href="answers-to-practical-exercises.html#cb658-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb658-3"><a href="answers-to-practical-exercises.html#cb658-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb658-4"><a href="answers-to-practical-exercises.html#cb658-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caper)</span>
<span id="cb658-5"><a href="answers-to-practical-exercises.html#cb658-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb658-6"><a href="answers-to-practical-exercises.html#cb658-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(OUwie)</span>
<span id="cb658-7"><a href="answers-to-practical-exercises.html#cb658-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb658-8"><a href="answers-to-practical-exercises.html#cb658-8" aria-hidden="true" tabindex="-1"></a><span class="do">### Prepare the tree and the data as shown in the Preparation answer above </span><span class="al">###</span></span>
<span id="cb658-9"><a href="answers-to-practical-exercises.html#cb658-9" aria-hidden="true" tabindex="-1"></a><span class="do">### This should mean you now have mydata and mytree ready to work with </span><span class="al">###</span></span>
<span id="cb658-10"><a href="answers-to-practical-exercises.html#cb658-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb658-11"><a href="answers-to-practical-exercises.html#cb658-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder the data so the tree and data are in the same order</span></span>
<span id="cb658-12"><a href="answers-to-practical-exercises.html#cb658-12" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> mydata[<span class="fu">match</span>(mytree<span class="sc">$</span>tip.label, mydata<span class="sc">$</span>Binomial), ]</span>
<span id="cb658-13"><a href="answers-to-practical-exercises.html#cb658-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb658-14"><a href="answers-to-practical-exercises.html#cb658-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Create Social containing just social status values</span></span>
<span id="cb658-15"><a href="answers-to-practical-exercises.html#cb658-15" aria-hidden="true" tabindex="-1"></a>Social <span class="ot">&lt;-</span> <span class="fu">pull</span>(mydata, SocialStatus)</span>
<span id="cb658-16"><a href="answers-to-practical-exercises.html#cb658-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Give social status names = species names at the tips of the phylogeny</span></span>
<span id="cb658-17"><a href="answers-to-practical-exercises.html#cb658-17" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(Social) <span class="ot">&lt;-</span> mydata<span class="sc">$</span>Binomial</span>
<span id="cb658-18"><a href="answers-to-practical-exercises.html#cb658-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb658-19"><a href="answers-to-practical-exercises.html#cb658-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate node and tip values for 500 trees using the ARD model</span></span>
<span id="cb658-20"><a href="answers-to-practical-exercises.html#cb658-20" aria-hidden="true" tabindex="-1"></a>trees.ARD <span class="ot">&lt;-</span> <span class="fu">make.simmap</span>(mytree, Social, <span class="at">model =</span> <span class="st">&quot;ARD&quot;</span>, <span class="at">nsim =</span> <span class="dv">500</span>)</span></code></pre></div>
<pre><code>## make.simmap is sampling character histories conditioned on
## the transition matrix
## 
## Q =
##              1           2
## 1 -0.009264911 0.009264911
## 2  0.000000000 0.000000000
## (estimated using likelihood);
## and (mean) root node prior probabilities
## pi =
##   1   2 
## 0.5 0.5</code></pre>
<div class="sourceCode" id="cb660"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb660-1"><a href="answers-to-practical-exercises.html#cb660-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Store the results of the summary of the 500 trees</span></span>
<span id="cb660-2"><a href="answers-to-practical-exercises.html#cb660-2" aria-hidden="true" tabindex="-1"></a>summary.ARD <span class="ot">&lt;-</span> <span class="fu">summary</span>(trees.ARD)</span>
<span id="cb660-3"><a href="answers-to-practical-exercises.html#cb660-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb660-4"><a href="answers-to-practical-exercises.html#cb660-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the max likelihood state at each node</span></span>
<span id="cb660-5"><a href="answers-to-practical-exercises.html#cb660-5" aria-hidden="true" tabindex="-1"></a>best <span class="ot">&lt;-</span> <span class="fu">apply</span>(summary.ARD<span class="sc">$</span>ace[<span class="dv">1</span><span class="sc">:</span><span class="dv">76</span>, ], <span class="dv">1</span>, which.max)</span>
<span id="cb660-6"><a href="answers-to-practical-exercises.html#cb660-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb660-7"><a href="answers-to-practical-exercises.html#cb660-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign regimes to the tree node labels</span></span>
<span id="cb660-8"><a href="answers-to-practical-exercises.html#cb660-8" aria-hidden="true" tabindex="-1"></a>mytree<span class="sc">$</span>node.label <span class="ot">&lt;-</span> <span class="fu">levels</span>(<span class="fu">as.factor</span>(mydata<span class="sc">$</span>SocialStatus))[best]</span>
<span id="cb660-9"><a href="answers-to-practical-exercises.html#cb660-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb660-10"><a href="answers-to-practical-exercises.html#cb660-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Make new dataset with just three columns for OUwie</span></span>
<span id="cb660-11"><a href="answers-to-practical-exercises.html#cb660-11" aria-hidden="true" tabindex="-1"></a>bodymass <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">species =</span> mydata<span class="sc">$</span>Binomial, </span>
<span id="cb660-12"><a href="answers-to-practical-exercises.html#cb660-12" aria-hidden="true" tabindex="-1"></a>                    <span class="at">regime =</span> mydata<span class="sc">$</span>SocialStatus, </span>
<span id="cb660-13"><a href="answers-to-practical-exercises.html#cb660-13" aria-hidden="true" tabindex="-1"></a>                    <span class="at">trait =</span> <span class="fu">log</span>(mydata<span class="sc">$</span>AdultBodyMass_g))</span>
<span id="cb660-14"><a href="answers-to-practical-exercises.html#cb660-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb660-15"><a href="answers-to-practical-exercises.html#cb660-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit BMS model</span></span>
<span id="cb660-16"><a href="answers-to-practical-exercises.html#cb660-16" aria-hidden="true" tabindex="-1"></a>BMS <span class="ot">&lt;-</span> <span class="fu">OUwie</span>(mytree, bodymass, <span class="at">model =</span> <span class="st">&quot;BMS&quot;</span>, <span class="at">root.station =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Warning: An algorithm was not specified. Defaulting to computing the determinant and
## inversion of the vcv.</code></pre>
<pre><code>## Initializing... 
## Finished. Begin thorough search... 
## Finished. Summarizing results.</code></pre>
<div class="sourceCode" id="cb663"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb663-1"><a href="answers-to-practical-exercises.html#cb663-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit OUM model</span></span>
<span id="cb663-2"><a href="answers-to-practical-exercises.html#cb663-2" aria-hidden="true" tabindex="-1"></a>OUM <span class="ot">&lt;-</span> <span class="fu">OUwie</span>(mytree, bodymass, <span class="at">model =</span> <span class="st">&quot;OUM&quot;</span>)</span></code></pre></div>
<pre><code>## Warning: An algorithm was not specified. Defaulting to computing the determinant and
## inversion of the vcv.</code></pre>
<pre><code>## Initializing... 
## Finished. Begin thorough search... 
## Finished. Summarizing results.</code></pre>
<p>Then answer the following questions:</p>
<ol style="list-style-type: decimal">
<li>Which is the best fitting model?</li>
</ol>
<div class="sourceCode" id="cb666"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb666-1"><a href="answers-to-practical-exercises.html#cb666-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract AICc scores from the models</span></span>
<span id="cb666-2"><a href="answers-to-practical-exercises.html#cb666-2" aria-hidden="true" tabindex="-1"></a>aic.scores <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">c</span>(BMS<span class="sc">$</span>AICc, OUM<span class="sc">$</span>AICc), </span>
<span id="cb666-3"><a href="answers-to-practical-exercises.html#cb666-3" aria-hidden="true" tabindex="-1"></a>                           <span class="fu">c</span>(<span class="st">&quot;BMS&quot;</span>, <span class="st">&quot;OUM&quot;</span>))</span>
<span id="cb666-4"><a href="answers-to-practical-exercises.html#cb666-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Get aic weights</span></span>
<span id="cb666-5"><a href="answers-to-practical-exercises.html#cb666-5" aria-hidden="true" tabindex="-1"></a><span class="fu">aicw</span>(aic.scores)</span></code></pre></div>
<p><em>OUM</em></p>
<ol start="2" style="list-style-type: decimal">
<li>If you were running these analyses for a publication, what additional models would you compare these models to?</li>
</ol>
<p><em>BM and single peak OU for sure. Possibly one or more of the more complex models if you had a good biological reason to</em></p>
<ol start="3" style="list-style-type: decimal">
<li>In the multi rate Brownian model, which social status state has the higher rate of log body mass evolution?</li>
</ol>
<div class="sourceCode" id="cb667"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb667-1"><a href="answers-to-practical-exercises.html#cb667-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the model output</span></span>
<span id="cb667-2"><a href="answers-to-practical-exercises.html#cb667-2" aria-hidden="true" tabindex="-1"></a>BMS</span></code></pre></div>
<p><em>Non-social (1) has the higher rate of body mass evolution 0.04031313</em></p>
<ol start="4" style="list-style-type: decimal">
<li>Look at the parameter estimates of the multi peak OU model. Do you see anything worrying there? It might help to look at a histogram showing the body sizes of species in the non-social and the social categories.</li>
</ol>
<div class="sourceCode" id="cb668"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb668-1"><a href="answers-to-practical-exercises.html#cb668-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the output</span></span>
<span id="cb668-2"><a href="answers-to-practical-exercises.html#cb668-2" aria-hidden="true" tabindex="-1"></a>OUM</span>
<span id="cb668-3"><a href="answers-to-practical-exercises.html#cb668-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb668-4"><a href="answers-to-practical-exercises.html#cb668-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the body sizes for the different groups</span></span>
<span id="cb668-5"><a href="answers-to-practical-exercises.html#cb668-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mydata, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log</span>(AdultBodyMass_g), <span class="at">fill =</span> <span class="fu">as.factor</span>(SocialStatus))) <span class="sc">+</span></span>
<span id="cb668-6"><a href="answers-to-practical-exercises.html#cb668-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">bins =</span> <span class="dv">20</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb668-7"><a href="answers-to-practical-exercises.html#cb668-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> social_colours) <span class="sc">+</span></span>
<span id="cb668-8"><a href="answers-to-practical-exercises.html#cb668-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>SocialStatus, <span class="at">ncol =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb668-9"><a href="answers-to-practical-exercises.html#cb668-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span></span>
<span id="cb668-10"><a href="answers-to-practical-exercises.html#cb668-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><em>The optima estimates are way off. Non-social species have a long-term mean log body size estimate of 5.918899, which makes sense looking at the histograms of the data. But social species have a long-term mean log body size estimate of 18607801, which if you exponentiate it is infinity. Even King Kong was not infinitely large! Something has clearly gone wrong here, but if we had not looked at the model parameters we would have said this was the best model…It’s likely we just don’t have enough data to fit a model that is as complex as this, so the parameter optimisations are not reliable.</em></p>
</div>
<div id="diversification" class="section level2 hasAnchor" number="12.11">
<h2><span class="header-section-number">12.11</span> Diversification<a href="answers-to-practical-exercises.html#diversification" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><code>BAMMtools</code> has a several inbuilt example datasets. For this practical exercise we will use the cetacean (<code>whales</code>; from <span class="citation">(<a href="#ref-steeman2009radiation" role="doc-biblioref"><strong>steeman2009radiation?</strong></a>)</span>) and the ray-finned fishes (<code>fishes</code>; from <span class="citation">(<a href="#ref-rabosky2013rates" role="doc-biblioref"><strong>rabosky2013rates?</strong></a>)</span>) datasets.</p>
<p>Load the whales and fishes trees and BAMM outputs data into R (<strong>there is no <code>mcmc.fishes</code> file</strong>_).</p>
<div class="sourceCode" id="cb669"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb669-1"><a href="answers-to-practical-exercises.html#cb669-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the packages</span></span>
<span id="cb669-2"><a href="answers-to-practical-exercises.html#cb669-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb669-3"><a href="answers-to-practical-exercises.html#cb669-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BAMMtools)</span>
<span id="cb669-4"><a href="answers-to-practical-exercises.html#cb669-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(coda)</span>
<span id="cb669-5"><a href="answers-to-practical-exercises.html#cb669-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phytools)</span>
<span id="cb669-6"><a href="answers-to-practical-exercises.html#cb669-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span>
<span id="cb669-7"><a href="answers-to-practical-exercises.html#cb669-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb669-8"><a href="answers-to-practical-exercises.html#cb669-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Load whales data from BAMMtools</span></span>
<span id="cb669-9"><a href="answers-to-practical-exercises.html#cb669-9" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(whales)</span>
<span id="cb669-10"><a href="answers-to-practical-exercises.html#cb669-10" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(mcmc.whales)</span>
<span id="cb669-11"><a href="answers-to-practical-exercises.html#cb669-11" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(events.whales)</span>
<span id="cb669-12"><a href="answers-to-practical-exercises.html#cb669-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb669-13"><a href="answers-to-practical-exercises.html#cb669-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Load fishes data from BAMMtools</span></span>
<span id="cb669-14"><a href="answers-to-practical-exercises.html#cb669-14" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(fishes)</span>
<span id="cb669-15"><a href="answers-to-practical-exercises.html#cb669-15" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(mcmc.fishes)</span></code></pre></div>
<pre><code>## Warning in data(mcmc.fishes): data set &#39;mcmc.fishes&#39; not found</code></pre>
<div class="sourceCode" id="cb671"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb671-1"><a href="answers-to-practical-exercises.html#cb671-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(events.fishes)</span>
<span id="cb671-2"><a href="answers-to-practical-exercises.html#cb671-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb671-3"><a href="answers-to-practical-exercises.html#cb671-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the tree, mcmc, and edata for whales</span></span>
<span id="cb671-4"><a href="answers-to-practical-exercises.html#cb671-4" aria-hidden="true" tabindex="-1"></a>whales_tree <span class="ot">&lt;-</span> whales</span>
<span id="cb671-5"><a href="answers-to-practical-exercises.html#cb671-5" aria-hidden="true" tabindex="-1"></a>whales_mcmc <span class="ot">&lt;-</span> mcmc.whales</span>
<span id="cb671-6"><a href="answers-to-practical-exercises.html#cb671-6" aria-hidden="true" tabindex="-1"></a>whales_edata <span class="ot">&lt;-</span> <span class="fu">getEventData</span>(whales, <span class="at">eventdata =</span> events.whales, <span class="at">burnin =</span> <span class="fl">0.1</span>)</span></code></pre></div>
<pre><code>## Processing event data from data.frame
## 
## Discarded as burnin: GENERATIONS &lt;  995000
## Analyzing  1801  samples from posterior
## 
## Setting recursive sequence on tree...
## 
## Done with recursive sequence</code></pre>
<div class="sourceCode" id="cb673"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb673-1"><a href="answers-to-practical-exercises.html#cb673-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the tree, and edata for fishes</span></span>
<span id="cb673-2"><a href="answers-to-practical-exercises.html#cb673-2" aria-hidden="true" tabindex="-1"></a>fishes_tree <span class="ot">&lt;-</span> fishes</span>
<span id="cb673-3"><a href="answers-to-practical-exercises.html#cb673-3" aria-hidden="true" tabindex="-1"></a>fishes_edata <span class="ot">&lt;-</span> <span class="fu">getEventData</span>(fishes, <span class="at">eventdata =</span> events.fishes, <span class="at">burnin =</span> <span class="fl">0.1</span>)</span></code></pre></div>
<pre><code>## Processing event data from data.frame
## 
## Discarded as burnin: GENERATIONS &lt;  5
## Analyzing  46  samples from posterior
## 
## Setting recursive sequence on tree...
## 
## Done with recursive sequence</code></pre>
<p>Then answer the following questions:</p>
<ol style="list-style-type: decimal">
<li>How can we tell that the BAMM model for whales has converged? What is the effective sample size for number of shifts?</li>
</ol>
<div class="sourceCode" id="cb675"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb675-1"><a href="answers-to-practical-exercises.html#cb675-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the plot</span></span>
<span id="cb675-2"><a href="answers-to-practical-exercises.html#cb675-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(whales_mcmc<span class="sc">$</span>logLik <span class="sc">~</span> whales_mcmc<span class="sc">$</span>generation)</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-388-1.png" width="672" /></p>
<div class="sourceCode" id="cb676"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb676-1"><a href="answers-to-practical-exercises.html#cb676-1" aria-hidden="true" tabindex="-1"></a><span class="co"># What is 10% of rows?</span></span>
<span id="cb676-2"><a href="answers-to-practical-exercises.html#cb676-2" aria-hidden="true" tabindex="-1"></a>burnstart <span class="ot">&lt;-</span> <span class="fu">floor</span>(<span class="fl">0.1</span> <span class="sc">*</span> <span class="fu">nrow</span>(whales_mcmc))</span>
<span id="cb676-3"><a href="answers-to-practical-exercises.html#cb676-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove the first 10% from the runs </span></span>
<span id="cb676-4"><a href="answers-to-practical-exercises.html#cb676-4" aria-hidden="true" tabindex="-1"></a>postburn <span class="ot">&lt;-</span> whales_mcmc[burnstart<span class="sc">:</span><span class="fu">nrow</span>(whales_mcmc), ]</span>
<span id="cb676-5"><a href="answers-to-practical-exercises.html#cb676-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb676-6"><a href="answers-to-practical-exercises.html#cb676-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract ESS for the number of rate shifts parameter</span></span>
<span id="cb676-7"><a href="answers-to-practical-exercises.html#cb676-7" aria-hidden="true" tabindex="-1"></a><span class="fu">effectiveSize</span>(postburn<span class="sc">$</span>N_shifts)</span></code></pre></div>
<pre><code>##     var1 
## 1587.998</code></pre>
<div class="sourceCode" id="cb678"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb678-1"><a href="answers-to-practical-exercises.html#cb678-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract ESS for the log likelihood</span></span>
<span id="cb678-2"><a href="answers-to-practical-exercises.html#cb678-2" aria-hidden="true" tabindex="-1"></a><span class="fu">effectiveSize</span>(postburn<span class="sc">$</span>logLik)</span></code></pre></div>
<pre><code>##     var1 
## 1091.479</code></pre>
<p><em>The plot looks like a furry caterpillar (!), and effective sample sizes are well over 200.</em></p>
<ol start="2" style="list-style-type: decimal">
<li>What is the most common number of rate shifts in the posterior for the whales data?</li>
</ol>
<div class="sourceCode" id="cb680"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb680-1"><a href="answers-to-practical-exercises.html#cb680-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the posterior for number of shifts</span></span>
<span id="cb680-2"><a href="answers-to-practical-exercises.html#cb680-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(whales_edata)</span></code></pre></div>
<pre><code>## 
## Analyzed 1801 posterior samples
## Shift posterior distribution:
## 
##          0     0.0770
##          1     0.6600
##          2     0.2100
##          3     0.0430
##          4     0.0110
##          5     0.0022
## 
## Compute credible set of shift configurations for more information:
##  See ?credibleShiftSet and ?getBestShiftConfiguration</code></pre>
<p><em>1</em></p>
<ol start="3" style="list-style-type: decimal">
<li>Plot the 95% credible set of rate shift configurations for whales.</li>
</ol>
<div class="sourceCode" id="cb682"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb682-1"><a href="answers-to-practical-exercises.html#cb682-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract 95% credible set of rate shift configurations</span></span>
<span id="cb682-2"><a href="answers-to-practical-exercises.html#cb682-2" aria-hidden="true" tabindex="-1"></a>whales_css <span class="ot">&lt;-</span> <span class="fu">credibleShiftSet</span>(whales_edata, <span class="at">expectedNumberOfShifts =</span> <span class="dv">1</span>, </span>
<span id="cb682-3"><a href="answers-to-practical-exercises.html#cb682-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">threshold =</span> <span class="dv">5</span>, <span class="at">set.limit =</span> <span class="fl">0.95</span>)</span>
<span id="cb682-4"><a href="answers-to-practical-exercises.html#cb682-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the summary</span></span>
<span id="cb682-5"><a href="answers-to-practical-exercises.html#cb682-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(whales_css)</span></code></pre></div>
<pre><code>## 
##  95 % credible set of rate shift configurations sampled with BAMM
## 
## Distinct shift configurations in credible set:  4
## 
## Frequency of 4 shift configurations with highest posterior probability:
## 
## 
##    rank     probability cumulative  Core_shifts
##          1 0.46418656  0.4641866          1
##          2 0.28928373  0.7534703          1
##          3 0.14158801  0.8950583          1
##          4 0.09272626  0.9877846          0</code></pre>
<div class="sourceCode" id="cb684"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb684-1"><a href="answers-to-practical-exercises.html#cb684-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot 95 % credible set of rate shift configurations</span></span>
<span id="cb684-2"><a href="answers-to-practical-exercises.html#cb684-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.credibleshiftset</span>(whales_css)</span></code></pre></div>
<pre><code>## Omitted 0 plots</code></pre>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-390-1.png" width="672" /></p>
<ol start="4" style="list-style-type: decimal">
<li>Plot the mean phylorate plot for fishes.</li>
</ol>
<div class="sourceCode" id="cb686"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb686-1"><a href="answers-to-practical-exercises.html#cb686-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot mean phylorate plot with legend and slighting larger line width (lwd)</span></span>
<span id="cb686-2"><a href="answers-to-practical-exercises.html#cb686-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.bammdata</span>(fishes_edata, <span class="at">legend =</span> <span class="cn">TRUE</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-391-1.png" width="672" /></p>
<ol start="5" style="list-style-type: decimal">
<li>What are the mean speciation rate and the 90% posterior density for mean speciation rate in fishes?</li>
</ol>
<div class="sourceCode" id="cb687"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb687-1"><a href="answers-to-practical-exercises.html#cb687-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract speciation and extinction rates from the whole posterior</span></span>
<span id="cb687-2"><a href="answers-to-practical-exercises.html#cb687-2" aria-hidden="true" tabindex="-1"></a>allrates <span class="ot">&lt;-</span> <span class="fu">getCladeRates</span>(fishes_edata)</span>
<span id="cb687-3"><a href="answers-to-practical-exercises.html#cb687-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate overall mean speciation rate</span></span>
<span id="cb687-4"><a href="answers-to-practical-exercises.html#cb687-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(allrates<span class="sc">$</span>lambda)</span></code></pre></div>
<pre><code>## [1] 0.09470252</code></pre>
<div class="sourceCode" id="cb689"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb689-1"><a href="answers-to-practical-exercises.html#cb689-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate 90% highest posterior density for speciation rate</span></span>
<span id="cb689-2"><a href="answers-to-practical-exercises.html#cb689-2" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(allrates<span class="sc">$</span>lambda, <span class="fu">c</span>(<span class="fl">0.05</span>, <span class="fl">0.95</span>))</span></code></pre></div>
<pre><code>##         5%        95% 
## 0.09020489 0.09960825</code></pre>
<p><em>Mean = 0.09470252, 5% = 0.09020489, 95% = 0.09960825</em></p>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-arnold201010ktrees" class="csl-entry">
Arnold, Christian, Luke J Matthews, and Charles L Nunn. 2010. <span>“The 10kTrees Website: A New Online Resource for Primate Phylogeny.”</span> <em>Evolutionary Anthropology: Issues, News, and Reviews</em> 19 (3): 114–18.
</div>
<div id="ref-feng2017phylogenomics" class="csl-entry">
Feng, Yan-Jie, David C Blackburn, Dan Liang, David M Hillis, David B Wake, David C Cannatella, and Peng Zhang. 2017. <span>“Phylogenomics Reveals Rapid, Simultaneous Diversification of Three Major Clades of Gondwanan Frogs at the Cretaceous–Paleogene Boundary.”</span> <em>Proceedings of the National Academy of Sciences</em> 114 (29): E5864–70.
</div>
<div id="ref-jones2009pantheria" class="csl-entry">
Jones, Kate E, Jon Bielby, Marcel Cardillo, Susanne A Fritz, Justin O’Dell, C David L Orme, Kamran Safi, et al. 2009. <span>“PanTHERIA: A Species-Level Database of Life History, Ecology, and Geography of Extant and Recently Extinct Mammals: Ecological Archives E090-184.”</span> <em>Ecology</em> 90 (9): 2648–48.
</div>
<div id="ref-thomas2020eye" class="csl-entry">
Thomas, Kate N, David J Gower, Rayna C Bell, Matthew K Fujita, Ryan K Schott, and Jeffrey W Streicher. 2020. <span>“Eye Size and Investment in Frogs and Toads Correlate with Adult Habitat, Activity Pattern and Breeding Ecology.”</span> <em>Proceedings of the Royal Society B</em> 287 (1935): 20201393.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="diversification-rate-analyses.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="common-mistakes-using-pcms-in-r.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/12-Answers.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["pcm-primer.pdf", "pcm-primer.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
