<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 9 Macroevolutionary models for discrete traits | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences</title>
  <meta name="description" content="Chapter 9 Macroevolutionary models for discrete traits | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 9 Macroevolutionary models for discrete traits | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 9 Macroevolutionary models for discrete traits | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  
  
  

<meta name="author" content="Natalie Cooper" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="macroevolutionary-models-for-continuous-traits.html"/>
<link rel="next" href="beyond-simple-macroevolutionary-models.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">PCM Primer</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#datasets-and-scripts"><i class="fa fa-check"></i>Datasets and scripts</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#how-to-use-these-materials"><i class="fa fa-check"></i>How to use these materials</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#boxes"><i class="fa fa-check"></i>Boxes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#example-datasets"><i class="fa fa-check"></i>Example datasets</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#frog-eye-size-evolution"><i class="fa fa-check"></i>Frog eye size evolution</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#natricine-snake-head-shape-evolution"><i class="fa fa-check"></i>Natricine snake head shape evolution</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#diversification-in-dragonflies"><i class="fa fa-check"></i>Diversification in dragonflies</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#citing-r-and-r-packages"><i class="fa fa-check"></i>Citing R and R packages</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#acknowledgements"><i class="fa fa-check"></i>Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><i class="fa fa-check"></i><b>1</b> What you need to be able to do in R before you start</a>
<ul>
<li class="chapter" data-level="1.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-r-and-rstudio"><i class="fa fa-check"></i><b>1.1</b> Installing R (and RStudio)</a></li>
<li class="chapter" data-level="1.2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#downloading-the-data"><i class="fa fa-check"></i><b>1.2</b> Downloading the data</a></li>
<li class="chapter" data-level="1.3" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#r-projects"><i class="fa fa-check"></i><b>1.3</b> R Projects</a></li>
<li class="chapter" data-level="1.4" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#using-a-script"><i class="fa fa-check"></i><b>1.4</b> Using a script</a></li>
<li class="chapter" data-level="1.5" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-and-loading-extra-packages-in-r"><i class="fa fa-check"></i><b>1.5</b> Installing and loading extra packages in R</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#loading-packages-into-r"><i class="fa fa-check"></i><b>1.5.1</b> Loading packages into R</a></li>
<li class="chapter" data-level="1.5.2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#ggtree-a-special-case-for-package-installation"><i class="fa fa-check"></i><b>1.5.2</b> <code>ggtree</code>, a special case for package installation</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#loading-and-viewing-your-data-in-r"><i class="fa fa-check"></i><b>1.6</b> Loading and viewing your data in R</a></li>
<li class="chapter" data-level="1.7" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#errors-and-warnings"><i class="fa fa-check"></i><b>1.7</b> Errors and warnings</a>
<ul>
<li class="chapter" data-level="1.7.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#removing-the"><i class="fa fa-check"></i><b>1.7.1</b> Removing the <code>+</code></a></li>
</ul></li>
<li class="chapter" data-level="1.8" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#summary"><i class="fa fa-check"></i><b>1.8</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html"><i class="fa fa-check"></i><b>2</b> Data wrangling comparative data in R</a>
<ul>
<li class="chapter" data-level="" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#before-you-start"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="2.1" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#chameleon-data"><i class="fa fa-check"></i><b>2.1</b> Chameleon data</a></li>
<li class="chapter" data-level="2.2" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#summary-1"><i class="fa fa-check"></i><b>2.2</b> Summary</a></li>
<li class="chapter" data-level="2.3" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#practical-exercise"><i class="fa fa-check"></i><b>2.3</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html"><i class="fa fa-check"></i><b>3</b> Understanding, plotting and manipulating phylogenies in R</a>
<ul>
<li class="chapter" data-level="" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#before-you-start-1"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="3.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#what-do-phylogeny-files-look-like"><i class="fa fa-check"></i><b>3.1</b> What do phylogeny files look like?</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#newick-trees"><i class="fa fa-check"></i><b>3.1.1</b> Newick trees</a></li>
<li class="chapter" data-level="3.1.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#nexus-trees"><i class="fa fa-check"></i><b>3.1.2</b> NEXUS trees</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#how-are-phylogenies-formatted-in-r"><i class="fa fa-check"></i><b>3.2</b> How are phylogenies formatted in R?</a></li>
<li class="chapter" data-level="3.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#plotting-phylogenies"><i class="fa fa-check"></i><b>3.3</b> Plotting phylogenies</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#basic-phylogeny-plotting-with-ape"><i class="fa fa-check"></i><b>3.3.1</b> Basic phylogeny plotting with APE</a></li>
<li class="chapter" data-level="3.3.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#basic-phylogeny-plotting-with-ggtree"><i class="fa fa-check"></i><b>3.3.2</b> Basic phylogeny plotting with <code>ggtree</code></a></li>
<li class="chapter" data-level="3.3.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#back-to-ggtree"><i class="fa fa-check"></i><b>3.3.3</b> Back to <code>ggtree</code></a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#phylogeny-manipulation"><i class="fa fa-check"></i><b>3.4</b> Phylogeny manipulation</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#removing-polytomies"><i class="fa fa-check"></i><b>3.4.1</b> Removing polytomies</a></li>
<li class="chapter" data-level="3.4.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#rooting-your-phylogeny"><i class="fa fa-check"></i><b>3.4.2</b> Rooting your phylogeny</a></li>
<li class="chapter" data-level="3.4.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#manipulating-the-species-in-your-phlyogeny"><i class="fa fa-check"></i><b>3.4.3</b> Manipulating the species in your phlyogeny</a></li>
<li class="chapter" data-level="3.4.4" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#writing-phylogenies-to-file"><i class="fa fa-check"></i><b>3.4.4</b> Writing phylogenies to file</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#summary-2"><i class="fa fa-check"></i><b>3.5</b> Summary</a></li>
<li class="chapter" data-level="3.6" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#practical-exercise-1"><i class="fa fa-check"></i><b>3.6</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html"><i class="fa fa-check"></i><b>4</b> Preparing your tree and data for PCMs in R</a>
<ul>
<li class="chapter" data-level="" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#before-you-start-2"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="4.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#preparation"><i class="fa fa-check"></i><b>4.1</b> Preparation</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#load-the-required-packages"><i class="fa fa-check"></i><b>4.1.1</b> Load the required packages</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-and-checking-your-phylogeny-in-r"><i class="fa fa-check"></i><b>4.2</b> Reading and checking your phylogeny in R</a></li>
<li class="chapter" data-level="4.3" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>4.3</b> Reading the data into R</a></li>
<li class="chapter" data-level="4.4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-your-data-to-your-phylogeny"><i class="fa fa-check"></i><b>4.4</b> Matching your data to your phylogeny</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#species-names-with-spaces"><i class="fa fa-check"></i><b>4.4.1</b> Species names with spaces</a></li>
<li class="chapter" data-level="4.4.2" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#mismatches-between-species-in-your-data-and-phylogeny"><i class="fa fa-check"></i><b>4.4.2</b> Mismatches between species in your data and phylogeny</a></li>
<li class="chapter" data-level="4.4.3" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-the-tree-and-the-data"><i class="fa fa-check"></i><b>4.4.3</b> Matching the tree and the data</a></li>
<li class="chapter" data-level="4.4.4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#nas-in-your-data"><i class="fa fa-check"></i><b>4.4.4</b> NAs in your data</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#quick-template-code"><i class="fa fa-check"></i><b>4.5</b> Quick template code</a></li>
<li class="chapter" data-level="4.6" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#summary-3"><i class="fa fa-check"></i><b>4.6</b> Summary</a></li>
<li class="chapter" data-level="4.7" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#practical-exercise-2"><i class="fa fa-check"></i><b>4.7</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html"><i class="fa fa-check"></i><b>5</b> Phylogenetic Signal in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#before-you-start-3"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="5.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#preparation-1"><i class="fa fa-check"></i><b>5.1</b> Preparation</a></li>
<li class="chapter" data-level="5.2" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#estimating-phylogenetic-signal-for-continuous-variables"><i class="fa fa-check"></i><b>5.2</b> Estimating phylogenetic signal for continuous variables</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#pagels-lambda"><i class="fa fa-check"></i><b>5.2.1</b> Pagel’s <span class="math inline">\(\lambda\)</span></a></li>
<li class="chapter" data-level="5.2.2" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#blombergs-k-blomberg-et-al-2003"><i class="fa fa-check"></i><b>5.2.2</b> Blomberg’s <em>K</em> (Blomberg et al 2003)</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#estimating-phylogenetic-signal-for-non-continuous-variables"><i class="fa fa-check"></i><b>5.3</b> Estimating phylogenetic signal for non-continuous variables</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#d"><i class="fa fa-check"></i><b>5.3.1</b> <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#summary-4"><i class="fa fa-check"></i><b>5.4</b> Summary</a></li>
<li class="chapter" data-level="5.5" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#practical-exercises"><i class="fa fa-check"></i><b>5.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html"><i class="fa fa-check"></i><b>6</b> Phylogenetic Generalised Least Squares (PGLS) in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#before-you-start-4"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="6.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#preparation-2"><i class="fa fa-check"></i><b>6.1</b> Preparation</a></li>
<li class="chapter" data-level="6.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#phylogenetic-generalized-least-squares-models-pgls"><i class="fa fa-check"></i><b>6.2</b> Phylogenetic generalized least squares models (PGLS)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#fitting-pgls-models-in-r"><i class="fa fa-check"></i><b>6.2.1</b> Fitting PGLS models in R</a></li>
<li class="chapter" data-level="6.2.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#model-diagnostics-for-pgls-models"><i class="fa fa-check"></i><b>6.2.2</b> Model diagnostics for PGLS models</a></li>
<li class="chapter" data-level="6.2.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#likelihood-profiles-for-lambda-in-pgls-models"><i class="fa fa-check"></i><b>6.2.3</b> Likelihood profiles for <span class="math inline">\(\lambda\)</span> in PGLS models</a></li>
<li class="chapter" data-level="6.2.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#a-pgls-example-with-more-than-one-predictor-variable"><i class="fa fa-check"></i><b>6.2.4</b> A PGLS example with more than one predictor variable</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#common-issues-when-running-pgls-models-in-r"><i class="fa fa-check"></i><b>6.3</b> Common issues when running PGLS models in R</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#optim-errors"><i class="fa fa-check"></i><b>6.3.1</b> Optim errors</a></li>
<li class="chapter" data-level="6.3.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#big-problems-with-small-datasets"><i class="fa fa-check"></i><b>6.3.2</b> Big problems with small datasets</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#summary-5"><i class="fa fa-check"></i><b>6.4</b> Summary</a></li>
<li class="chapter" data-level="6.5" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#practical-exercises-1"><i class="fa fa-check"></i><b>6.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html"><i class="fa fa-check"></i><b>7</b> Phylogenetic generalised linear mixed models in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#before-you-start-5"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="7.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#why-fit-phylogenetic-generalised-linear-mixed-models"><i class="fa fa-check"></i><b>7.1</b> Why fit phylogenetic generalised linear mixed models?</a></li>
<li class="chapter" data-level="7.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#a-really-quick-reminder-of-how-bayesian-analyses-work-with-apologies-to-bayesian-scholars-for-the-simplification"><i class="fa fa-check"></i><b>7.2</b> A really quick reminder of how Bayesian analyses work, with apologies to Bayesian scholars for the simplification!</a></li>
<li class="chapter" data-level="7.3" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#preparation-3"><i class="fa fa-check"></i><b>7.3</b> Preparation</a></li>
<li class="chapter" data-level="7.4" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmcglmm-specific-preparation"><i class="fa fa-check"></i><b>7.4</b> MCMCglmm specific preparation</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#inverse-phylogenetic-variance-covariance-matrix"><i class="fa fa-check"></i><b>7.4.1</b> Inverse phylogenetic variance-covariance matrix</a></li>
<li class="chapter" data-level="7.4.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#priors"><i class="fa fa-check"></i><b>7.4.2</b> Priors</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#other-arguments-in-mcmcglmm"><i class="fa fa-check"></i><b>7.5</b> Other arguments in <code>MCMCglmm</code></a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#nitt"><i class="fa fa-check"></i><b>7.5.1</b> nitt</a></li>
<li class="chapter" data-level="7.5.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#burnin"><i class="fa fa-check"></i><b>7.5.2</b> burnin</a></li>
<li class="chapter" data-level="7.5.3" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#thin"><i class="fa fa-check"></i><b>7.5.3</b> thin</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#fitting-the-mcmcglmm-model"><i class="fa fa-check"></i><b>7.6</b> Fitting the MCMCglmm model</a></li>
<li class="chapter" data-level="7.7" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#did-the-model-converge"><i class="fa fa-check"></i><b>7.7</b> Did the model converge?</a></li>
<li class="chapter" data-level="7.8" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmc-outputs"><i class="fa fa-check"></i><b>7.8</b> MCMC outputs</a></li>
<li class="chapter" data-level="7.9" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#extracting-lambda-for-mcmcglmm-models"><i class="fa fa-check"></i><b>7.9</b> Extracting lambda for MCMCglmm models</a></li>
<li class="chapter" data-level="7.10" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#checking-the-effects-of-priors"><i class="fa fa-check"></i><b>7.10</b> Checking the effects of priors</a></li>
<li class="chapter" data-level="7.11" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#an-example-using-a-pglmm"><i class="fa fa-check"></i><b>7.11</b> An example using a PGLMM</a></li>
<li class="chapter" data-level="7.12" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#adding-extra-fixed-and-random-effects"><i class="fa fa-check"></i><b>7.12</b> Adding extra fixed and random effects</a></li>
<li class="chapter" data-level="7.13" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#summary-6"><i class="fa fa-check"></i><b>7.13</b> Summary</a></li>
<li class="chapter" data-level="7.14" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#practical-exercises-2"><i class="fa fa-check"></i><b>7.14</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html"><i class="fa fa-check"></i><b>8</b> Macroevolutionary models for continuous traits</a>
<ul>
<li class="chapter" data-level="" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#before-you-start-6"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="8.1" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#preparation-4"><i class="fa fa-check"></i><b>8.1</b> Preparation</a></li>
<li class="chapter" data-level="8.2" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#models-of-evolution-for-continuous-traits"><i class="fa fa-check"></i><b>8.2</b> Models of evolution for continuous traits</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#look-at-the-data"><i class="fa fa-check"></i><b>8.2.1</b> Look at the data</a></li>
<li class="chapter" data-level="8.2.2" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#fitting-the-brownian-motion-and-ou-models-of-evolution-using-fitcontinuous"><i class="fa fa-check"></i><b>8.2.2</b> Fitting the Brownian motion and OU models of evolution using <code>fitContinuous</code></a></li>
<li class="chapter" data-level="8.2.3" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#comparing-models-of-evolution-using-aic"><i class="fa fa-check"></i><b>8.2.3</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#summary-7"><i class="fa fa-check"></i><b>8.3</b> Summary</a></li>
<li class="chapter" data-level="8.4" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#practical-exercises-3"><i class="fa fa-check"></i><b>8.4</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html"><i class="fa fa-check"></i><b>9</b> Macroevolutionary models for discrete traits</a>
<ul>
<li class="chapter" data-level="" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#before-you-start-7"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="9.1" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#preparation-5"><i class="fa fa-check"></i><b>9.1</b> Preparation</a></li>
<li class="chapter" data-level="9.2" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#models-of-evolution-for-discrete-traits"><i class="fa fa-check"></i><b>9.2</b> Models of evolution for discrete traits</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#fitting-the-er-sym-and-ard-models-of-evolution-using-fitdiscrete"><i class="fa fa-check"></i><b>9.2.1</b> Fitting the ER, SYM and ARD models of evolution using <code>fitDiscrete</code></a></li>
<li class="chapter" data-level="9.2.2" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#comparing-models-of-evolution-using-aic-1"><i class="fa fa-check"></i><b>9.2.2</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#ancestral-state-estimations"><i class="fa fa-check"></i><b>9.3</b> Ancestral state estimations</a></li>
<li class="chapter" data-level="9.4" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#summary-8"><i class="fa fa-check"></i><b>9.4</b> Summary</a></li>
<li class="chapter" data-level="9.5" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#practical-exercises-4"><i class="fa fa-check"></i><b>9.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html"><i class="fa fa-check"></i><b>10</b> Beyond simple macroevolutionary models</a>
<ul>
<li class="chapter" data-level="" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#before-you-start-8"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="10.1" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#preparation-6"><i class="fa fa-check"></i><b>10.1</b> Preparation</a></li>
<li class="chapter" data-level="10.2" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#more-complex-models-of-trait-evolution"><i class="fa fa-check"></i><b>10.2</b> More complex models of trait evolution</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#preparing-the-tree-and-data-for-ouwie"><i class="fa fa-check"></i><b>10.2.1</b> Preparing the tree and data for OUwie</a></li>
<li class="chapter" data-level="10.2.2" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#running-models-using-ouwie"><i class="fa fa-check"></i><b>10.2.2</b> Running models using OUwie</a></li>
<li class="chapter" data-level="10.2.3" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#single-rate-bm-models"><i class="fa fa-check"></i><b>10.2.3</b> Single rate BM models</a></li>
<li class="chapter" data-level="10.2.4" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#single-optimum-ou-models"><i class="fa fa-check"></i><b>10.2.4</b> Single optimum OU models</a></li>
<li class="chapter" data-level="10.2.5" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-bm-bms-models"><i class="fa fa-check"></i><b>10.2.5</b> Multi-rate BM (BMS) models</a></li>
<li class="chapter" data-level="10.2.6" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-optima-ou-oum-models"><i class="fa fa-check"></i><b>10.2.6</b> Multi-optima OU (OUM) models</a></li>
<li class="chapter" data-level="10.2.7" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-multi-optima-ou-motion-models-oumv"><i class="fa fa-check"></i><b>10.2.7</b> Multi-rate multi-optima OU motion models (OUMV)</a></li>
<li class="chapter" data-level="10.2.8" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-alpha-multi-optima-ou-motion-models-ouma"><i class="fa fa-check"></i><b>10.2.8</b> Multi-alpha multi-optima OU motion models (OUMA)</a></li>
<li class="chapter" data-level="10.2.9" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-multi-optima-ou-motion-models-oumva"><i class="fa fa-check"></i><b>10.2.9</b> Multi-rate multi-optima OU motion models (OUMVA)</a></li>
<li class="chapter" data-level="10.2.10" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#comparing-models-of-evolution-using-aic-2"><i class="fa fa-check"></i><b>10.2.10</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#summary-9"><i class="fa fa-check"></i><b>10.3</b> Summary</a></li>
<li class="chapter" data-level="10.4" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#practical-exercises-5"><i class="fa fa-check"></i><b>10.4</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html"><i class="fa fa-check"></i><b>11</b> Diversification rate analyses</a>
<ul>
<li class="chapter" data-level="" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#before-you-start-9"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="11.1" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#preparation-7"><i class="fa fa-check"></i><b>11.1</b> Preparation</a></li>
<li class="chapter" data-level="11.2" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#lineage-through-time-plots-and-the-gamma-statistic"><i class="fa fa-check"></i><b>11.2</b> Lineage through time plots and the <span class="math inline">\(\gamma\)</span> statistic</a></li>
<li class="chapter" data-level="11.3" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#bamm-bayesian-analysis-of-macroevolutionary-mixtures"><i class="fa fa-check"></i><b>11.3</b> BAMM: Bayesian Analysis of Macroevolutionary Mixtures</a></li>
<li class="chapter" data-level="11.4" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#running-bamm"><i class="fa fa-check"></i><b>11.4</b> Running BAMM</a></li>
<li class="chapter" data-level="11.5" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#looking-at-the-output-of-bamm-using-bammtools-in-r"><i class="fa fa-check"></i><b>11.5</b> Looking at the output of BAMM using BAMMtools in R</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#assessing-mcmc-convergence"><i class="fa fa-check"></i><b>11.5.1</b> Assessing MCMC convergence</a></li>
<li class="chapter" data-level="11.5.2" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#analysis-of-rate-shifts-in-the-dragonfly-phylogeny"><i class="fa fa-check"></i><b>11.5.2</b> Analysis of rate shifts in the dragonfly phylogeny</a></li>
<li class="chapter" data-level="11.5.3" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#mean-phylorate-plot"><i class="fa fa-check"></i><b>11.5.3</b> Mean phylorate plot</a></li>
<li class="chapter" data-level="11.5.4" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#best-single-distinct-shift-configuration"><i class="fa fa-check"></i><b>11.5.4</b> Best single distinct shift configuration</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#credible-sets-of-shifts"><i class="fa fa-check"></i><b>11.6</b> Credible sets of shifts</a></li>
<li class="chapter" data-level="11.7" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#clade-specific-evolutionary-rates"><i class="fa fa-check"></i><b>11.7</b> Clade specific evolutionary rates</a></li>
<li class="chapter" data-level="11.8" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#rate-through-time-analysis"><i class="fa fa-check"></i><b>11.8</b> Rate-through-time analysis</a></li>
<li class="chapter" data-level="11.9" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#macroevolutionary-cohort-analysis"><i class="fa fa-check"></i><b>11.9</b> Macroevolutionary cohort analysis</a></li>
<li class="chapter" data-level="11.10" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#assumptions-and-issues-with-bamm"><i class="fa fa-check"></i><b>11.10</b> Assumptions and issues with BAMM</a></li>
<li class="chapter" data-level="11.11" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#summary-10"><i class="fa fa-check"></i><b>11.11</b> Summary</a></li>
<li class="chapter" data-level="11.12" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#practical-exercises-6"><i class="fa fa-check"></i><b>11.12</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html"><i class="fa fa-check"></i><b>12</b> Answers to practical exercises</a>
<ul>
<li class="chapter" data-level="12.1" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#need-to-know-r"><i class="fa fa-check"></i><b>12.1</b> Need to know R</a></li>
<li class="chapter" data-level="12.2" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#data-wrangling"><i class="fa fa-check"></i><b>12.2</b> Data wrangling</a></li>
<li class="chapter" data-level="12.3" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#phylogenies"><i class="fa fa-check"></i><b>12.3</b> Phylogenies</a></li>
<li class="chapter" data-level="12.4" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#preparation-8"><i class="fa fa-check"></i><b>12.4</b> Preparation</a></li>
<li class="chapter" data-level="12.5" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#phylogenetic-signal"><i class="fa fa-check"></i><b>12.5</b> Phylogenetic signal</a></li>
<li class="chapter" data-level="12.6" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#pgls"><i class="fa fa-check"></i><b>12.6</b> PGLS</a></li>
<li class="chapter" data-level="12.7" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#mcmcglmm"><i class="fa fa-check"></i><b>12.7</b> MCMCglmm</a></li>
<li class="chapter" data-level="12.8" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#models-of-evolution-with-continuous-traits"><i class="fa fa-check"></i><b>12.8</b> Models of evolution with continuous traits</a></li>
<li class="chapter" data-level="12.9" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#models-of-evolution-with-discrete-traits"><i class="fa fa-check"></i><b>12.9</b> Models of evolution with discrete traits</a></li>
<li class="chapter" data-level="12.10" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#beyond-simple-macroevolutionary-models-of-trait-evolution"><i class="fa fa-check"></i><b>12.10</b> Beyond simple macroevolutionary models of trait evolution</a></li>
<li class="chapter" data-level="12.11" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#diversification"><i class="fa fa-check"></i><b>12.11</b> Diversification</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html"><i class="fa fa-check"></i><b>13</b> Common mistakes using PCMs in R</a>
<ul>
<li class="chapter" data-level="13.1" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#good-practice-to-prevent-errors"><i class="fa fa-check"></i><b>13.1</b> Good practice to prevent errors</a></li>
<li class="chapter" data-level="13.2" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#common-errors"><i class="fa fa-check"></i><b>13.2</b> Common errors</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#standard-r-issues"><i class="fa fa-check"></i><b>13.2.1</b> Standard R issues</a></li>
<li class="chapter" data-level="13.2.2" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#issues-with-the-tree-or-data"><i class="fa fa-check"></i><b>13.2.2</b> Issues with the tree or data</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#what-to-do-if-you-get-an-error-message"><i class="fa fa-check"></i><b>13.3</b> What to do if you get an Error message</a></li>
<li class="chapter" data-level="13.4" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#what-to-do-if-you-get-a-warning-message"><i class="fa fa-check"></i><b>13.4</b> What to do if you get a Warning message</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="macroevolutionary-models-for-discrete-traits" class="section level1 hasAnchor" number="9">
<h1><span class="header-section-number">Chapter 9</span> Macroevolutionary models for discrete traits<a href="macroevolutionary-models-for-discrete-traits.html#macroevolutionary-models-for-discrete-traits" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The aims of this exercise are to learn how to use R to fit macroevolutionary models in R to discrete traits.</p>
<div class="info">
<p>
We will be using the evolution of head shape in natricine snakes as
an example. The data and modified tree come from <span class="citation"><span class="citation">V. Deepak, Gower, and Cooper (<a href="#ref-deepak2023" role="doc-biblioref">2023</a>)</span></span>, and the tree comes from <span class="citation"><span class="citation">V. Deepak et al. (<a href="#ref-deepak2021" role="doc-biblioref">2021</a>)</span></span>. I’ve removed a few species and a
few variables to make things a bit more straightforward. If you want to
see the full results check out <span class="citation"><span class="citation">V. Deepak, Gower, and Cooper (<a href="#ref-deepak2023" role="doc-biblioref">2023</a>)</span></span>!
</p>
</div>
<div id="before-you-start-7" class="section level2 unnumbered hasAnchor">
<h2>Before you start<a href="macroevolutionary-models-for-discrete-traits.html#before-you-start-7" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li>Open the <code>09-ModelsDiscrete.RProj</code> file in the <code>09-ModelsDiscrete</code> folder to open your R Project for this exercise.</li>
</ul>
<p>You will also need to install the following packages:</p>
<ul>
<li><code>tidyverse</code> - for reading, manipulating and plotting data</li>
<li><code>ape</code> - functions for reading, plotting and manipulating phylogenies</li>
<li><code>geiger</code> - to fit models of evolution</li>
<li><code>phytools</code> - for plots of transition rates</li>
</ul>
</div>
<div id="preparation-5" class="section level2 hasAnchor" number="9.1">
<h2><span class="header-section-number">9.1</span> Preparation<a href="macroevolutionary-models-for-discrete-traits.html#preparation-5" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To begin we need to load the packages for this practical.</p>
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb343-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb343-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the packages</span></span>
<span id="cb343-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb343-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb343-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb343-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb343-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb343-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb343-5"><a href="macroevolutionary-models-for-discrete-traits.html#cb343-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phytools)</span></code></pre></div>
<p>Next we need to prepare the tree and data for the analyses. In the <code>04-Preparation</code> exercise we read in our tree and data, checked them, and matched them so only species in both were retained. Please refer to that exercise for more details on how and why we do these things, or run through it now if you haven’t previously.</p>
<p>It is important to do these things before beginning a phylogenetic comparative analysis, so let’s run through that code again here.</p>
<div class="sourceCode" id="cb344"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb344-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb344-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the data</span></span>
<span id="cb344-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb344-2" aria-hidden="true" tabindex="-1"></a>snakedata <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/snake-heads.csv&quot;</span>)</span>
<span id="cb344-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb344-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check everything loaded corrected</span></span>
<span id="cb344-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb344-4" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(snakedata)</span></code></pre></div>
<pre><code>## Rows: 206
## Columns: 9
## $ Species          &lt;chr&gt; &quot;Adelophis_foxi&quot;, &quot;Afronatrix_anoscopus_2&quot;, &quot;Amphiesma_beddomei&quot;, &quot;…
## $ Ecomorph         &lt;chr&gt; &quot;Burrowing&quot;, &quot;Aquatic&quot;, &quot;Terrestrial&quot;, &quot;Terrestrial&quot;, &quot;Terrestrial&quot;…
## $ Diet             &lt;chr&gt; &quot;annelids&quot;, &quot;aquatic generalist&quot;, &quot;anurans&quot;, &quot;anurans&quot;, &quot;generalist…
## $ ReproductiveMode &lt;chr&gt; &quot;Viviparous&quot;, &quot;Oviparous&quot;, &quot;Oviparous&quot;, &quot;Oviparous&quot;, &quot;Oviparous&quot;, &quot;…
## $ HeadLength       &lt;dbl&gt; 12.560, 25.100, 15.060, 20.105, 19.950, 18.810, 16.375, 22.640, 17.…
## $ HeadWidth        &lt;dbl&gt; 6.60, 15.80, 7.82, 9.13, 10.32, 9.82, 8.39, 11.30, 9.04, 9.55, 4.49…
## $ HeadHeight       &lt;dbl&gt; 5.43, 10.30, 5.40, 6.61, 7.56, 5.80, 6.40, 8.70, 5.68, 6.64, 4.19, …
## $ EyeDistance      &lt;dbl&gt; 4.01, 7.40, 4.85, 5.85, 5.84, 5.42, 6.00, 6.05, 5.01, 6.39, 3.12, 3…
## $ TotalLength      &lt;dbl&gt; 105, 194, 106, 185, 165, 155, NA, 163, 130, 309, 31, 62, 56, 35, 16…</code></pre>
<p>To load the tree we will use <code>read.nexus</code>.</p>
<div class="sourceCode" id="cb346"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb346-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb346-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the tree</span></span>
<span id="cb346-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb346-2" aria-hidden="true" tabindex="-1"></a>snaketree <span class="ot">&lt;-</span> <span class="fu">read.nexus</span>(<span class="st">&quot;data/snake-tree.nex&quot;</span>)</span>
<span id="cb346-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb346-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check it loaded correctly</span></span>
<span id="cb346-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb346-4" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(snaketree)</span></code></pre></div>
<pre><code>## List of 4
##  $ edge       : int [1:496, 1:2] 250 251 252 253 254 255 256 257 258 259 ...
##  $ edge.length: num [1:496] 4.695 2.225 0.587 0.307 0.832 ...
##  $ Nnode      : int 248
##  $ tip.label  : chr [1:249] &quot;Rhabdophis_leonardi&quot; &quot;Rhabdophis_nuchalis_1&quot; &quot;Rhabdophis_nuchalis_2&quot; &quot;Rhabdophis_adleri_1&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;cladewise&quot;</code></pre>
<p>Remember to check the tree is dichotomous, i.e. has no polytomies, rooted, and ultrametric.</p>
<div class="sourceCode" id="cb348"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb348-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb348-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is binary</span></span>
<span id="cb348-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb348-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE</span></span>
<span id="cb348-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb348-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.binary</span>(snaketree)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb350"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb350-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb350-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is rooted </span></span>
<span id="cb350-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb350-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE </span></span>
<span id="cb350-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb350-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(snaketree)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb352"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb352-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb352-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is ultrametric</span></span>
<span id="cb352-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb352-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE </span></span>
<span id="cb352-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb352-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.ultrametric</span>(snaketree)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Next check that the species names match up in the tree and the data. This should reveal any typos and/or taxonomic differences that need to be fixed before going any further. I haven’t printed <code>check</code> here as we already did this in the previous exercise and it’s a long list but we have already checked these in advance. Do not skip this step for your own analyses!</p>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb354-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb354-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the names match in the data and the tree</span></span>
<span id="cb354-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb354-2" aria-hidden="true" tabindex="-1"></a>check <span class="ot">&lt;-</span> <span class="fu">name.check</span>(<span class="at">phy =</span> snaketree, <span class="at">data =</span> snakedata, </span>
<span id="cb354-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb354-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">data.names =</span> snakedata<span class="sc">$</span>Species)</span>
<span id="cb354-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb354-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check</span></span></code></pre></div>
<p>Finally remove species that are not in the tree and the data and ensure the data is a data frame.</p>
<div class="sourceCode" id="cb355"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb355-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb355-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove species missing from the data</span></span>
<span id="cb355-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb355-2" aria-hidden="true" tabindex="-1"></a>mytree <span class="ot">&lt;-</span> <span class="fu">drop.tip</span>(snaketree, check<span class="sc">$</span>tree_not_data)</span>
<span id="cb355-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb355-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb355-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove species missing from the tree</span></span>
<span id="cb355-5"><a href="macroevolutionary-models-for-discrete-traits.html#cb355-5" aria-hidden="true" tabindex="-1"></a>matches <span class="ot">&lt;-</span> <span class="fu">match</span>(snakedata<span class="sc">$</span>Species, check<span class="sc">$</span>data_not_tree, <span class="at">nomatch =</span> <span class="dv">0</span>)</span>
<span id="cb355-6"><a href="macroevolutionary-models-for-discrete-traits.html#cb355-6" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">subset</span>(snakedata, matches <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb355-7"><a href="macroevolutionary-models-for-discrete-traits.html#cb355-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-8"><a href="macroevolutionary-models-for-discrete-traits.html#cb355-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the tree summary</span></span>
<span id="cb355-9"><a href="macroevolutionary-models-for-discrete-traits.html#cb355-9" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(mytree)</span></code></pre></div>
<pre><code>## List of 4
##  $ edge       : int [1:322, 1:2] 163 164 165 166 167 168 169 170 171 172 ...
##  $ edge.length: num [1:322] 4.695 2.225 0.587 0.307 0.832 ...
##  $ Nnode      : int 161
##  $ tip.label  : chr [1:162] &quot;Rhabdophis_leonardi&quot; &quot;Rhabdophis_nuchalis_3&quot; &quot;Rhabdophis_swinhonis_1&quot; &quot;Rhabdophis_nigrocinctus&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;cladewise&quot;</code></pre>
<div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb357-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb357-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the data</span></span>
<span id="cb357-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb357-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(mydata)</span></code></pre></div>
<pre><code>## Rows: 162
## Columns: 9
## $ Species          &lt;chr&gt; &quot;Adelophis_foxi&quot;, &quot;Afronatrix_anoscopus_2&quot;, &quot;Amphiesma_beddomei&quot;, &quot;…
## $ Ecomorph         &lt;chr&gt; &quot;Burrowing&quot;, &quot;Aquatic&quot;, &quot;Terrestrial&quot;, &quot;Terrestrial&quot;, &quot;Terrestrial&quot;…
## $ Diet             &lt;chr&gt; &quot;annelids&quot;, &quot;aquatic generalist&quot;, &quot;anurans&quot;, &quot;anurans&quot;, &quot;generalist…
## $ ReproductiveMode &lt;chr&gt; &quot;Viviparous&quot;, &quot;Oviparous&quot;, &quot;Oviparous&quot;, &quot;Oviparous&quot;, &quot;Oviparous&quot;, &quot;…
## $ HeadLength       &lt;dbl&gt; 12.560, 25.100, 15.060, 20.105, 18.810, 16.375, 22.640, 19.760, 10.…
## $ HeadWidth        &lt;dbl&gt; 6.60, 15.80, 7.82, 9.13, 9.82, 8.39, 11.30, 9.55, 4.03, 4.90, 3.44,…
## $ HeadHeight       &lt;dbl&gt; 5.43, 10.30, 5.40, 6.61, 5.80, 6.40, 8.70, 6.64, 3.24, 4.41, 3.12, …
## $ EyeDistance      &lt;dbl&gt; 4.01, 7.40, 4.85, 5.85, 5.42, 6.00, 6.05, 6.39, 3.34, 2.27, 1.70, 3…
## $ TotalLength      &lt;dbl&gt; 105, 194, 106, 185, 155, NA, 163, 309, 62, 35, 16, 23, NA, 33, 81, …</code></pre>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb359-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb359-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to a dataframe</span></span>
<span id="cb359-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb359-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(mydata)</span>
<span id="cb359-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb359-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check this is now a data frame</span></span>
<span id="cb359-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb359-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mydata)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<p>Overall we have 162 species in the data and the tree.</p>
<p>Now we’re ready to run our analyses!</p>
</div>
<div id="models-of-evolution-for-discrete-traits" class="section level2 hasAnchor" number="9.2">
<h2><span class="header-section-number">9.2</span> Models of evolution for discrete traits<a href="macroevolutionary-models-for-discrete-traits.html#models-of-evolution-for-discrete-traits" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For fitting simple models of evolution to discrete data we will use the <code>fitDiscrete</code> function in the R package <code>geiger</code>. <code>fitDiscrete</code> is a likelihood based method, so the output will give the maximum likelihood (ML) estimates of the parameters.</p>
<p>As an example, let’s look at the evolution of ecomorphs in our snake data. Before we start let’s make sure we have an idea of what these data look like.</p>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb361-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb361-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How many species are in each category of ecomorph?</span></span>
<span id="cb361-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb361-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="sc">%&gt;%</span></span>
<span id="cb361-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb361-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Ecomorph) <span class="sc">%&gt;%</span></span>
<span id="cb361-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb361-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">number =</span> <span class="fu">n</span>())</span></code></pre></div>
<pre><code>## # A tibble: 5 × 2
##   Ecomorph          number
##   &lt;chr&gt;              &lt;int&gt;
## 1 Aquatic               43
## 2 Aquatic Burrowing     11
## 3 Burrowing             17
## 4 Semiaquatic           54
## 5 Terrestrial           37</code></pre>
<p>We can also visualize these variables on our tree by plotting them with colours. We’ll then create a vector of five colours for the ecomorphs. These need to be in alphabetical order, so the colours should be in the order Aquatic, Aquatic burrower, Burrower, Semi-aquatic, Terrestrial.</p>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb363-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb363-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up list of colours</span></span>
<span id="cb363-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb363-2" aria-hidden="true" tabindex="-1"></a>ecomorph.colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;darkblue&quot;</span>, <span class="st">&quot;plum&quot;</span>, <span class="st">&quot;chocolate&quot;</span>, <span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;lightgreen&quot;</span>)</span></code></pre></div>
<p>We then need to make sure the variable <code>Ecomorph</code> is a factor rather than a character as R deals with these differently and we need it to be a factor for the next bit of code to work. We can do this using the function <code>as.factor</code> and we can change it in <code>mydata</code> usign the <code>tidyverse</code> function <code>mutate</code>.</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb364-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make Ecomorph a factor</span></span>
<span id="cb364-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb364-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> </span>
<span id="cb364-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb364-3" aria-hidden="true" tabindex="-1"></a>  mydata <span class="sc">%&gt;%</span></span>
<span id="cb364-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb364-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Ecomorph =</span> <span class="fu">as.factor</span>(Ecomorph))</span></code></pre></div>
<p>Finally we can plot the tree with coloured tip labels to match the ecomorphs. I’ve rearranged the colour order in the legend so it makes more sense biologically (i.e. putting the aquatic and semi-aquatic next to each other rather than ordering it alphabetically). To make things a bit easier to read I’ve left off the tip labels. I’ve done this using <code>ape</code> plotting functions but you can also do this using <code>ggtree</code>.</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb365-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the tree with colours at the tips to represent ecomorph</span></span>
<span id="cb365-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb365-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(mytree, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">adj =</span> <span class="fu">c</span>(<span class="fl">0.2</span>), <span class="at">type =</span> <span class="st">&quot;fan&quot;</span>, </span>
<span id="cb365-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb365-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">no.margin =</span> <span class="cn">TRUE</span>, <span class="at">show.tip.label =</span> <span class="cn">FALSE</span>)</span>
<span id="cb365-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb365-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tiplabels</span>(<span class="at">pch =</span> <span class="dv">16</span>, <span class="at">col =</span> ecomorph.colors[mydata<span class="sc">$</span>Ecomorph])</span>
<span id="cb365-5"><a href="macroevolutionary-models-for-discrete-traits.html#cb365-5" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomleft&quot;</span>, <span class="at">pch =</span> <span class="dv">15</span>, <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb365-6"><a href="macroevolutionary-models-for-discrete-traits.html#cb365-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">&quot;Aquatic&quot;</span>, <span class="st">&quot;Semi-aquatic&quot;</span>, <span class="st">&quot;Aquatic burrower&quot;</span>, </span>
<span id="cb365-7"><a href="macroevolutionary-models-for-discrete-traits.html#cb365-7" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;Burrower&quot;</span>, <span class="st">&quot;Terrestrial&quot;</span>), </span>
<span id="cb365-8"><a href="macroevolutionary-models-for-discrete-traits.html#cb365-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;darkblue&quot;</span>, <span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;plum&quot;</span>, </span>
<span id="cb365-9"><a href="macroevolutionary-models-for-discrete-traits.html#cb365-9" aria-hidden="true" tabindex="-1"></a>               <span class="st">&quot;chocolate&quot;</span>, <span class="st">&quot;lightgreen&quot;</span>))</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-252-1.png" width="672" /></p>
<p>Some ecomorphs are clustered, but others appear across the phylogeny (this is the result of some pretty cool convergent evolution - check out <span class="citation">V. Deepak, Gower, and Cooper (<a href="#ref-deepak2023" role="doc-biblioref">2023</a>)</span> for more details).</p>
<div id="fitting-the-er-sym-and-ard-models-of-evolution-using-fitdiscrete" class="section level3 hasAnchor" number="9.2.1">
<h3><span class="header-section-number">9.2.1</span> Fitting the ER, SYM and ARD models of evolution using <code>fitDiscrete</code><a href="macroevolutionary-models-for-discrete-traits.html#fitting-the-er-sym-and-ard-models-of-evolution-using-fitdiscrete" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We’ll fit three commonly used evolutionary models to the data; the equal rates (ER) model, the symmetric rates (SYM) model and the all rates different (ARD) model.</p>
<p>As a quick reminder, in the ER model, transitions between any pair of states occur at the same rate and are therefore equally probable. This is also known as the Mk1 model because it as a Markov model (Mk) with only 1 parameter (k).</p>
<p>In the SYM model transitions between any pair of states occurs at the same rate regardless of direction, but that the rate of change differs among states. A symmetric model with only 2 states becomes an equal rates (Mk1) model.</p>
<p>Finally, in the ARD model all transitions between pairs of states occurs at different rates. It’s important to be wary of over-fitting with this model as it can result in a lot of rates when you have more than a few states.</p>
<p>For more details on these models please see the <em>Primer</em>.</p>
<p>We are going to use the package <code>geiger</code> to fit models in this exercise, so we need the additional preparation step we first met in chapter <code>05-PhyloSignal</code>. We need to create an object in R that only contains the variable required, and the species names (so we can match it up to the tree).</p>
<p>We use the function <code>pull</code> to extract just the ecomorph values. We then name these values with the species names from <code>mydata</code> using the function <code>names</code>. Note that this requires the trait data is in the <em>same order</em> as the tree tip labels so we need to reorder the data first…</p>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb366-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb366-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check first few tip labels and species</span></span>
<span id="cb366-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb366-2" aria-hidden="true" tabindex="-1"></a>mytree<span class="sc">$</span>tip.label[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] &quot;Rhabdophis_leonardi&quot;     &quot;Rhabdophis_nuchalis_3&quot;   &quot;Rhabdophis_swinhonis_1&quot; 
## [4] &quot;Rhabdophis_nigrocinctus&quot; &quot;Rhabdophis_tigrinus&quot;</code></pre>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb368-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb368-1" aria-hidden="true" tabindex="-1"></a>mydata<span class="sc">$</span>Species[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] &quot;Adelophis_foxi&quot;         &quot;Afronatrix_anoscopus_2&quot; &quot;Amphiesma_beddomei&quot;    
## [4] &quot;Amphiesma_monticola&quot;    &quot;Amphiesma_stolatum_1&quot;</code></pre>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb370-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb370-1" aria-hidden="true" tabindex="-1"></a><span class="co"># These are different so we reorder the data by the tips</span></span>
<span id="cb370-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb370-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> mydata[<span class="fu">match</span>(mytree<span class="sc">$</span>tip.label, mydata<span class="sc">$</span>Species), ]</span>
<span id="cb370-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb370-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check this now matches the tip label order</span></span>
<span id="cb370-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb370-4" aria-hidden="true" tabindex="-1"></a>mydata<span class="sc">$</span>Species[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] &quot;Rhabdophis_leonardi&quot;     &quot;Rhabdophis_nuchalis_3&quot;   &quot;Rhabdophis_swinhonis_1&quot; 
## [4] &quot;Rhabdophis_nigrocinctus&quot; &quot;Rhabdophis_tigrinus&quot;</code></pre>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb372-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create ecomorph containing just ecomorph values</span></span>
<span id="cb372-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb372-2" aria-hidden="true" tabindex="-1"></a>ecomorph <span class="ot">&lt;-</span> <span class="fu">pull</span>(mydata, Ecomorph)</span>
<span id="cb372-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb372-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Make sure ecomorph is a character, not a factor</span></span>
<span id="cb372-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb372-4" aria-hidden="true" tabindex="-1"></a>ecomorph <span class="ot">&lt;-</span> <span class="fu">as.character</span>(ecomorph)</span>
<span id="cb372-5"><a href="macroevolutionary-models-for-discrete-traits.html#cb372-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the first few rows</span></span>
<span id="cb372-6"><a href="macroevolutionary-models-for-discrete-traits.html#cb372-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ecomorph)</span></code></pre></div>
<pre><code>## [1] &quot;Semiaquatic&quot; &quot;Terrestrial&quot; &quot;Semiaquatic&quot; &quot;Terrestrial&quot; &quot;Terrestrial&quot; &quot;Terrestrial&quot;</code></pre>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb374-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Give log head length names = species names at the tips of the phylogeny</span></span>
<span id="cb374-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb374-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(ecomorph) <span class="ot">&lt;-</span> mydata<span class="sc">$</span>Species</span>
<span id="cb374-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb374-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the first few rows</span></span>
<span id="cb374-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb374-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ecomorph)</span></code></pre></div>
<pre><code>##           Rhabdophis_leonardi         Rhabdophis_nuchalis_3        Rhabdophis_swinhonis_1 
##                 &quot;Semiaquatic&quot;                 &quot;Terrestrial&quot;                 &quot;Semiaquatic&quot; 
##       Rhabdophis_nigrocinctus           Rhabdophis_tigrinus Rhabdophis_tigrinus_lateralis 
##                 &quot;Terrestrial&quot;                 &quot;Terrestrial&quot;                 &quot;Terrestrial&quot;</code></pre>
<p>Now we have a list of values with associated species names.</p>
<p>To fit the ER model we can then use the code below.</p>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb376-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb376-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the ER model</span></span>
<span id="cb376-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb376-2" aria-hidden="true" tabindex="-1"></a>equal <span class="ot">&lt;-</span> <span class="fu">fitDiscrete</span>(mytree, ecomorph, <span class="at">model =</span> <span class="st">&quot;ER&quot;</span>)</span></code></pre></div>
<p>Let’s look at the output for the equal rates model:</p>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb377-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb377-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the output</span></span>
<span id="cb377-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb377-2" aria-hidden="true" tabindex="-1"></a>equal</span></code></pre></div>
<pre><code>## GEIGER-fitted comparative model of discrete data
##  fitted Q matrix:
##                            Aquatic Aquatic Burrowing    Burrowing  Semiaquatic  Terrestrial
##     Aquatic           -0.034973712       0.008743428  0.008743428  0.008743428  0.008743428
##     Aquatic Burrowing  0.008743428      -0.034973712  0.008743428  0.008743428  0.008743428
##     Burrowing          0.008743428       0.008743428 -0.034973712  0.008743428  0.008743428
##     Semiaquatic        0.008743428       0.008743428  0.008743428 -0.034973712  0.008743428
##     Terrestrial        0.008743428       0.008743428  0.008743428  0.008743428 -0.034973712
## 
##  model summary:
##  log-likelihood = -176.839091
##  AIC = 355.678181
##  AICc = 355.703181
##  free parameters = 1
## 
## Convergence diagnostics:
##  optimization iterations = 100
##  failed iterations = 0
##  number of iterations with same best fit = 100
##  frequency of best fit = 1.00
## 
##  object summary:
##  &#39;lik&#39; -- likelihood function
##  &#39;bnd&#39; -- bounds for likelihood search
##  &#39;res&#39; -- optimization iteration summary
##  &#39;opt&#39; -- maximum likelihood parameter estimates</code></pre>
<p>This looks very similar to the output from <code>fitContinuous</code> in the <code>08-ModelsContinuous</code> exercise. We’ve got a model summary, with log likelihoods and AIC scores, convergence diagnostics and an object summary. The only major difference is the first part, which gives us a fitted transition matrix or <span class="math inline">\(Q\)</span> matrix, rather than a summary of model parameters. In fact the <span class="math inline">\(Q\)</span> matrix <strong>is</strong> the model parameters for these models as it contains all the rates of change between different states. This is an equal rates model, so the off-diagonal elements are all same and show that the rate of change from one state to any other state is 0.00874.</p>
<div class="detail">
<p>
Why are the diagonal elements of the <span class="math inline"><span class="math inline">\(Q\)</span></span> matrix -0.034973712? The rows of the
<span class="math inline"><span class="math inline">\(Q\)</span></span> matrix must sum to zero:
-0.034973712 + (4 * 0.008743428) = 0.
</p>
</div>
<p>To fit the SYM model we just switch the model name. Note that this might be a bit slow on your computer as we have quite a lot of different ecomorphs, so a lot of different rates to fit. A good point to grab a coffee/tea/biscuit!</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb379-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb379-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the SYM model</span></span>
<span id="cb379-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb379-2" aria-hidden="true" tabindex="-1"></a>sym <span class="ot">&lt;-</span> <span class="fu">fitDiscrete</span>(mytree, ecomorph, <span class="at">model =</span> <span class="st">&quot;SYM&quot;</span>)</span>
<span id="cb379-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb379-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the output</span></span>
<span id="cb379-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb379-4" aria-hidden="true" tabindex="-1"></a>sym</span></code></pre></div>
<pre><code>## GEIGER-fitted comparative model of discrete data
##  fitted Q matrix:
##                             Aquatic Aquatic Burrowing     Burrowing   Semiaquatic
##     Aquatic           -3.410257e-02      6.787882e-03  8.536690e-03  1.877800e-02
##     Aquatic Burrowing  6.787882e-03     -6.787882e-03  1.163160e-19  3.309475e-18
##     Burrowing          8.536690e-03      1.163160e-19 -1.172452e-02  3.187830e-03
##     Semiaquatic        1.877800e-02      3.309475e-18  3.187830e-03 -6.519697e-02
##     Terrestrial        1.664242e-23      9.507585e-18  2.981126e-19  4.323114e-02
##                         Terrestrial
##     Aquatic            1.664242e-23
##     Aquatic Burrowing  9.507585e-18
##     Burrowing          2.981126e-19
##     Semiaquatic        4.323114e-02
##     Terrestrial       -4.323114e-02
## 
##  model summary:
##  log-likelihood = -148.281974
##  AIC = 316.563949
##  AICc = 318.020902
##  free parameters = 10
## 
## Convergence diagnostics:
##  optimization iterations = 100
##  failed iterations = 0
##  number of iterations with same best fit = 4
##  frequency of best fit = 0.04
## 
##  object summary:
##  &#39;lik&#39; -- likelihood function
##  &#39;bnd&#39; -- bounds for likelihood search
##  &#39;res&#39; -- optimization iteration summary
##  &#39;opt&#39; -- maximum likelihood parameter estimates</code></pre>
<p>Again the <span class="math inline">\(Q\)</span> matrix is the key output here. You could just read the numbers, we can see this more clearly if we plot the <span class="math inline">\(Q\)</span> matrix for the model:</p>
<div class="sourceCode" id="cb381"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb381-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb381-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot Q matrix as network</span></span>
<span id="cb381-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb381-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.gfit</span>(sym)</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-259-1.png" width="672" />
The transition rate between Semi-aquatic and Terrestrial is highest (0.043), followed by transitions between Semi-aquatic and Aquatic (0.019). Burrowing and Aquatic (0.009) and Aquatic Burrowing and Aquatic (0.007) have the next highest rates, followed by Burrowing and Semi-aquatic (0.003). All the other rates are &lt; 0.001 so they appear as zeros on this plot</p>
<p>Finally let’s fit the ARD model. Again this may take some time! Go and grab a tea/coffee/biscuit or take a short break.</p>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb382-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the ARD model</span></span>
<span id="cb382-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb382-2" aria-hidden="true" tabindex="-1"></a>ard <span class="ot">&lt;-</span> <span class="fu">fitDiscrete</span>(mytree, ecomorph, <span class="at">model =</span> <span class="st">&quot;ARD&quot;</span>)</span>
<span id="cb382-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb382-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the output</span></span>
<span id="cb382-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb382-4" aria-hidden="true" tabindex="-1"></a>ard</span></code></pre></div>
<pre><code>## GEIGER-fitted comparative model of discrete data
##  fitted Q matrix:
##                             Aquatic Aquatic Burrowing     Burrowing   Semiaquatic
##     Aquatic           -3.871262e-02      7.710796e-03  1.070672e-02  2.029510e-02
##     Aquatic Burrowing  3.070007e-16     -3.085515e-16  1.550706e-18  7.000260e-38
##     Burrowing          4.008351e-19      4.966449e-23 -4.116883e-19  4.081239e-23
##     Semiaquatic        1.592909e-02      3.156430e-19  2.496600e-03 -5.538242e-02
##     Terrestrial        2.472419e-19      3.097357e-21  1.042375e-20  2.673830e-03
##                         Terrestrial
##     Aquatic            4.967164e-21
##     Aquatic Burrowing  2.883901e-24
##     Burrowing          1.076274e-20
##     Semiaquatic        3.695673e-02
##     Terrestrial       -2.673830e-03
## 
##  model summary:
##  log-likelihood = -145.506018
##  AIC = 331.012035
##  AICc = 336.969482
##  free parameters = 20
## 
## Convergence diagnostics:
##  optimization iterations = 100
##  failed iterations = 0
##  number of iterations with same best fit = 1
##  frequency of best fit = 0.01
## 
##  object summary:
##  &#39;lik&#39; -- likelihood function
##  &#39;bnd&#39; -- bounds for likelihood search
##  &#39;res&#39; -- optimization iteration summary
##  &#39;opt&#39; -- maximum likelihood parameter estimates</code></pre>
<p>Again the <span class="math inline">\(Q\)</span> matrix is the key output. We can plot the <span class="math inline">\(Q\)</span> matrix for the model:</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb384-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot Q matrix as network</span></span>
<span id="cb384-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb384-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.gfit</span>(ard)</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-261-1.png" width="672" />
This is quite complex because in the ARD model we have different rates for each type of state change. Here, for example, the bottom left hand arrows joining Semi-aquatic and Terrestrial show that rates of change from Semi-aquatic to Terrestrial (0.037) are higher than those from Terrestrial to Semi-aquatic (0.003). These plots can be a little easier to read if we hide all the transitions where rates are &lt; 0.001 as follows.</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb385-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot Q matrix as network without zeros</span></span>
<span id="cb385-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb385-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.gfit</span>(ard, <span class="at">show.zeros =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-262-1.png" width="672" /></p>
<div class="detail">
<p>
For now these plots are pretty simple, but you can add some code to
make the arrows different sizes etc. if you need to using the code here
<a href="http://blog.phytools.org/2020/09/graphing-fitted-m-k-model-with.html" class="uri">http://blog.phytools.org/2020/09/graphing-fitted-m-k-model-with.html</a>.
Also check out future versions of <code>phytools</code> which should
have these functions added.
</p>
</div>
</div>
<div id="comparing-models-of-evolution-using-aic-1" class="section level3 hasAnchor" number="9.2.2">
<h3><span class="header-section-number">9.2.2</span> Comparing models of evolution using AIC<a href="macroevolutionary-models-for-discrete-traits.html#comparing-models-of-evolution-using-aic-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As we discussed in the continuous traits exercise in <code>08-ModelsContinuous</code>, we often want to know which of the models fits our variable best. We can use <code>fitDiscrete</code> to fit the models we are interested in and then compare them using AIC. We can extract the AICs from the models we fitted above as follows:</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb386-1" aria-hidden="true" tabindex="-1"></a>equal<span class="sc">$</span>opt<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 355.6782</code></pre>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb388-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb388-1" aria-hidden="true" tabindex="-1"></a>sym<span class="sc">$</span>opt<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 316.5639</code></pre>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb390-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb390-1" aria-hidden="true" tabindex="-1"></a>ard<span class="sc">$</span>opt<span class="sc">$</span>aic</span></code></pre></div>
<pre><code>## [1] 331.012</code></pre>
<p>The “best” model is the one with the smallest AIC. Here the best model of evolution appears to be the SYM model, whereby rates of transition are the same in both directions, but are different for the different pairs of ecomorphs.</p>
<p>Alternatively we can use <span class="math inline">\(\Delta\)</span>AIC or AIC weights to compare our models using the following code and the <code>geiger</code> function <code>aicw</code>:</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb392-1" aria-hidden="true" tabindex="-1"></a>aic.scores <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">c</span>(equal<span class="sc">$</span>opt<span class="sc">$</span>aic, sym<span class="sc">$</span>opt<span class="sc">$</span>aic, ard<span class="sc">$</span>opt<span class="sc">$</span>aic), </span>
<span id="cb392-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb392-2" aria-hidden="true" tabindex="-1"></a>                       <span class="fu">c</span>(<span class="st">&quot;equal&quot;</span>, <span class="st">&quot;symmetric&quot;</span>, <span class="st">&quot;different&quot;</span>))</span>
<span id="cb392-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb392-3" aria-hidden="true" tabindex="-1"></a><span class="fu">aicw</span>(aic.scores)</span></code></pre></div>
<pre><code>##                fit    delta            w
## equal     355.6782 39.11423 3.207273e-09
## symmetric 316.5639  0.00000 9.992717e-01
## different 331.0120 14.44809 7.283186e-04</code></pre>
<p><code>aicw</code> outputs the AIC (<code>fit</code>), <span class="math inline">\(\Delta\)</span>AIC (<code>delta</code>) and AIC weights (<code>w</code>) for each of the models we fitted. The best model is the model with <span class="math inline">\(\Delta\)</span>AIC = 0 or with AICw closest to 1. Using <span class="math inline">\(\Delta\)</span>AIC we can conclude that the SYM model is the best fit to the data.</p>
</div>
</div>
<div id="ancestral-state-estimations" class="section level2 hasAnchor" number="9.3">
<h2><span class="header-section-number">9.3</span> Ancestral state estimations<a href="macroevolutionary-models-for-discrete-traits.html#ancestral-state-estimations" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>I’m not a big fan of ancestral state estimation as you probably realised from the <em>Primer</em>! There are many reasons to be highly sceptical of ancestral state estimates and interpretations of macroevolutionary patterns and process that are based on them. However, if you want to know if evolutionary tempo or mode have varied over clade history based on the state of a discrete trait as we intend to do in the next chapter, you’ll need to do it so you can assign each node to a particular state.</p>
<p>We’ll use <code>ape</code>’s <code>ace</code> function here. There are other options out there, for example in the <code>phytools</code> package, but <code>ace</code> will work for our purposes.</p>
<p>To perform ancestral state estimation of ecomorphs for snakes we can use the <code>ecomorph</code> object we made earlier and used to fit the ER, SYM and ARD models. Recall that <code>ecomorph</code> is just the Ecomorph column from the snake heads dataset, in the same order as the tips in the tree, with names equal to the tip labels/Species names in the tree.</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb394-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb394-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a quick look at the first few rows of ecomorph</span></span>
<span id="cb394-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb394-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ecomorph)</span></code></pre></div>
<pre><code>##           Rhabdophis_leonardi         Rhabdophis_nuchalis_3        Rhabdophis_swinhonis_1 
##                 &quot;Semiaquatic&quot;                 &quot;Terrestrial&quot;                 &quot;Semiaquatic&quot; 
##       Rhabdophis_nigrocinctus           Rhabdophis_tigrinus Rhabdophis_tigrinus_lateralis 
##                 &quot;Terrestrial&quot;                 &quot;Terrestrial&quot;                 &quot;Terrestrial&quot;</code></pre>
<p>To perform an ancestral state estimation of ecomorphs under the symmetric model (the best fitting model for the data as discovered above) we use this code:</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb396-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb396-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ancestral state estimation of ecomorph under the SYM model</span></span>
<span id="cb396-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb396-2" aria-hidden="true" tabindex="-1"></a>ancestral_ecomorphs <span class="ot">&lt;-</span> <span class="fu">ace</span>(ecomorph, mytree, <span class="at">type =</span> <span class="st">&quot;discrete&quot;</span>, <span class="at">model =</span> <span class="st">&quot;SYM&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in log(comp[-TIPS]): NaNs produced</code></pre>
<pre><code>## Warning in sqrt(diag(solve(h))): NaNs produced</code></pre>
<p>You should notice a series of warning messages appear that say <code>NaNs produced</code> and <code>NA/Inf replaced by maximum positive value</code>. Don’t worry about this – it happens when rates for one transition are particularly low (and we know from above that many of the transition rates were &lt; 0.001) but doesn’t really affect our node state estimates.</p>
<div class="detail">
<p>
<code>ace</code> now defaults to a joint estimation procedure, where
the ancestral states are optimized based on all information in the tree,
not just the states at descendant tips. Older versions of
<code>ace</code> did not do this, so be wary of this when reading older
critiques of this method
</p>
</div>
<p>We can look at the first few rows of our ancestral state estimates by typing:</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb399-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb399-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ancestral_ecomorphs<span class="sc">$</span>lik.anc)</span></code></pre></div>
<p>In this matrix the rows correspond to nodes in the tree (although the numbering is bit off; we don’t need to worry about this here but it will come up in the next exercise) and the five columns give the <em>scaled likelihoods</em> that the node is in each of the different ecomorphs.</p>
<p>The scaled likelihoods are like probabilities, so for the first node, the probability of it being Aquatic is highest (probability = 0.5417), followed by the probability of it being Semi-aquatic (0.3587), Terrestrial (0.0755), Burrowing (0.0206), or Aquatic Burrowing (0.0035). These scaled likelihoods can be visualized with pie charts on the tree we plotted earlier byt adding the <code>nodelabels</code> line of code below.</p>
<div class="sourceCode" id="cb400"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb400-1"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the tree with colours at the tips and nodes to represent ecomorph</span></span>
<span id="cb400-2"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(mytree, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">adj =</span> <span class="fu">c</span>(<span class="fl">0.2</span>), <span class="at">type =</span> <span class="st">&quot;fan&quot;</span>, </span>
<span id="cb400-3"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">no.margin =</span> <span class="cn">TRUE</span>, <span class="at">show.tip.label =</span> <span class="cn">FALSE</span>)</span>
<span id="cb400-4"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tiplabels</span>(<span class="at">pch =</span> <span class="dv">16</span>, <span class="at">col =</span> ecomorph.colors[mydata<span class="sc">$</span>Ecomorph])</span>
<span id="cb400-5"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-5" aria-hidden="true" tabindex="-1"></a><span class="fu">nodelabels</span>(<span class="at">pie =</span> ancestral_ecomorphs<span class="sc">$</span>lik.anc, <span class="at">piecol =</span> ecomorph.colors, <span class="at">cex =</span> <span class="fl">0.5</span>)</span>
<span id="cb400-6"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-6" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomleft&quot;</span>, <span class="at">pch =</span> <span class="dv">15</span>, <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>,</span>
<span id="cb400-7"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">&quot;Aquatic&quot;</span>, <span class="st">&quot;Semi-aquatic&quot;</span>, <span class="st">&quot;Aquatic burrower&quot;</span>, </span>
<span id="cb400-8"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-8" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;Burrower&quot;</span>, <span class="st">&quot;Terrestrial&quot;</span>), </span>
<span id="cb400-9"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;darkblue&quot;</span>, <span class="st">&quot;lightblue&quot;</span>, <span class="st">&quot;plum&quot;</span>, </span>
<span id="cb400-10"><a href="macroevolutionary-models-for-discrete-traits.html#cb400-10" aria-hidden="true" tabindex="-1"></a>               <span class="st">&quot;chocolate&quot;</span>, <span class="st">&quot;lightgreen&quot;</span>))</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-270-1.png" width="672" /></p>
<p>Aquatic is most commonly estimated as the ancestral ecomorph for natricine snakes, with Semi-aquatic as a close second, and with multiple transitions to other ecomorphs throughout the tree.</p>
<div class="warning">
<p>
Ancestral state estimates (<strong>not</strong> reconstructions) are,
at best, weighted averages of your trait based on branch lengths of the
phylogeny and the model of evolution used. If your phylogeny or traits
or model have error (and they always will) then your ancestral state
estimates should be interpreted with care.
</p>
</div>
</div>
<div id="summary-8" class="section level2 hasAnchor" number="9.4">
<h2><span class="header-section-number">9.4</span> Summary<a href="macroevolutionary-models-for-discrete-traits.html#summary-8" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>You should now know how to fit simple models of evolution to discrete trait data in R.</p>
</div>
<div id="practical-exercises-4" class="section level2 hasAnchor" number="9.5">
<h2><span class="header-section-number">9.5</span> Practical exercises<a href="macroevolutionary-models-for-discrete-traits.html#practical-exercises-4" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Read in the tree and data, then prepare them for a PCM analysis (you may have already done this in a previous exercise which should save you some time). Fit equal rates, symmetric rates and all rates different models to the social status variable for Primates (<code>SocialStatus</code>). Note that social status has two states: non-social (1) or social (2).</p>
<p>Then answer the following questions.</p>
<ol style="list-style-type: decimal">
<li><p>Which model fits best?</p></li>
<li><p>What is the transition rate from non-social (1) to social (2), and vice versa, in the all rates different model? What does this mean biologically?</p></li>
<li><p>Plot ancestral state estimates of social status (from the ARD model) on the phylogeny. Is the root node estimated as being non-social or social?</p></li>
<li><p>Why didn’t I ask you to fit the symmetric model?</p></li>
</ol>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-arnold201010ktrees" class="csl-entry">
Arnold, Christian, Luke J Matthews, and Charles L Nunn. 2010. <span>“The 10kTrees Website: A New Online Resource for Primate Phylogeny.”</span> <em>Evolutionary Anthropology: Issues, News, and Reviews</em> 19 (3): 114–18.
</div>
<div id="ref-deepak2021" class="csl-entry">
Deepak, V, Natalie Cooper, Nikolay A Poyarkov, Fred Kraus, Gustavo Burin, Abhijit Das, Surya Narayanan, Jeffrey W Streicher, Sarah-Jane Smith, and David J Gower. 2021. <span>“<span class="nocase">Multilocus phylogeny, natural history traits and classification of natricine snakes (Serpentes: Natricinae)</span>.”</span> <em>Zoological Journal of the Linnean Society</em> 195 (1): 279–98. <a href="https://doi.org/10.1093/zoolinnean/zlab099">https://doi.org/10.1093/zoolinnean/zlab099</a>.
</div>
<div id="ref-deepak2023" class="csl-entry">
Deepak, V., D. J. Gower, and N. Cooper. 2023. <span>“Diet and Habit Explain Head-Shape Convergences in Natricine Snakes.”</span> <em>Journal of Evolutionary Biology</em> 36 (2): 399–411. https://doi.org/<a href="https://doi.org/10.1111/jeb.14139">https://doi.org/10.1111/jeb.14139</a>.
</div>
<div id="ref-jones2009pantheria" class="csl-entry">
Jones, Kate E, Jon Bielby, Marcel Cardillo, Susanne A Fritz, Justin O’Dell, C David L Orme, Kamran Safi, et al. 2009. <span>“PanTHERIA: A Species-Level Database of Life History, Ecology, and Geography of Extant and Recently Extinct Mammals: Ecological Archives E090-184.”</span> <em>Ecology</em> 90 (9): 2648–48.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="macroevolutionary-models-for-continuous-traits.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="beyond-simple-macroevolutionary-models.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/09-ModelsDiscrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["pcm-primer.pdf", "pcm-primer.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
