<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Phylogenetic generalised linear mixed models in R | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences</title>
  <meta name="description" content="Chapter 7 Phylogenetic generalised linear mixed models in R | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Phylogenetic generalised linear mixed models in R | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Phylogenetic generalised linear mixed models in R | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  
  
  

<meta name="author" content="Natalie Cooper" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="phylogenetic-generalised-least-squares-pgls-in-r.html"/>
<link rel="next" href="macroevolutionary-models-for-continuous-traits.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">PCM Primer</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#datasets-and-scripts"><i class="fa fa-check"></i>Datasets and scripts</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#how-to-use-these-materials"><i class="fa fa-check"></i>How to use these materials</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#boxes"><i class="fa fa-check"></i>Boxes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#example-datasets"><i class="fa fa-check"></i>Example datasets</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#frog-eye-size-evolution"><i class="fa fa-check"></i>Frog eye size evolution</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#natricine-snake-head-shape-evolution"><i class="fa fa-check"></i>Natricine snake head shape evolution</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#diversification-in-dragonflies"><i class="fa fa-check"></i>Diversification in dragonflies</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#citing-r-and-r-packages"><i class="fa fa-check"></i>Citing R and R packages</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#acknowledgements"><i class="fa fa-check"></i>Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><i class="fa fa-check"></i><b>1</b> What you need to be able to do in R before you start</a>
<ul>
<li class="chapter" data-level="1.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-r-and-rstudio"><i class="fa fa-check"></i><b>1.1</b> Installing R (and RStudio)</a></li>
<li class="chapter" data-level="1.2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#downloading-the-data"><i class="fa fa-check"></i><b>1.2</b> Downloading the data</a></li>
<li class="chapter" data-level="1.3" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#r-projects"><i class="fa fa-check"></i><b>1.3</b> R Projects</a></li>
<li class="chapter" data-level="1.4" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#using-a-script"><i class="fa fa-check"></i><b>1.4</b> Using a script</a></li>
<li class="chapter" data-level="1.5" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-and-loading-extra-packages-in-r"><i class="fa fa-check"></i><b>1.5</b> Installing and loading extra packages in R</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#loading-packages-into-r"><i class="fa fa-check"></i><b>1.5.1</b> Loading packages into R</a></li>
<li class="chapter" data-level="1.5.2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#ggtree-a-special-case-for-package-installation"><i class="fa fa-check"></i><b>1.5.2</b> <code>ggtree</code>, a special case for package installation</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#loading-and-viewing-your-data-in-r"><i class="fa fa-check"></i><b>1.6</b> Loading and viewing your data in R</a></li>
<li class="chapter" data-level="1.7" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#errors-and-warnings"><i class="fa fa-check"></i><b>1.7</b> Errors and warnings</a>
<ul>
<li class="chapter" data-level="1.7.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#removing-the"><i class="fa fa-check"></i><b>1.7.1</b> Removing the <code>+</code></a></li>
</ul></li>
<li class="chapter" data-level="1.8" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#summary"><i class="fa fa-check"></i><b>1.8</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html"><i class="fa fa-check"></i><b>2</b> Data wrangling comparative data in R</a>
<ul>
<li class="chapter" data-level="" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#before-you-start"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="2.1" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#chameleon-data"><i class="fa fa-check"></i><b>2.1</b> Chameleon data</a></li>
<li class="chapter" data-level="2.2" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#summary-1"><i class="fa fa-check"></i><b>2.2</b> Summary</a></li>
<li class="chapter" data-level="2.3" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#practical-exercise"><i class="fa fa-check"></i><b>2.3</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html"><i class="fa fa-check"></i><b>3</b> Understanding, plotting and manipulating phylogenies in R</a>
<ul>
<li class="chapter" data-level="" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#before-you-start-1"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="3.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#what-do-phylogeny-files-look-like"><i class="fa fa-check"></i><b>3.1</b> What do phylogeny files look like?</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#newick-trees"><i class="fa fa-check"></i><b>3.1.1</b> Newick trees</a></li>
<li class="chapter" data-level="3.1.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#nexus-trees"><i class="fa fa-check"></i><b>3.1.2</b> NEXUS trees</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#how-are-phylogenies-formatted-in-r"><i class="fa fa-check"></i><b>3.2</b> How are phylogenies formatted in R?</a></li>
<li class="chapter" data-level="3.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#plotting-phylogenies"><i class="fa fa-check"></i><b>3.3</b> Plotting phylogenies</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#basic-phylogeny-plotting-with-ape"><i class="fa fa-check"></i><b>3.3.1</b> Basic phylogeny plotting with APE</a></li>
<li class="chapter" data-level="3.3.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#basic-phylogeny-plotting-with-ggtree"><i class="fa fa-check"></i><b>3.3.2</b> Basic phylogeny plotting with <code>ggtree</code></a></li>
<li class="chapter" data-level="3.3.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#back-to-ggtree"><i class="fa fa-check"></i><b>3.3.3</b> Back to <code>ggtree</code></a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#phylogeny-manipulation"><i class="fa fa-check"></i><b>3.4</b> Phylogeny manipulation</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#removing-polytomies"><i class="fa fa-check"></i><b>3.4.1</b> Removing polytomies</a></li>
<li class="chapter" data-level="3.4.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#rooting-your-phylogeny"><i class="fa fa-check"></i><b>3.4.2</b> Rooting your phylogeny</a></li>
<li class="chapter" data-level="3.4.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#manipulating-the-species-in-your-phlyogeny"><i class="fa fa-check"></i><b>3.4.3</b> Manipulating the species in your phlyogeny</a></li>
<li class="chapter" data-level="3.4.4" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#writing-phylogenies-to-file"><i class="fa fa-check"></i><b>3.4.4</b> Writing phylogenies to file</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#summary-2"><i class="fa fa-check"></i><b>3.5</b> Summary</a></li>
<li class="chapter" data-level="3.6" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#practical-exercise-1"><i class="fa fa-check"></i><b>3.6</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html"><i class="fa fa-check"></i><b>4</b> Preparing your tree and data for PCMs in R</a>
<ul>
<li class="chapter" data-level="" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#before-you-start-2"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="4.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#preparation"><i class="fa fa-check"></i><b>4.1</b> Preparation</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#load-the-required-packages"><i class="fa fa-check"></i><b>4.1.1</b> Load the required packages</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-and-checking-your-phylogeny-in-r"><i class="fa fa-check"></i><b>4.2</b> Reading and checking your phylogeny in R</a></li>
<li class="chapter" data-level="4.3" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>4.3</b> Reading the data into R</a></li>
<li class="chapter" data-level="4.4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-your-data-to-your-phylogeny"><i class="fa fa-check"></i><b>4.4</b> Matching your data to your phylogeny</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#species-names-with-spaces"><i class="fa fa-check"></i><b>4.4.1</b> Species names with spaces</a></li>
<li class="chapter" data-level="4.4.2" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#mismatches-between-species-in-your-data-and-phylogeny"><i class="fa fa-check"></i><b>4.4.2</b> Mismatches between species in your data and phylogeny</a></li>
<li class="chapter" data-level="4.4.3" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-the-tree-and-the-data"><i class="fa fa-check"></i><b>4.4.3</b> Matching the tree and the data</a></li>
<li class="chapter" data-level="4.4.4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#nas-in-your-data"><i class="fa fa-check"></i><b>4.4.4</b> NAs in your data</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#quick-template-code"><i class="fa fa-check"></i><b>4.5</b> Quick template code</a></li>
<li class="chapter" data-level="4.6" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#summary-3"><i class="fa fa-check"></i><b>4.6</b> Summary</a></li>
<li class="chapter" data-level="4.7" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#practical-exercise-2"><i class="fa fa-check"></i><b>4.7</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html"><i class="fa fa-check"></i><b>5</b> Phylogenetic Signal in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#before-you-start-3"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="5.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#preparation-1"><i class="fa fa-check"></i><b>5.1</b> Preparation</a></li>
<li class="chapter" data-level="5.2" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#estimating-phylogenetic-signal-for-continuous-variables"><i class="fa fa-check"></i><b>5.2</b> Estimating phylogenetic signal for continuous variables</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#pagels-lambda"><i class="fa fa-check"></i><b>5.2.1</b> Pagel’s <span class="math inline">\(\lambda\)</span></a></li>
<li class="chapter" data-level="5.2.2" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#blombergs-k-blomberg-et-al-2003"><i class="fa fa-check"></i><b>5.2.2</b> Blomberg’s <em>K</em> (Blomberg et al 2003)</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#estimating-phylogenetic-signal-for-non-continuous-variables"><i class="fa fa-check"></i><b>5.3</b> Estimating phylogenetic signal for non-continuous variables</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#d"><i class="fa fa-check"></i><b>5.3.1</b> <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#summary-4"><i class="fa fa-check"></i><b>5.4</b> Summary</a></li>
<li class="chapter" data-level="5.5" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#practical-exercises"><i class="fa fa-check"></i><b>5.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html"><i class="fa fa-check"></i><b>6</b> Phylogenetic Generalised Least Squares (PGLS) in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#before-you-start-4"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="6.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#preparation-2"><i class="fa fa-check"></i><b>6.1</b> Preparation</a></li>
<li class="chapter" data-level="6.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#phylogenetic-generalized-least-squares-models-pgls"><i class="fa fa-check"></i><b>6.2</b> Phylogenetic generalized least squares models (PGLS)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#fitting-pgls-models-in-r"><i class="fa fa-check"></i><b>6.2.1</b> Fitting PGLS models in R</a></li>
<li class="chapter" data-level="6.2.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#model-diagnostics-for-pgls-models"><i class="fa fa-check"></i><b>6.2.2</b> Model diagnostics for PGLS models</a></li>
<li class="chapter" data-level="6.2.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#likelihood-profiles-for-lambda-in-pgls-models"><i class="fa fa-check"></i><b>6.2.3</b> Likelihood profiles for <span class="math inline">\(\lambda\)</span> in PGLS models</a></li>
<li class="chapter" data-level="6.2.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#a-pgls-example-with-more-than-one-predictor-variable"><i class="fa fa-check"></i><b>6.2.4</b> A PGLS example with more than one predictor variable</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#common-issues-when-running-pgls-models-in-r"><i class="fa fa-check"></i><b>6.3</b> Common issues when running PGLS models in R</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#optim-errors"><i class="fa fa-check"></i><b>6.3.1</b> Optim errors</a></li>
<li class="chapter" data-level="6.3.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#big-problems-with-small-datasets"><i class="fa fa-check"></i><b>6.3.2</b> Big problems with small datasets</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#summary-5"><i class="fa fa-check"></i><b>6.4</b> Summary</a></li>
<li class="chapter" data-level="6.5" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#practical-exercises-1"><i class="fa fa-check"></i><b>6.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html"><i class="fa fa-check"></i><b>7</b> Phylogenetic generalised linear mixed models in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#before-you-start-5"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="7.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#why-fit-phylogenetic-generalised-linear-mixed-models"><i class="fa fa-check"></i><b>7.1</b> Why fit phylogenetic generalised linear mixed models?</a></li>
<li class="chapter" data-level="7.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#a-really-quick-reminder-of-how-bayesian-analyses-work-with-apologies-to-bayesian-scholars-for-the-simplification"><i class="fa fa-check"></i><b>7.2</b> A really quick reminder of how Bayesian analyses work, with apologies to Bayesian scholars for the simplification!</a></li>
<li class="chapter" data-level="7.3" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#preparation-3"><i class="fa fa-check"></i><b>7.3</b> Preparation</a></li>
<li class="chapter" data-level="7.4" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmcglmm-specific-preparation"><i class="fa fa-check"></i><b>7.4</b> MCMCglmm specific preparation</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#inverse-phylogenetic-variance-covariance-matrix"><i class="fa fa-check"></i><b>7.4.1</b> Inverse phylogenetic variance-covariance matrix</a></li>
<li class="chapter" data-level="7.4.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#priors"><i class="fa fa-check"></i><b>7.4.2</b> Priors</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#other-arguments-in-mcmcglmm"><i class="fa fa-check"></i><b>7.5</b> Other arguments in <code>MCMCglmm</code></a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#nitt"><i class="fa fa-check"></i><b>7.5.1</b> nitt</a></li>
<li class="chapter" data-level="7.5.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#burnin"><i class="fa fa-check"></i><b>7.5.2</b> burnin</a></li>
<li class="chapter" data-level="7.5.3" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#thin"><i class="fa fa-check"></i><b>7.5.3</b> thin</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#fitting-the-mcmcglmm-model"><i class="fa fa-check"></i><b>7.6</b> Fitting the MCMCglmm model</a></li>
<li class="chapter" data-level="7.7" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#did-the-model-converge"><i class="fa fa-check"></i><b>7.7</b> Did the model converge?</a></li>
<li class="chapter" data-level="7.8" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmc-outputs"><i class="fa fa-check"></i><b>7.8</b> MCMC outputs</a></li>
<li class="chapter" data-level="7.9" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#extracting-lambda-for-mcmcglmm-models"><i class="fa fa-check"></i><b>7.9</b> Extracting lambda for MCMCglmm models</a></li>
<li class="chapter" data-level="7.10" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#checking-the-effects-of-priors"><i class="fa fa-check"></i><b>7.10</b> Checking the effects of priors</a></li>
<li class="chapter" data-level="7.11" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#an-example-using-a-pglmm"><i class="fa fa-check"></i><b>7.11</b> An example using a PGLMM</a></li>
<li class="chapter" data-level="7.12" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#adding-extra-fixed-and-random-effects"><i class="fa fa-check"></i><b>7.12</b> Adding extra fixed and random effects</a></li>
<li class="chapter" data-level="7.13" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#summary-6"><i class="fa fa-check"></i><b>7.13</b> Summary</a></li>
<li class="chapter" data-level="7.14" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#practical-exercises-2"><i class="fa fa-check"></i><b>7.14</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html"><i class="fa fa-check"></i><b>8</b> Macroevolutionary models for continuous traits</a>
<ul>
<li class="chapter" data-level="" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#before-you-start-6"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="8.1" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#preparation-4"><i class="fa fa-check"></i><b>8.1</b> Preparation</a></li>
<li class="chapter" data-level="8.2" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#models-of-evolution-for-continuous-traits"><i class="fa fa-check"></i><b>8.2</b> Models of evolution for continuous traits</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#look-at-the-data"><i class="fa fa-check"></i><b>8.2.1</b> Look at the data</a></li>
<li class="chapter" data-level="8.2.2" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#fitting-the-brownian-motion-and-ou-models-of-evolution-using-fitcontinuous"><i class="fa fa-check"></i><b>8.2.2</b> Fitting the Brownian motion and OU models of evolution using <code>fitContinuous</code></a></li>
<li class="chapter" data-level="8.2.3" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#comparing-models-of-evolution-using-aic"><i class="fa fa-check"></i><b>8.2.3</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#summary-7"><i class="fa fa-check"></i><b>8.3</b> Summary</a></li>
<li class="chapter" data-level="8.4" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#practical-exercises-3"><i class="fa fa-check"></i><b>8.4</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html"><i class="fa fa-check"></i><b>9</b> Macroevolutionary models for discrete traits</a>
<ul>
<li class="chapter" data-level="" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#before-you-start-7"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="9.1" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#preparation-5"><i class="fa fa-check"></i><b>9.1</b> Preparation</a></li>
<li class="chapter" data-level="9.2" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#models-of-evolution-for-discrete-traits"><i class="fa fa-check"></i><b>9.2</b> Models of evolution for discrete traits</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#fitting-the-er-sym-and-ard-models-of-evolution-using-fitdiscrete"><i class="fa fa-check"></i><b>9.2.1</b> Fitting the ER, SYM and ARD models of evolution using <code>fitDiscrete</code></a></li>
<li class="chapter" data-level="9.2.2" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#comparing-models-of-evolution-using-aic-1"><i class="fa fa-check"></i><b>9.2.2</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#ancestral-state-estimations"><i class="fa fa-check"></i><b>9.3</b> Ancestral state estimations</a></li>
<li class="chapter" data-level="9.4" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#summary-8"><i class="fa fa-check"></i><b>9.4</b> Summary</a></li>
<li class="chapter" data-level="9.5" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#practical-exercises-4"><i class="fa fa-check"></i><b>9.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html"><i class="fa fa-check"></i><b>10</b> Beyond simple macroevolutionary models</a>
<ul>
<li class="chapter" data-level="" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#before-you-start-8"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="10.1" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#preparation-6"><i class="fa fa-check"></i><b>10.1</b> Preparation</a></li>
<li class="chapter" data-level="10.2" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#more-complex-models-of-trait-evolution"><i class="fa fa-check"></i><b>10.2</b> More complex models of trait evolution</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#preparing-the-tree-and-data-for-ouwie"><i class="fa fa-check"></i><b>10.2.1</b> Preparing the tree and data for OUwie</a></li>
<li class="chapter" data-level="10.2.2" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#running-models-using-ouwie"><i class="fa fa-check"></i><b>10.2.2</b> Running models using OUwie</a></li>
<li class="chapter" data-level="10.2.3" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#single-rate-bm-models"><i class="fa fa-check"></i><b>10.2.3</b> Single rate BM models</a></li>
<li class="chapter" data-level="10.2.4" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#single-optimum-ou-models"><i class="fa fa-check"></i><b>10.2.4</b> Single optimum OU models</a></li>
<li class="chapter" data-level="10.2.5" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-bm-bms-models"><i class="fa fa-check"></i><b>10.2.5</b> Multi-rate BM (BMS) models</a></li>
<li class="chapter" data-level="10.2.6" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-optima-ou-oum-models"><i class="fa fa-check"></i><b>10.2.6</b> Multi-optima OU (OUM) models</a></li>
<li class="chapter" data-level="10.2.7" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-multi-optima-ou-motion-models-oumv"><i class="fa fa-check"></i><b>10.2.7</b> Multi-rate multi-optima OU motion models (OUMV)</a></li>
<li class="chapter" data-level="10.2.8" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-alpha-multi-optima-ou-motion-models-ouma"><i class="fa fa-check"></i><b>10.2.8</b> Multi-alpha multi-optima OU motion models (OUMA)</a></li>
<li class="chapter" data-level="10.2.9" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-multi-optima-ou-motion-models-oumva"><i class="fa fa-check"></i><b>10.2.9</b> Multi-rate multi-optima OU motion models (OUMVA)</a></li>
<li class="chapter" data-level="10.2.10" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#comparing-models-of-evolution-using-aic-2"><i class="fa fa-check"></i><b>10.2.10</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#summary-9"><i class="fa fa-check"></i><b>10.3</b> Summary</a></li>
<li class="chapter" data-level="10.4" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#practical-exercises-5"><i class="fa fa-check"></i><b>10.4</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html"><i class="fa fa-check"></i><b>11</b> Diversification rate analyses</a>
<ul>
<li class="chapter" data-level="" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#before-you-start-9"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="11.1" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#preparation-7"><i class="fa fa-check"></i><b>11.1</b> Preparation</a></li>
<li class="chapter" data-level="11.2" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#lineage-through-time-plots-and-the-gamma-statistic"><i class="fa fa-check"></i><b>11.2</b> Lineage through time plots and the <span class="math inline">\(\gamma\)</span> statistic</a></li>
<li class="chapter" data-level="11.3" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#bamm-bayesian-analysis-of-macroevolutionary-mixtures"><i class="fa fa-check"></i><b>11.3</b> BAMM: Bayesian Analysis of Macroevolutionary Mixtures</a></li>
<li class="chapter" data-level="11.4" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#running-bamm"><i class="fa fa-check"></i><b>11.4</b> Running BAMM</a></li>
<li class="chapter" data-level="11.5" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#looking-at-the-output-of-bamm-using-bammtools-in-r"><i class="fa fa-check"></i><b>11.5</b> Looking at the output of BAMM using BAMMtools in R</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#assessing-mcmc-convergence"><i class="fa fa-check"></i><b>11.5.1</b> Assessing MCMC convergence</a></li>
<li class="chapter" data-level="11.5.2" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#analysis-of-rate-shifts-in-the-dragonfly-phylogeny"><i class="fa fa-check"></i><b>11.5.2</b> Analysis of rate shifts in the dragonfly phylogeny</a></li>
<li class="chapter" data-level="11.5.3" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#mean-phylorate-plot"><i class="fa fa-check"></i><b>11.5.3</b> Mean phylorate plot</a></li>
<li class="chapter" data-level="11.5.4" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#best-single-distinct-shift-configuration"><i class="fa fa-check"></i><b>11.5.4</b> Best single distinct shift configuration</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#credible-sets-of-shifts"><i class="fa fa-check"></i><b>11.6</b> Credible sets of shifts</a></li>
<li class="chapter" data-level="11.7" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#clade-specific-evolutionary-rates"><i class="fa fa-check"></i><b>11.7</b> Clade specific evolutionary rates</a></li>
<li class="chapter" data-level="11.8" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#rate-through-time-analysis"><i class="fa fa-check"></i><b>11.8</b> Rate-through-time analysis</a></li>
<li class="chapter" data-level="11.9" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#macroevolutionary-cohort-analysis"><i class="fa fa-check"></i><b>11.9</b> Macroevolutionary cohort analysis</a></li>
<li class="chapter" data-level="11.10" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#assumptions-and-issues-with-bamm"><i class="fa fa-check"></i><b>11.10</b> Assumptions and issues with BAMM</a></li>
<li class="chapter" data-level="11.11" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#summary-10"><i class="fa fa-check"></i><b>11.11</b> Summary</a></li>
<li class="chapter" data-level="11.12" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#practical-exercises-6"><i class="fa fa-check"></i><b>11.12</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html"><i class="fa fa-check"></i><b>12</b> Answers to practical exercises</a>
<ul>
<li class="chapter" data-level="12.1" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#need-to-know-r"><i class="fa fa-check"></i><b>12.1</b> Need to know R</a></li>
<li class="chapter" data-level="12.2" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#data-wrangling"><i class="fa fa-check"></i><b>12.2</b> Data wrangling</a></li>
<li class="chapter" data-level="12.3" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#phylogenies"><i class="fa fa-check"></i><b>12.3</b> Phylogenies</a></li>
<li class="chapter" data-level="12.4" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#preparation-8"><i class="fa fa-check"></i><b>12.4</b> Preparation</a></li>
<li class="chapter" data-level="12.5" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#phylogenetic-signal"><i class="fa fa-check"></i><b>12.5</b> Phylogenetic signal</a></li>
<li class="chapter" data-level="12.6" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#pgls"><i class="fa fa-check"></i><b>12.6</b> PGLS</a></li>
<li class="chapter" data-level="12.7" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#mcmcglmm"><i class="fa fa-check"></i><b>12.7</b> MCMCglmm</a></li>
<li class="chapter" data-level="12.8" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#models-of-evolution-with-continuous-traits"><i class="fa fa-check"></i><b>12.8</b> Models of evolution with continuous traits</a></li>
<li class="chapter" data-level="12.9" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#models-of-evolution-with-discrete-traits"><i class="fa fa-check"></i><b>12.9</b> Models of evolution with discrete traits</a></li>
<li class="chapter" data-level="12.10" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#beyond-simple-macroevolutionary-models-of-trait-evolution"><i class="fa fa-check"></i><b>12.10</b> Beyond simple macroevolutionary models of trait evolution</a></li>
<li class="chapter" data-level="12.11" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#diversification"><i class="fa fa-check"></i><b>12.11</b> Diversification</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html"><i class="fa fa-check"></i><b>13</b> Common mistakes using PCMs in R</a>
<ul>
<li class="chapter" data-level="13.1" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#good-practice-to-prevent-errors"><i class="fa fa-check"></i><b>13.1</b> Good practice to prevent errors</a></li>
<li class="chapter" data-level="13.2" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#common-errors"><i class="fa fa-check"></i><b>13.2</b> Common errors</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#standard-r-issues"><i class="fa fa-check"></i><b>13.2.1</b> Standard R issues</a></li>
<li class="chapter" data-level="13.2.2" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#issues-with-the-tree-or-data"><i class="fa fa-check"></i><b>13.2.2</b> Issues with the tree or data</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#what-to-do-if-you-get-an-error-message"><i class="fa fa-check"></i><b>13.3</b> What to do if you get an Error message</a></li>
<li class="chapter" data-level="13.4" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#what-to-do-if-you-get-a-warning-message"><i class="fa fa-check"></i><b>13.4</b> What to do if you get a Warning message</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="phylogenetic-generalised-linear-mixed-models-in-r" class="section level1 hasAnchor" number="7">
<h1><span class="header-section-number">Chapter 7</span> Phylogenetic generalised linear mixed models in R<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#phylogenetic-generalised-linear-mixed-models-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The aims of this exercise are to learn how to use R to perform Phylogenetic Generalised Linear Mixed models (PGLMM)in R using <code>MCMCglmm</code>.</p>
<div class="info">
<p>
We will be using the evolution of eye size in frogs as an example.
The data and modified tree come from <span class="citation"><span class="citation">K. N. Thomas et al. (<a href="#ref-thomas2020eye" role="doc-biblioref">2020</a>)</span></span>, and the original tree comes from
<span class="citation"><span class="citation">Feng et al. (<a href="#ref-feng2017phylogenomics" role="doc-biblioref">2017</a>)</span></span>. I’ve removed a few
species and a few variables to make things a bit more straightforward.
If you want to see the full results check out <span class="citation"><span class="citation">K. N. Thomas et al. (<a href="#ref-thomas2020eye" role="doc-biblioref">2020</a>)</span></span>!
</p>
</div>
<div id="before-you-start-5" class="section level2 unnumbered hasAnchor">
<h2>Before you start<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#before-you-start-5" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li>Open the <code>07-MCMCglmm.RProj</code> file in the <code>07-MCMCglmm</code> folder to open your R Project for this exercise.</li>
</ul>
<p>You will also need to install the following packages:</p>
<ul>
<li><code>tidyverse</code> - for reading, manipulating and plotting data</li>
<li><code>ape</code> - functions for reading, plotting and manipulating phylogenies</li>
<li><code>geiger</code> - to fit models of evolution</li>
<li><code>MCMCglmm</code> - to fit GLMMs</li>
<li><code>coda</code> - to assess convergence of MCMC runs</li>
</ul>
</div>
<div id="why-fit-phylogenetic-generalised-linear-mixed-models" class="section level2 hasAnchor" number="7.1">
<h2><span class="header-section-number">7.1</span> Why fit phylogenetic generalised linear mixed models?<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#why-fit-phylogenetic-generalised-linear-mixed-models" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Up this point we have been working with models where our diagnostic plots look pretty good, and where the residuals are roughly normally distributed. Unfortunately this is not the case for all models, especially those with particular kinds of response variables.</p>
<p>Where response variables are counts (e.g. number of individuals) we don’t expect normally distributed residuals because counts can only be whole numbers, and cannot go below zero. Instead count data are usually pretty well described by a Poisson distribution instead. Where response variables are binary (e.g. 0 or 1) or proportion data, we also don’t expect to see normally distributed residuals because again these variables cannot be negative, and in the case of binary data cannot be anything other than 0 or 1, and in the case of proportion data can only be between 0 and 1. Binary/proportion data are usually pretty well described by a binomial distribution instead. In your statistics classes you may have heard that the solution in these cases is to fit a <em>generalised</em> linear model, <code>glm</code> in R (rather than just a linear model; <code>lm</code> in R). We need to add this generalised linear modelling framework if we want to use count, or proportion, or binary response variables in our models.</p>
<p>Unfortunately, there is not an easy way to extend PGLS to do this, so we need another kind of model. As described in the <em>Primer</em>, there are a number of ways to incorporate phylogenetic information into statistical methods based on a variety of statistical models. We focus on the PGLS model because it’s quite easy to understand how it works, and uses standard models that we (should be) are already familiar with from basic statistics.</p>
<p>Another way we can incorporate phylogenetic information into a model is by using a mixed effects model. Mixed effects models have fixed effects, usually the variables we are interested in investigating, and random effects, often “nuisance” variables that we might need to control for in our analysis, or information about the hierarchical nature of the model set up. We can add information about the phlyogeny into the random effects componenet of a mixed effects model. Mixed effects models are very useful, but very hard to fully understand. See the <em>Primer</em> for more details.</p>
<p>Luckily, there is a package in R called <code>MCMCglmm</code> that performs <em>generalised linear mixed models</em> (GLMM), i.e. mixed effects models that are generalised to deal with response variables that result in non-normally distributed residuals. This package also allows us to add our phylogenetic information as a random effect in the model. Great!</p>
<p>Unfortunately, fitting these models is quite tricky, in part because of the <em>MCMC</em> part of the <code>MCMCglmm</code> package name. MCMC stands for Markov Chain Monte Carlo, and this is a Bayesian method which means it works somewhat differently to the models we’ve met before. I’ll try to explain as we go along, and see the <em>Primer</em> for more details.</p>
<div class="warning">
<p>
Bayesian phylogenetic generalised mixed models are very powerful
tools, but can be complicated to understand and difficult to use
properly. Before jumping in to these it is vital that you have a good
understanding of <strong>generalised linear models (GLMs)</strong>, and
<strong>generalised linear mixed models (GLMMs)</strong> including
<strong>how to fit and interpret the outputs of these models in
R</strong>. This is beyond the scope of this book. If you feel
uncomfortable with these we suggest you go back and learn these methods
first. You also need a basic understanding of <strong>Bayesian
statistics</strong>.
</p>
</div>
</div>
<div id="a-really-quick-reminder-of-how-bayesian-analyses-work-with-apologies-to-bayesian-scholars-for-the-simplification" class="section level2 hasAnchor" number="7.2">
<h2><span class="header-section-number">7.2</span> A really quick reminder of how Bayesian analyses work, with apologies to Bayesian scholars for the simplification!<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#a-really-quick-reminder-of-how-bayesian-analyses-work-with-apologies-to-bayesian-scholars-for-the-simplification" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Bayesian methods calculate the probability of the model given the data, <em>P(model|data)</em>, i.e. the reverse of Maximum Likelihood.</p>
<p>To fit Bayesian models we first assign a set of <em>priors</em>. These are our hypotheses about how we think the world works. They provide a starting point for the MCMC algorithm to search model space.</p>
<p>I like to think of Bayesian MCMC algorithms as a little robot friend wandering about. The robot starts where the priors suggest it should, then searches the virtual space of all possible model fits to find the parameters for your model (e.g. slopes and intercepts). The first few (hundreds, thousands, sometimes millions depending on the model!) tend to be poor attempts as the robot is still getting oriented, so we exclude these from the output as <em>burnin</em>. The robot keeps going until it’s tried as many models as you asked for (again this can be thousands to millions depending on the model). Because the robot will find very similar results close together due to the nature of the way it searches, we say that the results are autocorrelated. To deal with this, we only sample the results every so often, rather than sampling all of them. This is called <em>thinning</em> and we tell the robot in advance how we want to thin the sample (usually we ask for an answer every couple of thousand attempts).</p>
<p>Finally the robot gives us a distribution of results (often thousands or even millions), that we call the <em>posterior</em> distribution (or just the posterior). The posterior is a sample of all possible results across all of models fitted, in proportion to their posterior probability. This means there will be a range of results in the posterior, but those the robot encountered more often will appear more often, and those it encountered rarely will appear less often.</p>
<p>Before we look at the results, we need to check that the robot did a good job and <em>converged</em> on a sensible set of results, rather than wandering all over the place like it did during the burnin phase. We generally look at trace plots to assess this (see below) and also calculate <em>Effective Sample Size</em> or ESS. ESS of over 200 suggests the model has converged, and we can then go on to look at our results.</p>
<p>This is a simplification of course! But the take home is that to fit MCMCglmm models we need to consider priors, burnin, number of iterations, thinning, convergence and the posterior. In addition to the model, data and phylogeny that we’ve been considering in previous examples…</p>
<div class="info">
<p>
If you want to learn more about Bayesian statistics we highly
recommend <span class="citation"><span class="citation">McElreath (<a href="#ref-mcelreath2020statistical" role="doc-biblioref">2020</a>)</span></span> and
the accompanying online course.
</p>
</div>
<p>Now let’s try a practical example…</p>
</div>
<div id="preparation-3" class="section level2 hasAnchor" number="7.3">
<h2><span class="header-section-number">7.3</span> Preparation<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#preparation-3" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To begin we need to load the packages for this practical.</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb232-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the packages</span></span>
<span id="cb232-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb232-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb232-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb232-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb232-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb232-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb232-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb232-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MCMCglmm)</span>
<span id="cb232-6"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb232-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(coda)</span></code></pre></div>
<p>Next we need to prepare the tree and data for the analyses. In the <code>04-Preparation</code> exercise we read in our tree and data, checked them, and matched them so only species in both were retained. Please refer to that exercise for more details on how and why we do these things, or run through it now if you haven’t previously. Hopefully you’re pretty comfortable with this by now.</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb233-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the data</span></span>
<span id="cb233-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb233-2" aria-hidden="true" tabindex="-1"></a>frogdata <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/frog-eyes.csv&quot;</span>)</span>
<span id="cb233-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb233-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check everything loaded corrected</span></span>
<span id="cb233-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb233-4" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(frogdata)</span></code></pre></div>
<pre><code>## Rows: 215
## Columns: 11
## $ Binomial         &lt;chr&gt; &quot;Allophryne_ruthveni&quot;, &quot;Eupsophus_roseus&quot;, &quot;Alytes_obstetricans&quot;, &quot;…
## $ Family           &lt;chr&gt; &quot;Allophrynidae&quot;, &quot;Alsodidae&quot;, &quot;Alytidae&quot;, &quot;Alytidae&quot;, &quot;Aromobatidae…
## $ Genus            &lt;chr&gt; &quot;Allophryne&quot;, &quot;Eupsophus&quot;, &quot;Alytes&quot;, &quot;Discoglossus&quot;, &quot;Allobates&quot;, &quot;…
## $ tiplabel         &lt;chr&gt; &quot;Allophryne_ruthveni_Allophrynidae&quot;, &quot;Eupsophus_roseus_Alsodidae&quot;, …
## $ Adult_habitat    &lt;chr&gt; &quot;Scansorial&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwellin…
## $ Life_history     &lt;chr&gt; &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;…
## $ Sex_dichromatism &lt;chr&gt; &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent…
## $ SVL              &lt;dbl&gt; 23.76667, 38.37500, 37.46667, 62.64000, 25.60000, 27.03333, 38.4000…
## $ mass             &lt;dbl&gt; 1.0000000, 7.5000000, 6.6666667, 24.4000000, 2.1000000, 2.4666667, …
## $ rootmass         &lt;dbl&gt; 0.9917748, 1.9273451, 1.8690601, 2.8885465, 1.2549697, 1.3507613, 1…
## $ eyesize          &lt;dbl&gt; 3.200000, 5.362500, 6.366667, 7.550000, 4.116667, 3.783333, 4.75000…</code></pre>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb235-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the tree</span></span>
<span id="cb235-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb235-2" aria-hidden="true" tabindex="-1"></a>frogtree <span class="ot">&lt;-</span> <span class="fu">read.nexus</span>(<span class="st">&quot;data/frog-tree.nex&quot;</span>)</span>
<span id="cb235-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb235-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check it loaded correctly</span></span>
<span id="cb235-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb235-4" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(frogtree)</span></code></pre></div>
<pre><code>## List of 4
##  $ edge       : int [1:426, 1:2] 215 216 217 218 219 220 221 222 223 224 ...
##  $ edge.length: num [1:426] 0.166 0.114 0.102 0.4 0.133 ...
##  $ Nnode      : int 213
##  $ tip.label  : chr [1:214] &quot;Ascaphus_truei_Ascaphidae&quot; &quot;Leiopelma_hochstetteri_Leiopelmatidae&quot; &quot;Alytes_obstetricans_Alytidae&quot; &quot;Discoglossus_pictus_Alytidae&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;cladewise&quot;</code></pre>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb237-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is binary</span></span>
<span id="cb237-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb237-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE</span></span>
<span id="cb237-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb237-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.binary</span>(frogtree)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb239-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is rooted </span></span>
<span id="cb239-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb239-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE </span></span>
<span id="cb239-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb239-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(frogtree)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb241-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is ultrametric</span></span>
<span id="cb241-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb241-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE </span></span>
<span id="cb241-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb241-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.ultrametric</span>(frogtree)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb243-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the names match in the data and the tree</span></span>
<span id="cb243-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb243-2" aria-hidden="true" tabindex="-1"></a>check <span class="ot">&lt;-</span> <span class="fu">name.check</span>(<span class="at">phy =</span> frogtree, <span class="at">data =</span> frogdata, </span>
<span id="cb243-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb243-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">data.names =</span> frogdata<span class="sc">$</span>tiplabel)</span>
<span id="cb243-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb243-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at check</span></span>
<span id="cb243-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb243-5" aria-hidden="true" tabindex="-1"></a>check</span></code></pre></div>
<pre><code>## $tree_not_data
## [1] &quot;Incilius_nebulifer_Bufonidae&quot;            &quot;Leptobrachella_bidoupensis_Megophryidae&quot;
## [3] &quot;Microhyla_fissipes_Microhylidae&quot;         &quot;Microhyla_marmorata_Microhylidae&quot;       
## 
## $data_not_tree
## [1] &quot;Gastrophryne_carolinensis_Microhylidae&quot; &quot;Leptobrachella_dringi_Megophryidae&quot;    
## [3] &quot;Megophrys_gerti_Megophryidae&quot;           &quot;Microhyla_pulverata_Microhylidae&quot;      
## [5] &quot;Oreobates_quixensis_Strabomantidae&quot;</code></pre>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb245-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove species missing from the data</span></span>
<span id="cb245-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb245-2" aria-hidden="true" tabindex="-1"></a>mytree <span class="ot">&lt;-</span> <span class="fu">drop.tip</span>(frogtree, check<span class="sc">$</span>tree_not_data)</span>
<span id="cb245-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb245-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb245-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb245-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove species missing from the tree</span></span>
<span id="cb245-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb245-5" aria-hidden="true" tabindex="-1"></a>matches <span class="ot">&lt;-</span> <span class="fu">match</span>(frogdata<span class="sc">$</span>tiplabel, check<span class="sc">$</span>data_not_tree, <span class="at">nomatch =</span> <span class="dv">0</span>)</span>
<span id="cb245-6"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb245-6" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">subset</span>(frogdata, matches <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb245-7"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb245-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb245-8"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb245-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the tree summary</span></span>
<span id="cb245-9"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb245-9" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(mytree)</span></code></pre></div>
<pre><code>## List of 4
##  $ edge       : int [1:418, 1:2] 211 212 213 214 215 216 217 218 219 220 ...
##  $ edge.length: num [1:418] 0.166 0.114 0.102 0.4 0.133 ...
##  $ Nnode      : int 209
##  $ tip.label  : chr [1:210] &quot;Ascaphus_truei_Ascaphidae&quot; &quot;Leiopelma_hochstetteri_Leiopelmatidae&quot; &quot;Alytes_obstetricans_Alytidae&quot; &quot;Discoglossus_pictus_Alytidae&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;cladewise&quot;</code></pre>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb247-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the data</span></span>
<span id="cb247-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb247-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(mydata)</span></code></pre></div>
<pre><code>## Rows: 210
## Columns: 11
## $ Binomial         &lt;chr&gt; &quot;Allophryne_ruthveni&quot;, &quot;Eupsophus_roseus&quot;, &quot;Alytes_obstetricans&quot;, &quot;…
## $ Family           &lt;chr&gt; &quot;Allophrynidae&quot;, &quot;Alsodidae&quot;, &quot;Alytidae&quot;, &quot;Alytidae&quot;, &quot;Aromobatidae…
## $ Genus            &lt;chr&gt; &quot;Allophryne&quot;, &quot;Eupsophus&quot;, &quot;Alytes&quot;, &quot;Discoglossus&quot;, &quot;Allobates&quot;, &quot;…
## $ tiplabel         &lt;chr&gt; &quot;Allophryne_ruthveni_Allophrynidae&quot;, &quot;Eupsophus_roseus_Alsodidae&quot;, …
## $ Adult_habitat    &lt;chr&gt; &quot;Scansorial&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwellin…
## $ Life_history     &lt;chr&gt; &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;…
## $ Sex_dichromatism &lt;chr&gt; &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent…
## $ SVL              &lt;dbl&gt; 23.76667, 38.37500, 37.46667, 62.64000, 25.60000, 27.03333, 38.4000…
## $ mass             &lt;dbl&gt; 1.0000000, 7.5000000, 6.6666667, 24.4000000, 2.1000000, 2.4666667, …
## $ rootmass         &lt;dbl&gt; 0.9917748, 1.9273451, 1.8690601, 2.8885465, 1.2549697, 1.3507613, 1…
## $ eyesize          &lt;dbl&gt; 3.200000, 5.362500, 6.366667, 7.550000, 4.116667, 3.783333, 4.75000…</code></pre>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb249-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to a dataframe</span></span>
<span id="cb249-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb249-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(mydata)</span>
<span id="cb249-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb249-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check this is now a data frame</span></span>
<span id="cb249-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb249-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mydata)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
</div>
<div id="mcmcglmm-specific-preparation" class="section level2 hasAnchor" number="7.4">
<h2><span class="header-section-number">7.4</span> MCMCglmm specific preparation<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmcglmm-specific-preparation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="inverse-phylogenetic-variance-covariance-matrix" class="section level3 hasAnchor" number="7.4.1">
<h3><span class="header-section-number">7.4.1</span> Inverse phylogenetic variance-covariance matrix<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#inverse-phylogenetic-variance-covariance-matrix" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For <code>MCMCglmm</code> we need the inverse of the phylogenetic variance-covariance matrix.</p>
<p>To do this, we first need the tree to have no zero-length branches. In <code>04-Preparation</code> you may recall that we deal with polytomies by replacing them with zero length branches using the <code>ape</code> function <code>multi2di</code>. To remove the zero length branches we are going to use the opposite function: <code>di2multi</code>. This deletes all zero length branches and collapses them back into polytomies.</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb251-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove zero length branches and replace with polytomies</span></span>
<span id="cb251-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb251-2" aria-hidden="true" tabindex="-1"></a>mytree2 <span class="ot">&lt;-</span> <span class="fu">di2multi</span>(mytree)</span></code></pre></div>
<p>Although not an issue here, <code>inverseA</code> (see below) can break if your tree has node labels. We can remove these using:</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb252-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove node labels </span></span>
<span id="cb252-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb252-2" aria-hidden="true" tabindex="-1"></a>mytree2<span class="sc">$</span>node.label <span class="ot">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
<p>Now we can get the inverse of the phylogenetic variance-covariance matrix using the function <code>inverseA</code>.</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb253-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the inverse vcv matrix for the phylogeny</span></span>
<span id="cb253-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb253-2" aria-hidden="true" tabindex="-1"></a>inv.phylo <span class="ot">&lt;-</span> <span class="fu">inverseA</span>(mytree2, <span class="at">nodes =</span> <span class="st">&quot;TIPS&quot;</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>Ainv</span></code></pre></div>
<p><code>nodes = "TIPS"</code> means that we are only including the tips in the calculation. We can use <code>nodes = "ALL"</code> to include ancestral nodes in the calculation. <code>scale = TRUE</code> means the tree is scaled so that total branch length from root to tips, is equal to one. This makes the model a bit easier to fit, but only works with ultrametric trees. For non-ultrametric trees you need to set this as <code>FALSE</code>.</p>
<p>Using <code>nodes = "TIPS"</code> can mean it takes longer for the model to run. Very large phylogenies usually require <code>nodes = "ALL"</code> or <code>MCMCglmm</code> won’t even run.
However, <code>nodes = "ALL"</code> can result in higher auto-correlation (see below).</p>
</div>
<div id="priors" class="section level3 hasAnchor" number="7.4.2">
<h3><span class="header-section-number">7.4.2</span> Priors<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#priors" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>MCMCglmm</code> requires us to set prior distributions for the fixed and random effects. For the random effects this is actually a bit trickier than many help tutorials might suggest. If you’re fitting any model with any degree of complexity you’ll need to read up on this to get the most appropriate priors (e.g. see <a href="http://www.wildanimalmodels.org/tiki-download_wiki_attachment.php?attId=24" class="uri">http://www.wildanimalmodels.org/tiki-download_wiki_attachment.php?attId=24</a> for some examples using non-phylogenetic models).</p>
<p>The default prior for the fixed effects is a Normal distribution with a mean (<code>mu</code>) of 0 and variance (<code>V</code>) of 1^10. This is known as a diffuse prior, as the variance is so large the prior is basically saying that the fixed effects can be pretty much any number, positive or negative. Generally this is fine so we don’t tend to change this from the default.</p>
<p>For the random effects we need to define a set of priors for the variance components of each random effect. <code>MCMCglmm</code> uses something called an <em>inverse Wishart</em> distribution for the random effects prior, and this is described by two parameters: the variance (<code>V</code>) and degree of belief (<code>nu</code>). This distribution tends to be right-skewed when <code>nu</code> is small.</p>
<div class="figure">
<img src="images/invwishart.png" alt="" />
<p class="caption">The inverse Wishart distribution</p>
</div>
<p>Often in examples you’ll see people using priors where <code>V = 1</code> and <code>nu = 0.02</code>. This prior specification is used a lot because it was believed to be relatively uninformative (it is also equivalent to an inverse-gamma prior with shape and scale of 0.001, for anyone more familiar with that distribution). It is generally is fairly uninformative, but can behave inappropriately when the posterior distribution for the variances has support close to zero. In these cases it’s worth looking at the parameter expanded priors section of the <a href="https://cran.r-project.org/web/packages/MCMCglmm/vignettes/CourseNotes.pdf"><code>MCMCglmm</code> course notes</a> for some solutions.</p>
<p>Using priors where <code>V = 1</code> and <code>nu = 0.02</code> tells R that it shouldn’t pay too much attention to our prior of <code>V = 1</code>, because our degree of belief is very low (<code>nu = 0.02</code>). This is why we say it is a fairly uniformative prior. Why do we want priors to be uniformative? We don’t always. If we have a pretty good idea of what answer we should get through theory, we might use more informative priors. However, in the analyses we do in most of biology, we don’t have good theory to back them up, so using an uniformative prior means we’re allowing the model to find the best answer without biasing the possible choices.</p>
<p>To set up priors for MCMCglmm we use a list, as follows (<code>G</code> is for the random effect(s), and <code>R</code> is for the residual variance):</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb254-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up priors for MCMCglmm</span></span>
<span id="cb254-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb254-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Inverse Wishart with V = 1 and nu = 0.02</span></span>
<span id="cb254-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb254-3" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. fairly uninformative priors</span></span>
<span id="cb254-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb254-4" aria-hidden="true" tabindex="-1"></a>prior <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">G =</span> <span class="fu">list</span>(<span class="at">G1 =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">1</span>, <span class="at">nu =</span> <span class="fl">0.02</span>)),</span>
<span id="cb254-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb254-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">R =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">1</span>, <span class="at">nu =</span> <span class="fl">0.02</span>))</span></code></pre></div>
</div>
</div>
<div id="other-arguments-in-mcmcglmm" class="section level2 hasAnchor" number="7.5">
<h2><span class="header-section-number">7.5</span> Other arguments in <code>MCMCglmm</code><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#other-arguments-in-mcmcglmm" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We also need to decide on a couple of other things before using <code>MCMCglmm</code>.</p>
<div id="nitt" class="section level3 hasAnchor" number="7.5.1">
<h3><span class="header-section-number">7.5.1</span> nitt<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#nitt" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>How many iterations of the model do we want to run? It’s standard to run quite a lot, sometimes in the millions, meaning it can take a really long time to run these models. Note that you don’t need to run millions of iterations if your model converges fairly quickly. Often a good plan is to choose a number of iterations, run the model, check for convergence (see below), then increase the number of iterations if needed. Bayesian analyses often involve a lot of tinkering with parameters, and this can be useful if you have lots of models to run and don’t want to edit for them unnecessarily.</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb255-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set number of iterations</span></span>
<span id="cb255-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb255-2" aria-hidden="true" tabindex="-1"></a>nitt <span class="ot">&lt;-</span> <span class="dv">1000000</span></span></code></pre></div>
</div>
<div id="burnin" class="section level3 hasAnchor" number="7.5.2">
<h3><span class="header-section-number">7.5.2</span> burnin<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#burnin" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The variables <code>nitt</code>, <code>burnin</code> and <code>thin</code> are used to calibrate the MCMCM algorithm. If we set <code>burnin = 1000</code>, <code>nitt = 1,000,000</code> and <code>thin = 500</code>, it will iterate for 1000 iterations (burnin) before recording samples (to ensure convergence), and then iterate 1,000,000 times. The parameter <code>thin</code> helps us to save memory by saving only every 500th value in this example and thus, dropping highly auto-correlated values.</p>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb256-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set burnin</span></span>
<span id="cb256-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb256-2" aria-hidden="true" tabindex="-1"></a>burnin <span class="ot">&lt;-</span> <span class="dv">1000</span></span></code></pre></div>
</div>
<div id="thin" class="section level3 hasAnchor" number="7.5.3">
<h3><span class="header-section-number">7.5.3</span> thin<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#thin" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb257-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sampling interval</span></span>
<span id="cb257-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb257-2" aria-hidden="true" tabindex="-1"></a>thin <span class="ot">&lt;-</span> <span class="dv">500</span></span></code></pre></div>
</div>
</div>
<div id="fitting-the-mcmcglmm-model" class="section level2 hasAnchor" number="7.6">
<h2><span class="header-section-number">7.6</span> Fitting the MCMCglmm model<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#fitting-the-mcmcglmm-model" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>OK now let’s fit a simple model that we’ve looked at before, eye size as a function of body size…</p>
<p>To tell <code>MCMCglmm</code> that we want to fit a model with a phylogeny we use <code>random = ~ tiplabel</code>. <code>tiplabel</code> is the name of the column in our dataset that matches the tip labels of the phylogeny. We use <code>ginverse = list(tiplabel = inv.phylo)</code> to tell it to use <code>inv.phylo</code>, the inverse phylogenetic variance-covariance matrix we made above.</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit MCMCglmm model</span></span>
<span id="cb258-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-2" aria-hidden="true" tabindex="-1"></a>model_mcmcglmm <span class="ot">&lt;-</span> <span class="fu">MCMCglmm</span>(<span class="fu">log</span>(eyesize) <span class="sc">~</span> <span class="fu">log</span>(mass), </span>
<span id="cb258-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> mydata, </span>
<span id="cb258-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">random =</span> <span class="sc">~</span> tiplabel,</span>
<span id="cb258-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ginverse =</span> <span class="fu">list</span>(<span class="at">tiplabel =</span> inv.phylo), </span>
<span id="cb258-6"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">prior =</span> prior,</span>
<span id="cb258-7"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">nitt =</span> nitt, <span class="at">thin =</span> thin, <span class="at">burnin =</span> burnin,</span>
<span id="cb258-8"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb258-9"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb258-10"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Save the model</span></span>
<span id="cb258-11"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb258-11" aria-hidden="true" tabindex="-1"></a><span class="co">#saveRDS(model_mcmcglmm, file = &quot;data/model_mcmcglmm_output1.rda&quot;)</span></span></code></pre></div>
<p>This will take a long time to run, dependent on your computer power. You’ll know it is working because you’ll see text appearing, e.g. <code>MCMC iteration = 1000</code>, every time the MCMC takes a sample for the posterior. You can turn this off with <code>verbose = FALSE</code> if it gets annoying. In fact this takes so long to run (at least an hour on my machine), that to save you some time here I’ve run the model myself and saved it using <code>saveRDS</code>. <strong>You do not need to run the code above if you don’t have an hour to spare!</strong>. We can then read it back in using:</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb259-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the saved model output run in advance to save time</span></span>
<span id="cb259-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb259-2" aria-hidden="true" tabindex="-1"></a>model_mcmcglmm <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/model_mcmcglmm_output1.rda&quot;</span>)</span></code></pre></div>
</div>
<div id="did-the-model-converge" class="section level2 hasAnchor" number="7.7">
<h2><span class="header-section-number">7.7</span> Did the model converge?<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#did-the-model-converge" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Before we look at the outputs we need to know if the model converged. If it didn’t you’ll need to fiddle with your parameters and run it again (usually the first thing to try is increasing <code>nitt</code> - this also means you might want to increase <code>burnin</code> and <code>thin</code> too).</p>
<p>We can check for convergence in two ways. First we can look at the model diagnostics using <code>plot</code> as we have done for other models in earlier exercises.</p>
<p>We can look at the fixed effects first, then the random effects:</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb260-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot model diagnostics for MCMCglmm</span></span>
<span id="cb260-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb260-2" aria-hidden="true" tabindex="-1"></a><span class="co"># For fixed effects</span></span>
<span id="cb260-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb260-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(model_mcmcglmm<span class="sc">$</span>Sol)  </span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-192-1.png" width="672" /></p>
<p>The plots on the left show a time series of the values of samples of the posterior distribution. The plots on the right show the same data as a distribution. We are looking for a “furry caterpillar” on the left hand plots, suggesting that the values are fluctuating about a largely unchanging mean value of each parameter. On the right hand plots we hope to see a distribution with one clear peak. The rows show the different parameters of the model from the fixed effects component, i.e. the intercept, the slope (labelled as log(mass)).</p>
<p>If it doesn’t look like the caterpillar is converging near the start of the plots that suggests you might need to increase the burnin. If there’s still a lot of fluctuation up and down you might need to increase burnin and the number of iterations. If the autocorrelation is high (see below) you might need to increase the thinning interval.</p>
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb261-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot model diagnostics for MCMCglmm</span></span>
<span id="cb261-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb261-2" aria-hidden="true" tabindex="-1"></a><span class="co"># For random effects</span></span>
<span id="cb261-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb261-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(model_mcmcglmm<span class="sc">$</span>VCV)  </span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-193-1.png" width="672" />
We interpret these in the same way as above except the first row is the random effect that contains the information from the phylogeny, and the second row is the residual variance. This is called units in <code>MCMCglmm</code>.</p>
<div class="info">
<p>
<strong>Mean centring and scaling your variables to unit
variance</strong>. Often people advise that you rescale numeric
variables so they are mean centred and scaled to unit variance. This can
be really helpful if you’ve got models with multiple numeric variables
that are quite different in scale (i.e. one ranges from 1 to 3 and one
from 100 to 1000). If your models are not converging this may also help.
To do this in R using functions from the <code>tidyverse</code>
package:
</p>
<p>
<code>mydata &lt;- mydata %&gt;% mutate(variable1_Z = scale(log(mydata<span class="math inline">\(variable1))) %&amp;gt;% mutate(variable2_Z = scale(abs(mydata\)</span>variable2)))</code>
</p>
<p>
Just remember to reverse the transformation when trying to interpret
any parameter estimates on the scale of the original variable.
</p>
</div>
<p>These plots look fine to me! :). Remember because of the nature of MCMC we won’t have the exact same plots here, don’t worry!</p>
<p>We can also calculate the effective sample size (ESS) as follows.</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb262-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate ESS</span></span>
<span id="cb262-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb262-2" aria-hidden="true" tabindex="-1"></a><span class="fu">effectiveSize</span>(model_mcmcglmm<span class="sc">$</span>Sol[, <span class="dv">1</span><span class="sc">:</span>model_mcmcglmm<span class="sc">$</span>Fixed<span class="sc">$</span>nfl, </span>
<span id="cb262-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb262-3" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">drop =</span> <span class="cn">FALSE</span>])[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## [1] 1998</code></pre>
<p>This is above 200 so the model has converged.</p>
<p>We should also check the validity of the posterior by looking at levels of autocorrelation…</p>
<div class="sourceCode" id="cb264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb264-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb264-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look for autocorrelation</span></span>
<span id="cb264-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb264-2" aria-hidden="true" tabindex="-1"></a><span class="fu">autocorr</span>(model_mcmcglmm<span class="sc">$</span>VCV)</span></code></pre></div>
<pre><code>## , , tiplabel
## 
##              tiplabel        units
## Lag 0      1.00000000 -0.526793220
## Lag 500   -0.01961702  0.001163258
## Lag 2500   0.02760427 -0.013775871
## Lag 5000  -0.02899116  0.006129946
## Lag 25000 -0.01119349 -0.006134928
## 
## , , units
## 
##                tiplabel        units
## Lag 0     -0.5267932199  1.000000000
## Lag 500   -0.0006145319 -0.003037141
## Lag 2500  -0.0187737699 -0.016918298
## Lag 5000   0.0064124393 -0.015999863
## Lag 25000 -0.0173105728  0.015958313</code></pre>
<p>Ideally, all samples of the posterior distribution should be independent,
and the autocorrelation for all lag values greater than zero should be near zero. These values look OK - they’re mostly below 0.03 which is pretty close to zero - but we could try and improve them by increasing the thinning interval and the number of iterations if we needed to.</p>
</div>
<div id="mcmc-outputs" class="section level2 hasAnchor" number="7.8">
<h2><span class="header-section-number">7.8</span> MCMC outputs<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmc-outputs" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now we can finally look at the results using <code>summary</code>.</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb266-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at summary of fixed effect results</span></span>
<span id="cb266-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb266-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model_mcmcglmm)<span class="sc">$</span>solutions</span></code></pre></div>
<pre><code>##             post.mean  l-95% CI  u-95% CI eff.samp        pMCMC
## (Intercept) 1.0375044 0.7484659 1.2955290 1998.000 0.0005005005
## log(mass)   0.2768096 0.2581998 0.2955905 3277.781 0.0005005005</code></pre>
<p>This gives us a similar output to the other models we’ve used previously, except the numbers come from the posterior distribution. We get the mean value of the intercept from the posterior, its 95% credible interval (the Bayesian equivalent of confidence intervals), its effective sample size, and an associated p value testing whether the value is signficantly different from zero. The second row shows the same information for the slope (log(mass)).</p>
<p>Here neither of the posterior distributions for the intercept or slope overlaps zero, so we can consider them both statistically supported. In a paper/report/thesis we’d report these parameter estimates and their credible intervals. I’d rarely report the MCMCp values. p values don’t really belong in a Bayesian world.</p>
<p>Remember we also need to look at results for the random effect variances. Generally the mode of the posterior is a better thing to look at here than the mean of the posterior, as the mode tells us where the peak of the posterior is. We can look at these using <code>posterior.mode</code>. Recall that <code>units</code> is what <code>MCMCglmm</code> calls the residual variance.</p>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb268-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at G-structure and R-structure posterior modes</span></span>
<span id="cb268-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb268-2" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior.mode</span>(model_mcmcglmm<span class="sc">$</span>VCV)</span></code></pre></div>
<pre><code>##   tiplabel      units 
## 0.11949510 0.00491087</code></pre>
<p>We also need the 95% credible intervals for these components:</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb270-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at G-structure and R-structure 95% credible intervals</span></span>
<span id="cb270-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb270-2" aria-hidden="true" tabindex="-1"></a><span class="fu">HPDinterval</span>(model_mcmcglmm<span class="sc">$</span>VCV)</span></code></pre></div>
<pre><code>##                lower      upper
## tiplabel 0.095286046 0.16795035
## units    0.002140634 0.01023753
## attr(,&quot;Probability&quot;)
## [1] 0.9499499</code></pre>
<p>While testing the significance of fixed effects by evaluating whether or not their posterior distributions overlap zero is simple and valid, this approach does not work for variance components. Variance components should be positive so even when a random effect is not meaningful, the posterior distribution will never overlap zero.</p>
<p>If we really care about whether a random effect is “significant”, we can fit a model <em>without</em> the random effect, and a model <em>with</em> the random effect, and then compare them using DIC (smaller DIC = better model). To get the DIC of the model we just do:</p>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb272-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract DIC</span></span>
<span id="cb272-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb272-2" aria-hidden="true" tabindex="-1"></a>model_mcmcglmm<span class="sc">$</span>DIC</span></code></pre></div>
<pre><code>## [1] -354.1069</code></pre>
<p>If the model with the random effect has the lowest DIC, we can say that inclusion of the term statistically justifiable. DIC has a lot of issues however, and this won’t be appropriate in all cases. Luckily in the case of our phylogenetic random effect we already have a very good reason for including it in the model, whether or not it is “significant”. We can also look at <span class="math inline">\(\lambda\)</span> to assess the amount of phylogenetic signal in the model. This should give us a good idea of occasions where adding the phylogenetic random effect is not appropriate.</p>
</div>
<div id="extracting-lambda-for-mcmcglmm-models" class="section level2 hasAnchor" number="7.9">
<h2><span class="header-section-number">7.9</span> Extracting lambda for MCMCglmm models<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#extracting-lambda-for-mcmcglmm-models" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We can estimate the posterior probability of the phylogenetic signal <span class="math inline">\(\lambda\)</span> as follows.</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb274-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get lambda</span></span>
<span id="cb274-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb274-2" aria-hidden="true" tabindex="-1"></a>lambda <span class="ot">&lt;-</span> model_mcmcglmm<span class="sc">$</span>VCV[,<span class="st">&#39;tiplabel&#39;</span>]<span class="sc">/</span></span>
<span id="cb274-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb274-3" aria-hidden="true" tabindex="-1"></a>          (model_mcmcglmm<span class="sc">$</span>VCV[,<span class="st">&#39;tiplabel&#39;</span>] <span class="sc">+</span> model_mcmcglmm<span class="sc">$</span>VCV[,<span class="st">&#39;units&#39;</span>])</span></code></pre></div>
<p>We can then get the mean and mode (for Bayesian analyses the mode is often a better measure of central tendency than the mean as it tells us where the peak of the posterior distribution is) of the posterior distribution for <span class="math inline">\(\lambda\)</span>, along with its 95% credible interval:</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb275-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Posterior mean of lambda</span></span>
<span id="cb275-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb275-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(lambda)</span></code></pre></div>
<pre><code>## [1] 0.9551432</code></pre>
<div class="sourceCode" id="cb277"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb277-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb277-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Posterior mode of lambda</span></span>
<span id="cb277-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb277-2" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior.mode</span>(lambda)</span></code></pre></div>
<pre><code>##      var1 
## 0.9658262</code></pre>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb279-1" aria-hidden="true" tabindex="-1"></a><span class="co"># HPD interval</span></span>
<span id="cb279-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb279-2" aria-hidden="true" tabindex="-1"></a><span class="fu">HPDinterval</span>(lambda)</span></code></pre></div>
<pre><code>##          lower     upper
## var1 0.9146033 0.9884231
## attr(,&quot;Probability&quot;)
## [1] 0.9499499</code></pre>
<p>We could also quickly plot the whole posterior, with the mode and 95% credible intervals.</p>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb281-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the posterior</span></span>
<span id="cb281-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb281-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(lambda), <span class="fu">aes</span>(<span class="at">x =</span> lambda))<span class="sc">+</span></span>
<span id="cb281-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb281-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb281-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb281-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">posterior.mode</span>(lambda), <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb281-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb281-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">HPDinterval</span>(lambda)[[<span class="dv">1</span>]], </span>
<span id="cb281-6"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb281-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb281-7"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb281-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">HPDinterval</span>(lambda)[[<span class="dv">2</span>]], </span>
<span id="cb281-8"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb281-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="at">colour =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb281-9"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb281-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-203-1.png" width="672" /></p>
<p>Note that this isn’t precisely <span class="math inline">\(\lambda\)</span> as formulated by Pagel. It’s the proportion of the variance components (G + R) compromised by the phylogenetic variance component (G). But we can interpret it in the same way, i.e. 0 = no phylogenetic signal, 1 = expectation under Brownian motion.</p>
<div class="info">
<p>
The MCMCglmm results are very close to what we got from PGLS where
intercept = 1.03, slope = 0.277, and lambda was 0.976 (95% CI: 0.924 -
1).
</p>
</div>
</div>
<div id="checking-the-effects-of-priors" class="section level2 hasAnchor" number="7.10">
<h2><span class="header-section-number">7.10</span> Checking the effects of priors<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#checking-the-effects-of-priors" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>You shouldn’t stop here! It’s important to repeat the analysis with different priors to see if that would influence your conclusions. Generally I’d increase the value for <code>nu</code>, maybe to 1 and see what happens.</p>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb282-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set up new priors for MCMCglmm</span></span>
<span id="cb282-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Inverse Wishart with V = 1 and nu = 1</span></span>
<span id="cb282-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-3" aria-hidden="true" tabindex="-1"></a>prior <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">G =</span> <span class="fu">list</span>(<span class="at">G1 =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">1</span>, <span class="at">nu =</span> <span class="dv">1</span>)),</span>
<span id="cb282-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">R =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">1</span>, <span class="at">nu =</span> <span class="dv">1</span>))</span>
<span id="cb282-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb282-6"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the model</span></span>
<span id="cb282-7"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-7" aria-hidden="true" tabindex="-1"></a>model_mcmcglmm2 <span class="ot">&lt;-</span> <span class="fu">MCMCglmm</span>(<span class="fu">log</span>(eyesize) <span class="sc">~</span> <span class="fu">log</span>(mass), </span>
<span id="cb282-8"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> mydata, </span>
<span id="cb282-9"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">random =</span> <span class="sc">~</span> tiplabel,</span>
<span id="cb282-10"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ginverse =</span> <span class="fu">list</span>(<span class="at">tiplabel =</span> inv.phylo), </span>
<span id="cb282-11"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">prior =</span> prior,</span>
<span id="cb282-12"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-12" aria-hidden="true" tabindex="-1"></a>                   <span class="at">nitt =</span> nitt, <span class="at">thin =</span> thin, <span class="at">burnin =</span> burnin,</span>
<span id="cb282-13"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-13" aria-hidden="true" tabindex="-1"></a>                   <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb282-14"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb282-15"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Save the model</span></span>
<span id="cb282-16"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb282-16" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(model_mcmcglmm2, <span class="at">file =</span> <span class="st">&quot;data/model_mcmcglmm_output2.rda&quot;</span>)</span></code></pre></div>
<p>Again this takes ages to run (at least an hour on my machine), so to save you some time here I’ve run the model myself and saved it using <code>saveRDS</code>. <strong>You do not need to run the code above if you don’t have an hour to spare!</strong>. We can then read it back in using:</p>
<div class="sourceCode" id="cb283"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb283-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the saved model output run in advance to save time</span></span>
<span id="cb283-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb283-2" aria-hidden="true" tabindex="-1"></a>model_mcmcglmm2 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/model_mcmcglmm_output2.rda&quot;</span>)</span></code></pre></div>
<p>Let’s look at the results:</p>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb284-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fixed effects results</span></span>
<span id="cb284-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb284-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model_mcmcglmm2)<span class="sc">$</span>solutions</span></code></pre></div>
<pre><code>##             post.mean  l-95% CI  u-95% CI eff.samp        pMCMC
## (Intercept) 1.0317751 0.7789332 1.2671116 2097.988 0.0005005005
## log(mass)   0.2733169 0.2518094 0.2962124 1998.000 0.0005005005</code></pre>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb286-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at G-structure and R-structure 95% credible intervals</span></span>
<span id="cb286-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb286-2" aria-hidden="true" tabindex="-1"></a><span class="fu">posterior.mode</span>(model_mcmcglmm2<span class="sc">$</span>VCV)</span></code></pre></div>
<pre><code>##   tiplabel      units 
## 0.09779188 0.02390362</code></pre>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb288-1" aria-hidden="true" tabindex="-1"></a><span class="fu">HPDinterval</span>(model_mcmcglmm2<span class="sc">$</span>VCV)</span></code></pre></div>
<pre><code>##               lower      upper
## tiplabel 0.06703434 0.12782712
## units    0.01708881 0.03047269
## attr(,&quot;Probability&quot;)
## [1] 0.9499499</code></pre>
<p>There’s no major change. Phew! We can therefore say that the difference in the priors has little effect on the outcome of the analysis (this is usual for an analysis where lots of data are available relative to the complexity of the model, but it is important to check!).</p>
<p>Hopefully you now see why these models are complicated and should be considered very carefully before using them. You should never skip the checking and tinkering steps.</p>
</div>
<div id="an-example-using-a-pglmm" class="section level2 hasAnchor" number="7.11">
<h2><span class="header-section-number">7.11</span> An example using a PGLMM<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#an-example-using-a-pglmm" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="warning">
<p>
<strong>A reminder</strong>. Before jumping in to these models it is
vital that you have a good understanding of <strong>generalised linear
models (GLMs)</strong>, and <strong>generalised linear mixed models
(GLMMs)</strong> including <strong>how to fit and interpret the outputs
of these models in R</strong>. This is beyond the scope of this book. If
you feel uncomfortable with these we suggest you go back and learn these
methods first.
</p>
</div>
<p>The example above is technically only a LMM (linear mixed model) as we used a model with Gaussian or Normal errors. But what if our response is a count? Or a proportion? We can then use the GLMM architecture of <code>MCMCglmm</code> to help.</p>
<p>For Poisson models (used generally when the response variable is a count) we actually only need to change one thing, we need to add <code>family = "poisson"</code> to tell <code>MCMCglmm</code> to fit this model. This is the same as in standard <code>glm</code> analyses in R. Make sure you understand GLMs before attempting PGLMMs!</p>
<p>For a fake example here, if we had a variable called <code>number_of_eggs</code> that contained count data, we could fit a model like this:</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb290-1" aria-hidden="true" tabindex="-1"></a>model_eggs <span class="ot">&lt;-</span> <span class="fu">MCMCglmm</span>(number_of_eggs <span class="sc">~</span> <span class="fu">log</span>(mass), </span>
<span id="cb290-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb290-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> mydata, <span class="at">family =</span> <span class="st">&quot;poisson&quot;</span>,</span>
<span id="cb290-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb290-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">random =</span> <span class="sc">~</span> Species,</span>
<span id="cb290-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb290-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ginverse =</span> <span class="fu">list</span>(<span class="at">Species =</span> inv.phylo), </span>
<span id="cb290-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb290-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">prior =</span> prior,</span>
<span id="cb290-6"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb290-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">nitt =</span> nitt, <span class="at">thin =</span> thin, <span class="at">burnin =</span> burnin,</span>
<span id="cb290-7"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb290-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Other than these changes, everything runs as above including the priors and all the model diagnostics and checking.</p>
<p>If we had proportion data instead, for example the number of individuals with blue eyes and the number with red eyes, we could use a binomial model. Again this works just like in a basic <code>glm</code> model. Again, make sure you understand GLMs before attempting PGLMMs!</p>
<p>We tell R we want to use a binomial error structure using <code>family = "binomial"</code>, and we include successes (number of frogs with red eyes) and failures (number of frogs with blue eyes) as one variable as the response using <code>cbind</code> to stick them together. If you’re not sure why we would do this, go back and read up on GLMs.</p>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb291-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit MCMCglmm</span></span>
<span id="cb291-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb291-2" aria-hidden="true" tabindex="-1"></a>model_eyes <span class="ot">&lt;-</span> <span class="fu">MCMCglmm</span>(<span class="fu">cbind</span>(numberred, numberblue) <span class="sc">~</span> <span class="fu">log</span>(mass), </span>
<span id="cb291-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb291-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> mydata, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>,</span>
<span id="cb291-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb291-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">random =</span> <span class="sc">~</span> Species,</span>
<span id="cb291-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb291-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ginverse =</span> <span class="fu">list</span>(<span class="at">Species =</span> inv.phylo), </span>
<span id="cb291-6"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb291-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">prior =</span> prior,</span>
<span id="cb291-7"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb291-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">nitt =</span> nitt, <span class="at">thin =</span> thin, <span class="at">burnin =</span> burnin,</span>
<span id="cb291-8"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb291-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Again, other than these changes, everything runs as above including the priors and all the model diagnostics and checking. Remember the parameter estimates for binomial GLMs are log odds ratios, so might look a little strange. Again, go back and read up on GLMs if this is all sounding unfamiliar to you.</p>
<div class="info">
<p>
If you’re worried about whether you’ve fitted the right model, and
the additional Bayesian components of <code>MCMCglmm</code> are causing
you confusion, it’s fine to fit the model using a standard GLM (using
the <code>glm</code> function in R) to see if the results are similar.
Remember that we expect the results to be somewhat different because
<code>glm</code> does not incorporate the phylogeny.
</p>
</div>
</div>
<div id="adding-extra-fixed-and-random-effects" class="section level2 hasAnchor" number="7.12">
<h2><span class="header-section-number">7.12</span> Adding extra fixed and random effects<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#adding-extra-fixed-and-random-effects" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>I won’t run these models here; this is just to give you an indication of how we can add complexity to these models.</p>
<p>Adding fixed effects to models is easy, we just add variables as we would in any other model in R. For example, if we wanted to fit this model from the PGLS exercise:</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb292-1" aria-hidden="true" tabindex="-1"></a>model.pgls2 <span class="ot">&lt;-</span> <span class="fu">pgls</span>(<span class="fu">log</span>(eyesize) <span class="sc">~</span> <span class="fu">log</span>(mass) <span class="sc">*</span> <span class="fu">as.factor</span>(Adult_habitat), </span>
<span id="cb292-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb292-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> frog, <span class="at">lambda =</span> <span class="st">&quot;ML&quot;</span>)</span></code></pre></div>
<p>We could fit a <code>MCMCglmm</code> model like this:</p>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb293-1" aria-hidden="true" tabindex="-1"></a>model_multiple<span class="ot">&lt;-</span> model_mcmcglmm <span class="ot">&lt;-</span> <span class="fu">MCMCglmm</span>(<span class="fu">log</span>(eyesize) <span class="sc">~</span> <span class="fu">log</span>(mass) <span class="sc">*</span> </span>
<span id="cb293-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb293-2" aria-hidden="true" tabindex="-1"></a>                                         <span class="fu">as.factor</span>(Adult_habitat), </span>
<span id="cb293-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb293-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> mydata, </span>
<span id="cb293-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb293-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">random =</span> <span class="sc">~</span> Species,</span>
<span id="cb293-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb293-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ginverse =</span> <span class="fu">list</span>(<span class="at">Species =</span> inv.phylo), </span>
<span id="cb293-6"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb293-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">prior =</span> prior,</span>
<span id="cb293-7"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb293-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">nitt =</span> nitt, <span class="at">thin =</span> thin, <span class="at">burnin =</span> burnin)</span></code></pre></div>
<p>Note that we are adding <em>fixed effects</em> in this example. If we want to add more <em>random effects</em> we need to be a bit more careful because we need to remember to define the priors for those random effects too.</p>
<p>Let’s say we wanted to fit a model with phylogeny as a random effect <em>and</em> a random effect of Sex. We’d now have two random effects in the G-structure.</p>
<p>We would define the priors as follows. Note that Sex is referred to as G2.</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb294-1" aria-hidden="true" tabindex="-1"></a>prior2 <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">G =</span> <span class="fu">list</span>(<span class="at">G1 =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">1</span>, <span class="at">n =</span> <span class="fl">0.002</span>), </span>
<span id="cb294-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb294-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">G2 =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">1</span>, <span class="at">n =</span> <span class="fl">0.002</span>)), </span>
<span id="cb294-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb294-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">R =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">1</span>, <span class="at">n =</span> <span class="fl">0.002</span>))</span></code></pre></div>
<p>We’d then fit the model as follows.</p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb295-1" aria-hidden="true" tabindex="-1"></a>model_new <span class="ot">&lt;-</span> model_mcmcglmm <span class="ot">&lt;-</span> <span class="fu">MCMCglmm</span>(<span class="fu">log</span>(eyesize) <span class="sc">~</span> <span class="fu">log</span>(mass), </span>
<span id="cb295-2"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb295-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">data =</span> mydata, </span>
<span id="cb295-3"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb295-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">random =</span> <span class="sc">~</span> Species <span class="sc">+</span> Sex,</span>
<span id="cb295-4"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb295-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ginverse =</span> <span class="fu">list</span>(<span class="at">Species =</span> inv.phylo), </span>
<span id="cb295-5"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb295-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">prior =</span> prior2,</span>
<span id="cb295-6"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#cb295-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">nitt =</span> nitt, <span class="at">thin =</span> thin, <span class="at">burnin =</span> burnin)</span></code></pre></div>
<p>Note that this gets more complex if you want to fit hierarchical or repeated measures type models so we recommend reading the materials below, and getting a thorough understanding of mixed models, before attempting this.</p>
</div>
<div id="summary-6" class="section level2 hasAnchor" number="7.13">
<h2><span class="header-section-number">7.13</span> Summary<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#summary-6" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>You should now know how to perform a GLMM analysis in R using the package <code>MCMCglmm</code>.</p>
<p>I also recommend reading the following if you want to know more: <a href="http://www.wildanimalmodels.org/tiki-download_wiki_attachment.php?attId=24" class="uri">http://www.wildanimalmodels.org/tiki-download_wiki_attachment.php?attId=24</a>
<a href="https://cran.r-project.org/web/packages/MCMCglmm/vignettes/CourseNotes.pdf" class="uri">https://cran.r-project.org/web/packages/MCMCglmm/vignettes/CourseNotes.pdf</a></p>
</div>
<div id="practical-exercises-2" class="section level2 hasAnchor" number="7.14">
<h2><span class="header-section-number">7.14</span> Practical exercises<a href="phylogenetic-generalised-linear-mixed-models-in-r.html#practical-exercises-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Let’s repeat the analyses we did in the PGLS chapter but using <code>MCMCglmm</code>. We will investigate the relationship between gestation length in Primates and their body size.</p>
<p>Read in the tree and data, then prepare them for a PCM analysis (you may have already done this in a previous exercise which should save you some time). Fit a MCMCglmm model to investigate the relationship between log gestation length (<code>y = log(GestationLen_d)</code>) and log body size (<code>x = log(AdultBodyMass_g)</code>) in Primates. Don’t forget to check that the model has converged, and to look at the model diagnostics and for autocorrelation.</p>
<p>Then answer the following questions.</p>
<ol style="list-style-type: decimal">
<li><p>What is <span class="math inline">\(\lambda\)</span> in the model?</p></li>
<li><p>What are the 95% credible intervals (HPD) around <span class="math inline">\(\lambda\)</span>?</p></li>
<li><p>Plot the <span class="math inline">\(\lambda\)</span> profile for the the maximum likelihood estimate of <span class="math inline">\(\lambda\)</span>.</p></li>
<li><p>Is there a significant relationship between log gestation length and log body size? What is the slope of this relationship?</p></li>
</ol>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-arnold201010ktrees" class="csl-entry">
Arnold, Christian, Luke J Matthews, and Charles L Nunn. 2010. <span>“The 10kTrees Website: A New Online Resource for Primate Phylogeny.”</span> <em>Evolutionary Anthropology: Issues, News, and Reviews</em> 19 (3): 114–18.
</div>
<div id="ref-feng2017phylogenomics" class="csl-entry">
Feng, Yan-Jie, David C Blackburn, Dan Liang, David M Hillis, David B Wake, David C Cannatella, and Peng Zhang. 2017. <span>“Phylogenomics Reveals Rapid, Simultaneous Diversification of Three Major Clades of Gondwanan Frogs at the Cretaceous–Paleogene Boundary.”</span> <em>Proceedings of the National Academy of Sciences</em> 114 (29): E5864–70.
</div>
<div id="ref-jones2009pantheria" class="csl-entry">
Jones, Kate E, Jon Bielby, Marcel Cardillo, Susanne A Fritz, Justin O’Dell, C David L Orme, Kamran Safi, et al. 2009. <span>“PanTHERIA: A Species-Level Database of Life History, Ecology, and Geography of Extant and Recently Extinct Mammals: Ecological Archives E090-184.”</span> <em>Ecology</em> 90 (9): 2648–48.
</div>
<div id="ref-mcelreath2020statistical" class="csl-entry">
McElreath, R. 2020. <em>Statistical Rethinking: A Bayesian Course with Examples in r and Stan</em>. CRC press.
</div>
<div id="ref-thomas2020eye" class="csl-entry">
Thomas, Kate N, David J Gower, Rayna C Bell, Matthew K Fujita, Ryan K Schott, and Jeffrey W Streicher. 2020. <span>“Eye Size and Investment in Frogs and Toads Correlate with Adult Habitat, Activity Pattern and Breeding Ecology.”</span> <em>Proceedings of the Royal Society B</em> 287 (1935): 20201393.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="phylogenetic-generalised-least-squares-pgls-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="macroevolutionary-models-for-continuous-traits.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/07-MCMCglmm.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["pcm-primer.pdf", "pcm-primer.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
