<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Preparing your tree and data for PCMs in R | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences</title>
  <meta name="description" content="Chapter 4 Preparing your tree and data for PCMs in R | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Preparing your tree and data for PCMs in R | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Preparing your tree and data for PCMs in R | Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences" />
  
  
  

<meta name="author" content="Natalie Cooper" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="understanding-plotting-and-manipulating-phylogenies-in-r.html"/>
<link rel="next" href="phylogenetic-signal-in-r.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">PCM Primer</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#datasets-and-scripts"><i class="fa fa-check"></i>Datasets and scripts</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#how-to-use-these-materials"><i class="fa fa-check"></i>How to use these materials</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#boxes"><i class="fa fa-check"></i>Boxes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#example-datasets"><i class="fa fa-check"></i>Example datasets</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#frog-eye-size-evolution"><i class="fa fa-check"></i>Frog eye size evolution</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#natricine-snake-head-shape-evolution"><i class="fa fa-check"></i>Natricine snake head shape evolution</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#diversification-in-dragonflies"><i class="fa fa-check"></i>Diversification in dragonflies</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#citing-r-and-r-packages"><i class="fa fa-check"></i>Citing R and R packages</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#acknowledgements"><i class="fa fa-check"></i>Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><i class="fa fa-check"></i><b>1</b> What you need to be able to do in R before you start</a>
<ul>
<li class="chapter" data-level="1.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-r-and-rstudio"><i class="fa fa-check"></i><b>1.1</b> Installing R (and RStudio)</a></li>
<li class="chapter" data-level="1.2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#downloading-the-data"><i class="fa fa-check"></i><b>1.2</b> Downloading the data</a></li>
<li class="chapter" data-level="1.3" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#r-projects"><i class="fa fa-check"></i><b>1.3</b> R Projects</a></li>
<li class="chapter" data-level="1.4" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#using-a-script"><i class="fa fa-check"></i><b>1.4</b> Using a script</a></li>
<li class="chapter" data-level="1.5" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#installing-and-loading-extra-packages-in-r"><i class="fa fa-check"></i><b>1.5</b> Installing and loading extra packages in R</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#loading-packages-into-r"><i class="fa fa-check"></i><b>1.5.1</b> Loading packages into R</a></li>
<li class="chapter" data-level="1.5.2" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#ggtree-a-special-case-for-package-installation"><i class="fa fa-check"></i><b>1.5.2</b> <code>ggtree</code>, a special case for package installation</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#loading-and-viewing-your-data-in-r"><i class="fa fa-check"></i><b>1.6</b> Loading and viewing your data in R</a></li>
<li class="chapter" data-level="1.7" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#errors-and-warnings"><i class="fa fa-check"></i><b>1.7</b> Errors and warnings</a>
<ul>
<li class="chapter" data-level="1.7.1" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#removing-the"><i class="fa fa-check"></i><b>1.7.1</b> Removing the <code>+</code></a></li>
</ul></li>
<li class="chapter" data-level="1.8" data-path="what-you-need-to-be-able-to-do-in-r-before-you-start.html"><a href="what-you-need-to-be-able-to-do-in-r-before-you-start.html#summary"><i class="fa fa-check"></i><b>1.8</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html"><i class="fa fa-check"></i><b>2</b> Data wrangling comparative data in R</a>
<ul>
<li class="chapter" data-level="" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#before-you-start"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="2.1" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#chameleon-data"><i class="fa fa-check"></i><b>2.1</b> Chameleon data</a></li>
<li class="chapter" data-level="2.2" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#summary-1"><i class="fa fa-check"></i><b>2.2</b> Summary</a></li>
<li class="chapter" data-level="2.3" data-path="data-wrangling-comparative-data-in-r.html"><a href="data-wrangling-comparative-data-in-r.html#practical-exercise"><i class="fa fa-check"></i><b>2.3</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html"><i class="fa fa-check"></i><b>3</b> Understanding, plotting and manipulating phylogenies in R</a>
<ul>
<li class="chapter" data-level="" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#before-you-start-1"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="3.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#what-do-phylogeny-files-look-like"><i class="fa fa-check"></i><b>3.1</b> What do phylogeny files look like?</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#newick-trees"><i class="fa fa-check"></i><b>3.1.1</b> Newick trees</a></li>
<li class="chapter" data-level="3.1.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#nexus-trees"><i class="fa fa-check"></i><b>3.1.2</b> NEXUS trees</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#how-are-phylogenies-formatted-in-r"><i class="fa fa-check"></i><b>3.2</b> How are phylogenies formatted in R?</a></li>
<li class="chapter" data-level="3.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#plotting-phylogenies"><i class="fa fa-check"></i><b>3.3</b> Plotting phylogenies</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#basic-phylogeny-plotting-with-ape"><i class="fa fa-check"></i><b>3.3.1</b> Basic phylogeny plotting with APE</a></li>
<li class="chapter" data-level="3.3.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#basic-phylogeny-plotting-with-ggtree"><i class="fa fa-check"></i><b>3.3.2</b> Basic phylogeny plotting with <code>ggtree</code></a></li>
<li class="chapter" data-level="3.3.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#back-to-ggtree"><i class="fa fa-check"></i><b>3.3.3</b> Back to <code>ggtree</code></a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#phylogeny-manipulation"><i class="fa fa-check"></i><b>3.4</b> Phylogeny manipulation</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#removing-polytomies"><i class="fa fa-check"></i><b>3.4.1</b> Removing polytomies</a></li>
<li class="chapter" data-level="3.4.2" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#rooting-your-phylogeny"><i class="fa fa-check"></i><b>3.4.2</b> Rooting your phylogeny</a></li>
<li class="chapter" data-level="3.4.3" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#manipulating-the-species-in-your-phlyogeny"><i class="fa fa-check"></i><b>3.4.3</b> Manipulating the species in your phlyogeny</a></li>
<li class="chapter" data-level="3.4.4" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#writing-phylogenies-to-file"><i class="fa fa-check"></i><b>3.4.4</b> Writing phylogenies to file</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#summary-2"><i class="fa fa-check"></i><b>3.5</b> Summary</a></li>
<li class="chapter" data-level="3.6" data-path="understanding-plotting-and-manipulating-phylogenies-in-r.html"><a href="understanding-plotting-and-manipulating-phylogenies-in-r.html#practical-exercise-1"><i class="fa fa-check"></i><b>3.6</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html"><i class="fa fa-check"></i><b>4</b> Preparing your tree and data for PCMs in R</a>
<ul>
<li class="chapter" data-level="" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#before-you-start-2"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="4.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#preparation"><i class="fa fa-check"></i><b>4.1</b> Preparation</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#load-the-required-packages"><i class="fa fa-check"></i><b>4.1.1</b> Load the required packages</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-and-checking-your-phylogeny-in-r"><i class="fa fa-check"></i><b>4.2</b> Reading and checking your phylogeny in R</a></li>
<li class="chapter" data-level="4.3" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-the-data-into-r"><i class="fa fa-check"></i><b>4.3</b> Reading the data into R</a></li>
<li class="chapter" data-level="4.4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-your-data-to-your-phylogeny"><i class="fa fa-check"></i><b>4.4</b> Matching your data to your phylogeny</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#species-names-with-spaces"><i class="fa fa-check"></i><b>4.4.1</b> Species names with spaces</a></li>
<li class="chapter" data-level="4.4.2" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#mismatches-between-species-in-your-data-and-phylogeny"><i class="fa fa-check"></i><b>4.4.2</b> Mismatches between species in your data and phylogeny</a></li>
<li class="chapter" data-level="4.4.3" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-the-tree-and-the-data"><i class="fa fa-check"></i><b>4.4.3</b> Matching the tree and the data</a></li>
<li class="chapter" data-level="4.4.4" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#nas-in-your-data"><i class="fa fa-check"></i><b>4.4.4</b> NAs in your data</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#quick-template-code"><i class="fa fa-check"></i><b>4.5</b> Quick template code</a></li>
<li class="chapter" data-level="4.6" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#summary-3"><i class="fa fa-check"></i><b>4.6</b> Summary</a></li>
<li class="chapter" data-level="4.7" data-path="preparing-your-tree-and-data-for-pcms-in-r.html"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#practical-exercise-2"><i class="fa fa-check"></i><b>4.7</b> Practical exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html"><i class="fa fa-check"></i><b>5</b> Phylogenetic Signal in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#before-you-start-3"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="5.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#preparation-1"><i class="fa fa-check"></i><b>5.1</b> Preparation</a></li>
<li class="chapter" data-level="5.2" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#estimating-phylogenetic-signal-for-continuous-variables"><i class="fa fa-check"></i><b>5.2</b> Estimating phylogenetic signal for continuous variables</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#pagels-lambda"><i class="fa fa-check"></i><b>5.2.1</b> Pagel’s <span class="math inline">\(\lambda\)</span></a></li>
<li class="chapter" data-level="5.2.2" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#blombergs-k-blomberg-et-al-2003"><i class="fa fa-check"></i><b>5.2.2</b> Blomberg’s <em>K</em> (Blomberg et al 2003)</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#estimating-phylogenetic-signal-for-non-continuous-variables"><i class="fa fa-check"></i><b>5.3</b> Estimating phylogenetic signal for non-continuous variables</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#d"><i class="fa fa-check"></i><b>5.3.1</b> <em>D</em></a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#summary-4"><i class="fa fa-check"></i><b>5.4</b> Summary</a></li>
<li class="chapter" data-level="5.5" data-path="phylogenetic-signal-in-r.html"><a href="phylogenetic-signal-in-r.html#practical-exercises"><i class="fa fa-check"></i><b>5.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html"><i class="fa fa-check"></i><b>6</b> Phylogenetic Generalised Least Squares (PGLS) in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#before-you-start-4"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="6.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#preparation-2"><i class="fa fa-check"></i><b>6.1</b> Preparation</a></li>
<li class="chapter" data-level="6.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#phylogenetic-generalized-least-squares-models-pgls"><i class="fa fa-check"></i><b>6.2</b> Phylogenetic generalized least squares models (PGLS)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#fitting-pgls-models-in-r"><i class="fa fa-check"></i><b>6.2.1</b> Fitting PGLS models in R</a></li>
<li class="chapter" data-level="6.2.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#model-diagnostics-for-pgls-models"><i class="fa fa-check"></i><b>6.2.2</b> Model diagnostics for PGLS models</a></li>
<li class="chapter" data-level="6.2.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#likelihood-profiles-for-lambda-in-pgls-models"><i class="fa fa-check"></i><b>6.2.3</b> Likelihood profiles for <span class="math inline">\(\lambda\)</span> in PGLS models</a></li>
<li class="chapter" data-level="6.2.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#a-pgls-example-with-more-than-one-predictor-variable"><i class="fa fa-check"></i><b>6.2.4</b> A PGLS example with more than one predictor variable</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#common-issues-when-running-pgls-models-in-r"><i class="fa fa-check"></i><b>6.3</b> Common issues when running PGLS models in R</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#optim-errors"><i class="fa fa-check"></i><b>6.3.1</b> Optim errors</a></li>
<li class="chapter" data-level="6.3.2" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#big-problems-with-small-datasets"><i class="fa fa-check"></i><b>6.3.2</b> Big problems with small datasets</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#summary-5"><i class="fa fa-check"></i><b>6.4</b> Summary</a></li>
<li class="chapter" data-level="6.5" data-path="phylogenetic-generalised-least-squares-pgls-in-r.html"><a href="phylogenetic-generalised-least-squares-pgls-in-r.html#practical-exercises-1"><i class="fa fa-check"></i><b>6.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html"><i class="fa fa-check"></i><b>7</b> Phylogenetic generalised linear mixed models in R</a>
<ul>
<li class="chapter" data-level="" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#before-you-start-5"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="7.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#why-fit-phylogenetic-generalised-linear-mixed-models"><i class="fa fa-check"></i><b>7.1</b> Why fit phylogenetic generalised linear mixed models?</a></li>
<li class="chapter" data-level="7.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#a-really-quick-reminder-of-how-bayesian-analyses-work-with-apologies-to-bayesian-scholars-for-the-simplification"><i class="fa fa-check"></i><b>7.2</b> A really quick reminder of how Bayesian analyses work, with apologies to Bayesian scholars for the simplification!</a></li>
<li class="chapter" data-level="7.3" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#preparation-3"><i class="fa fa-check"></i><b>7.3</b> Preparation</a></li>
<li class="chapter" data-level="7.4" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmcglmm-specific-preparation"><i class="fa fa-check"></i><b>7.4</b> MCMCglmm specific preparation</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#inverse-phylogenetic-variance-covariance-matrix"><i class="fa fa-check"></i><b>7.4.1</b> Inverse phylogenetic variance-covariance matrix</a></li>
<li class="chapter" data-level="7.4.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#priors"><i class="fa fa-check"></i><b>7.4.2</b> Priors</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#other-arguments-in-mcmcglmm"><i class="fa fa-check"></i><b>7.5</b> Other arguments in <code>MCMCglmm</code></a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#nitt"><i class="fa fa-check"></i><b>7.5.1</b> nitt</a></li>
<li class="chapter" data-level="7.5.2" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#burnin"><i class="fa fa-check"></i><b>7.5.2</b> burnin</a></li>
<li class="chapter" data-level="7.5.3" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#thin"><i class="fa fa-check"></i><b>7.5.3</b> thin</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#fitting-the-mcmcglmm-model"><i class="fa fa-check"></i><b>7.6</b> Fitting the MCMCglmm model</a></li>
<li class="chapter" data-level="7.7" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#did-the-model-converge"><i class="fa fa-check"></i><b>7.7</b> Did the model converge?</a></li>
<li class="chapter" data-level="7.8" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#mcmc-outputs"><i class="fa fa-check"></i><b>7.8</b> MCMC outputs</a></li>
<li class="chapter" data-level="7.9" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#extracting-lambda-for-mcmcglmm-models"><i class="fa fa-check"></i><b>7.9</b> Extracting lambda for MCMCglmm models</a></li>
<li class="chapter" data-level="7.10" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#checking-the-effects-of-priors"><i class="fa fa-check"></i><b>7.10</b> Checking the effects of priors</a></li>
<li class="chapter" data-level="7.11" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#an-example-using-a-pglmm"><i class="fa fa-check"></i><b>7.11</b> An example using a PGLMM</a></li>
<li class="chapter" data-level="7.12" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#adding-extra-fixed-and-random-effects"><i class="fa fa-check"></i><b>7.12</b> Adding extra fixed and random effects</a></li>
<li class="chapter" data-level="7.13" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#summary-6"><i class="fa fa-check"></i><b>7.13</b> Summary</a></li>
<li class="chapter" data-level="7.14" data-path="phylogenetic-generalised-linear-mixed-models-in-r.html"><a href="phylogenetic-generalised-linear-mixed-models-in-r.html#practical-exercises-2"><i class="fa fa-check"></i><b>7.14</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html"><i class="fa fa-check"></i><b>8</b> Macroevolutionary models for continuous traits</a>
<ul>
<li class="chapter" data-level="" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#before-you-start-6"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="8.1" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#preparation-4"><i class="fa fa-check"></i><b>8.1</b> Preparation</a></li>
<li class="chapter" data-level="8.2" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#models-of-evolution-for-continuous-traits"><i class="fa fa-check"></i><b>8.2</b> Models of evolution for continuous traits</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#look-at-the-data"><i class="fa fa-check"></i><b>8.2.1</b> Look at the data</a></li>
<li class="chapter" data-level="8.2.2" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#fitting-the-brownian-motion-and-ou-models-of-evolution-using-fitcontinuous"><i class="fa fa-check"></i><b>8.2.2</b> Fitting the Brownian motion and OU models of evolution using <code>fitContinuous</code></a></li>
<li class="chapter" data-level="8.2.3" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#comparing-models-of-evolution-using-aic"><i class="fa fa-check"></i><b>8.2.3</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#summary-7"><i class="fa fa-check"></i><b>8.3</b> Summary</a></li>
<li class="chapter" data-level="8.4" data-path="macroevolutionary-models-for-continuous-traits.html"><a href="macroevolutionary-models-for-continuous-traits.html#practical-exercises-3"><i class="fa fa-check"></i><b>8.4</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html"><i class="fa fa-check"></i><b>9</b> Macroevolutionary models for discrete traits</a>
<ul>
<li class="chapter" data-level="" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#before-you-start-7"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="9.1" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#preparation-5"><i class="fa fa-check"></i><b>9.1</b> Preparation</a></li>
<li class="chapter" data-level="9.2" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#models-of-evolution-for-discrete-traits"><i class="fa fa-check"></i><b>9.2</b> Models of evolution for discrete traits</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#fitting-the-er-sym-and-ard-models-of-evolution-using-fitdiscrete"><i class="fa fa-check"></i><b>9.2.1</b> Fitting the ER, SYM and ARD models of evolution using <code>fitDiscrete</code></a></li>
<li class="chapter" data-level="9.2.2" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#comparing-models-of-evolution-using-aic-1"><i class="fa fa-check"></i><b>9.2.2</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#ancestral-state-estimations"><i class="fa fa-check"></i><b>9.3</b> Ancestral state estimations</a></li>
<li class="chapter" data-level="9.4" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#summary-8"><i class="fa fa-check"></i><b>9.4</b> Summary</a></li>
<li class="chapter" data-level="9.5" data-path="macroevolutionary-models-for-discrete-traits.html"><a href="macroevolutionary-models-for-discrete-traits.html#practical-exercises-4"><i class="fa fa-check"></i><b>9.5</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html"><i class="fa fa-check"></i><b>10</b> Beyond simple macroevolutionary models</a>
<ul>
<li class="chapter" data-level="" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#before-you-start-8"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="10.1" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#preparation-6"><i class="fa fa-check"></i><b>10.1</b> Preparation</a></li>
<li class="chapter" data-level="10.2" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#more-complex-models-of-trait-evolution"><i class="fa fa-check"></i><b>10.2</b> More complex models of trait evolution</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#preparing-the-tree-and-data-for-ouwie"><i class="fa fa-check"></i><b>10.2.1</b> Preparing the tree and data for OUwie</a></li>
<li class="chapter" data-level="10.2.2" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#running-models-using-ouwie"><i class="fa fa-check"></i><b>10.2.2</b> Running models using OUwie</a></li>
<li class="chapter" data-level="10.2.3" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#single-rate-bm-models"><i class="fa fa-check"></i><b>10.2.3</b> Single rate BM models</a></li>
<li class="chapter" data-level="10.2.4" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#single-optimum-ou-models"><i class="fa fa-check"></i><b>10.2.4</b> Single optimum OU models</a></li>
<li class="chapter" data-level="10.2.5" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-bm-bms-models"><i class="fa fa-check"></i><b>10.2.5</b> Multi-rate BM (BMS) models</a></li>
<li class="chapter" data-level="10.2.6" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-optima-ou-oum-models"><i class="fa fa-check"></i><b>10.2.6</b> Multi-optima OU (OUM) models</a></li>
<li class="chapter" data-level="10.2.7" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-multi-optima-ou-motion-models-oumv"><i class="fa fa-check"></i><b>10.2.7</b> Multi-rate multi-optima OU motion models (OUMV)</a></li>
<li class="chapter" data-level="10.2.8" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-alpha-multi-optima-ou-motion-models-ouma"><i class="fa fa-check"></i><b>10.2.8</b> Multi-alpha multi-optima OU motion models (OUMA)</a></li>
<li class="chapter" data-level="10.2.9" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#multi-rate-multi-optima-ou-motion-models-oumva"><i class="fa fa-check"></i><b>10.2.9</b> Multi-rate multi-optima OU motion models (OUMVA)</a></li>
<li class="chapter" data-level="10.2.10" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#comparing-models-of-evolution-using-aic-2"><i class="fa fa-check"></i><b>10.2.10</b> Comparing models of evolution using AIC</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#summary-9"><i class="fa fa-check"></i><b>10.3</b> Summary</a></li>
<li class="chapter" data-level="10.4" data-path="beyond-simple-macroevolutionary-models.html"><a href="beyond-simple-macroevolutionary-models.html#practical-exercises-5"><i class="fa fa-check"></i><b>10.4</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html"><i class="fa fa-check"></i><b>11</b> Diversification rate analyses</a>
<ul>
<li class="chapter" data-level="" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#before-you-start-9"><i class="fa fa-check"></i>Before you start</a></li>
<li class="chapter" data-level="11.1" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#preparation-7"><i class="fa fa-check"></i><b>11.1</b> Preparation</a></li>
<li class="chapter" data-level="11.2" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#lineage-through-time-plots-and-the-gamma-statistic"><i class="fa fa-check"></i><b>11.2</b> Lineage through time plots and the <span class="math inline">\(\gamma\)</span> statistic</a></li>
<li class="chapter" data-level="11.3" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#bamm-bayesian-analysis-of-macroevolutionary-mixtures"><i class="fa fa-check"></i><b>11.3</b> BAMM: Bayesian Analysis of Macroevolutionary Mixtures</a></li>
<li class="chapter" data-level="11.4" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#running-bamm"><i class="fa fa-check"></i><b>11.4</b> Running BAMM</a></li>
<li class="chapter" data-level="11.5" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#looking-at-the-output-of-bamm-using-bammtools-in-r"><i class="fa fa-check"></i><b>11.5</b> Looking at the output of BAMM using BAMMtools in R</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#assessing-mcmc-convergence"><i class="fa fa-check"></i><b>11.5.1</b> Assessing MCMC convergence</a></li>
<li class="chapter" data-level="11.5.2" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#analysis-of-rate-shifts-in-the-dragonfly-phylogeny"><i class="fa fa-check"></i><b>11.5.2</b> Analysis of rate shifts in the dragonfly phylogeny</a></li>
<li class="chapter" data-level="11.5.3" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#mean-phylorate-plot"><i class="fa fa-check"></i><b>11.5.3</b> Mean phylorate plot</a></li>
<li class="chapter" data-level="11.5.4" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#best-single-distinct-shift-configuration"><i class="fa fa-check"></i><b>11.5.4</b> Best single distinct shift configuration</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#credible-sets-of-shifts"><i class="fa fa-check"></i><b>11.6</b> Credible sets of shifts</a></li>
<li class="chapter" data-level="11.7" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#clade-specific-evolutionary-rates"><i class="fa fa-check"></i><b>11.7</b> Clade specific evolutionary rates</a></li>
<li class="chapter" data-level="11.8" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#rate-through-time-analysis"><i class="fa fa-check"></i><b>11.8</b> Rate-through-time analysis</a></li>
<li class="chapter" data-level="11.9" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#macroevolutionary-cohort-analysis"><i class="fa fa-check"></i><b>11.9</b> Macroevolutionary cohort analysis</a></li>
<li class="chapter" data-level="11.10" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#assumptions-and-issues-with-bamm"><i class="fa fa-check"></i><b>11.10</b> Assumptions and issues with BAMM</a></li>
<li class="chapter" data-level="11.11" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#summary-10"><i class="fa fa-check"></i><b>11.11</b> Summary</a></li>
<li class="chapter" data-level="11.12" data-path="diversification-rate-analyses.html"><a href="diversification-rate-analyses.html#practical-exercises-6"><i class="fa fa-check"></i><b>11.12</b> Practical exercises</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html"><i class="fa fa-check"></i><b>12</b> Answers to practical exercises</a>
<ul>
<li class="chapter" data-level="12.1" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#need-to-know-r"><i class="fa fa-check"></i><b>12.1</b> Need to know R</a></li>
<li class="chapter" data-level="12.2" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#data-wrangling"><i class="fa fa-check"></i><b>12.2</b> Data wrangling</a></li>
<li class="chapter" data-level="12.3" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#phylogenies"><i class="fa fa-check"></i><b>12.3</b> Phylogenies</a></li>
<li class="chapter" data-level="12.4" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#preparation-8"><i class="fa fa-check"></i><b>12.4</b> Preparation</a></li>
<li class="chapter" data-level="12.5" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#phylogenetic-signal"><i class="fa fa-check"></i><b>12.5</b> Phylogenetic signal</a></li>
<li class="chapter" data-level="12.6" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#pgls"><i class="fa fa-check"></i><b>12.6</b> PGLS</a></li>
<li class="chapter" data-level="12.7" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#mcmcglmm"><i class="fa fa-check"></i><b>12.7</b> MCMCglmm</a></li>
<li class="chapter" data-level="12.8" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#models-of-evolution-with-continuous-traits"><i class="fa fa-check"></i><b>12.8</b> Models of evolution with continuous traits</a></li>
<li class="chapter" data-level="12.9" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#models-of-evolution-with-discrete-traits"><i class="fa fa-check"></i><b>12.9</b> Models of evolution with discrete traits</a></li>
<li class="chapter" data-level="12.10" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#beyond-simple-macroevolutionary-models-of-trait-evolution"><i class="fa fa-check"></i><b>12.10</b> Beyond simple macroevolutionary models of trait evolution</a></li>
<li class="chapter" data-level="12.11" data-path="answers-to-practical-exercises.html"><a href="answers-to-practical-exercises.html#diversification"><i class="fa fa-check"></i><b>12.11</b> Diversification</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html"><i class="fa fa-check"></i><b>13</b> Common mistakes using PCMs in R</a>
<ul>
<li class="chapter" data-level="13.1" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#good-practice-to-prevent-errors"><i class="fa fa-check"></i><b>13.1</b> Good practice to prevent errors</a></li>
<li class="chapter" data-level="13.2" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#common-errors"><i class="fa fa-check"></i><b>13.2</b> Common errors</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#standard-r-issues"><i class="fa fa-check"></i><b>13.2.1</b> Standard R issues</a></li>
<li class="chapter" data-level="13.2.2" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#issues-with-the-tree-or-data"><i class="fa fa-check"></i><b>13.2.2</b> Issues with the tree or data</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#what-to-do-if-you-get-an-error-message"><i class="fa fa-check"></i><b>13.3</b> What to do if you get an Error message</a></li>
<li class="chapter" data-level="13.4" data-path="common-mistakes-using-pcms-in-r.html"><a href="common-mistakes-using-pcms-in-r.html#what-to-do-if-you-get-a-warning-message"><i class="fa fa-check"></i><b>13.4</b> What to do if you get a Warning message</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Online materials for Primer on Phylogenetic Comparative Methods for the Biological Sciences</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="preparing-your-tree-and-data-for-pcms-in-r" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">Chapter 4</span> Preparing your tree and data for PCMs in R<a href="preparing-your-tree-and-data-for-pcms-in-r.html#preparing-your-tree-and-data-for-pcms-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Before we do any analysis in R we will generally need to clean and tidy our data. Data cleaning involves looking for errors, for example we might need to remove duplicate columns, or correct typos in species names, or make sure column headers are all unique and easy enough for us to type in R. We also often need to tidy our data. For data to be <em>tidy</em>:</p>
<ul>
<li>Each variable must have its own column.</li>
<li>Each observation must have its own row.</li>
<li>Each value must have its own cell.</li>
</ul>
<p>The R packages <code>dplyr</code>, <code>tidyr</code> are great for data cleaning and tidying. See <a href="https://r4ds.had.co.nz/tidy-data.html">https://r4ds.had.co.nz/tidy-data.html</a> for more details on what tidy data is and how to do this in R. Of course you are welcome to do this in Excel or another package if you’re more comfortable with that. The important thing is that when you begin your analyses, your data are in a suitable format. Note that a detailed introduction to data manipulation in R is beyond the scope of this <em>Primer</em>, but if in doubt about your data, try and make it look as much like our example datasets as possible. If you can do that, these R analyses should work for you! :)</p>
<p>In this exercise we will take an existing tree and some data and prepare them for a PCM analysis. Note that prior to this step, it is important that your data is in a tidy format, and has been thoroughly checked for errors. The exercise below only shows you how to deal with additional issues caused by using trees and then matching species data to those trees. We assume the tree and data themselves are fit for purpose.</p>
<div class="info">
<p>
We will be using the evolution of eye size in frogs as an example.
The data and modified tree come from <span class="citation"><span class="citation">K. N. Thomas et al. (<a href="#ref-thomas2020eye" role="doc-biblioref">2020</a>)</span></span>, and the original tree comes from
<span class="citation"><span class="citation">Feng et al. (<a href="#ref-feng2017phylogenomics" role="doc-biblioref">2017</a>)</span></span>. I’ve removed a few
species and a few variables to make things a bit more straightforward.
If you want to see the full results check out <span class="citation"><span class="citation">K. N. Thomas et al. (<a href="#ref-thomas2020eye" role="doc-biblioref">2020</a>)</span></span>!
</p>
</div>
<div id="before-you-start-2" class="section level2 unnumbered hasAnchor">
<h2>Before you start<a href="preparing-your-tree-and-data-for-pcms-in-r.html#before-you-start-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li>Open the <code>04-Preparation.RProj</code> file in the <code>04-Preparation</code> folder to open your R Project for this exercise.</li>
</ul>
<p>You will also need to install the following packages:</p>
<ul>
<li><code>tidyverse</code> - for reading, manipulating and plotting data</li>
<li><code>ape</code> - functions for reading, plotting and manipulating phylogenies</li>
<li><code>geiger</code> - to check species in the tree and data match</li>
<li><code>phytools</code> - for force.ultrametric</li>
</ul>
</div>
<div id="preparation" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> Preparation<a href="preparing-your-tree-and-data-for-pcms-in-r.html#preparation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="load-the-required-packages" class="section level3 hasAnchor" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Load the required packages<a href="preparing-your-tree-and-data-for-pcms-in-r.html#load-the-required-packages" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To begin we need to load the packages we need.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load packages</span></span>
<span id="cb91-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb91-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb91-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb91-4"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
</div>
</div>
<div id="reading-and-checking-your-phylogeny-in-r" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> Reading and checking your phylogeny in R<a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-and-checking-your-phylogeny-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We already learned how to do these things in <code>03-Phylogenies</code>. If you haven’t looked at that exercise I suggest you check it out before doing the steps below.</p>
<p>First let’s read in the tree and look at its structure:</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the tree</span></span>
<span id="cb92-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb92-2" aria-hidden="true" tabindex="-1"></a>frogtree <span class="ot">&lt;-</span> <span class="fu">read.nexus</span>(<span class="st">&quot;data/frog-tree.nex&quot;</span>)</span>
<span id="cb92-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb92-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the tree summary</span></span>
<span id="cb92-4"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb92-4" aria-hidden="true" tabindex="-1"></a>frogtree</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 214 tips and 213 internal nodes.
## 
## Tip labels:
##   Ascaphus_truei_Ascaphidae, Leiopelma_hochstetteri_Leiopelmatidae, Alytes_obstetricans_Alytidae, Discoglossus_pictus_Alytidae, Barbourula_busuangensis_Bombinatoridae, Bombina_orientalis_Bombinatoridae, ...
## 
## Rooted; includes branch lengths.</code></pre>
<p>It’s usually a good idea to quickly plot the tree too…</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the tree as a circular/fan phylogeny with small labels</span></span>
<span id="cb94-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(frogtree, <span class="at">cex =</span> <span class="fl">0.2</span>, <span class="at">typ =</span> <span class="st">&quot;fan&quot;</span>, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="pcm-primer_files/figure-html/unnamed-chunk-85-1.png" width="672" /></p>
<p><code>frogtree</code> is a fully resolved tree with branch lengths. There are 214 species and 213 internal nodes. Note that the species names at the tips also have family names added to them…</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the first 5 tip labels</span></span>
<span id="cb95-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb95-2" aria-hidden="true" tabindex="-1"></a>frogtree<span class="sc">$</span>tip.label[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>## [1] &quot;Ascaphus_truei_Ascaphidae&quot;              &quot;Leiopelma_hochstetteri_Leiopelmatidae&quot; 
## [3] &quot;Alytes_obstetricans_Alytidae&quot;           &quot;Discoglossus_pictus_Alytidae&quot;          
## [5] &quot;Barbourula_busuangensis_Bombinatoridae&quot;</code></pre>
<p>Most trees will have just the genus and species names, but some will have additional information like here, or numbering etc. This is not a problem as long as the names match those in your data (see below).</p>
<p>Most R functions require your tree to be dichotomous, i.e. to have no polytomies. To check whether your tree is dichotomous use <code>is.binary</code>.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is binary</span></span>
<span id="cb97-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb97-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE</span></span>
<span id="cb97-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.binary</span>(frogtree) </span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>If this was FALSE, we’d use <code>multi2di</code> to make the tree dichotomous, but here it is TRUE so we can leave it as it is.</p>
<p>Most functions also require the tree to be rooted, i.e., to have one taxon designated as the outgroup. We can check whether the tree is rooted as follows.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is rooted</span></span>
<span id="cb99-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb99-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE</span></span>
<span id="cb99-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb99-3" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(frogtree)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Our tree is rooted so it’s ready to go.</p>
<p>Finally we might want to check that the tree is ultrametric. Most functions will assume this, and although we can see the tree looks ultrametric when we plotted it above, we should still check.</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the tree is ultrametric</span></span>
<span id="cb101-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb101-2" aria-hidden="true" tabindex="-1"></a><span class="fu">is.ultrametric</span>(frogtree)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>If this is <code>FALSE</code> there are two options. If the tree really is non-ultrametric, for example if it contains fossil species, then we can’t use it in methods that require an ultrametric tree. However, there are cases (the dragonfly tree in <em>Chapter 11</em> is one example) where the tree is non-ultrametric due to a rounding error. In the latter case we can use the <code>phytools</code> function <code>force.ultrametric</code> to fix this.</p>
<div class="detail">
<p>
<strong>When can a tree that looks ultrametric appear to be
non-ultrametric?</strong> Most of the methods we use expect trees to be
ultrametric, i.e. that all the tips line up, generally at the present
day (time 0). Sometimes (see <em>Chapter 11</em> the tree and it looks
ultrametric, but when we check using <code>is.ultrametric</code> the
answer returned is <code>FALSE</code>. What is going on? The tree is
actually ultrametric, and will run in most analyses with R treating it
as an ultrametric tree. The reason <code>is.ultrametric</code> tells us
it is not ultrametric is related to rounding errors. When you save a
tree file, it will save the branch lengths to a certain level of
accuracy, but not always the full level of accuracy if your branch
lengths have lots and lots of decimal places. When you read these trees
back into R, a teeny tiny bit of the accuracy is lost which sometimes
means that when R adds up the root-to-tip distances for each tip, they
aren’t all exactly the same length, so therefore technically the tree is
not ultrametric. For most implementations this tiny difference is not a
big deal. A quick fix to this problem is to use the function
<code>force.ultrametric</code> which essentially fudges the numbers to
force the tree to be ultrametric. <strong>If your tree is genuinely
non-ultrametric you should not use this function</strong> as it can
introduce negative branch lengths which will break most functions you
might want to use in R.
</p>
</div>
</div>
<div id="reading-the-data-into-r" class="section level2 hasAnchor" number="4.3">
<h2><span class="header-section-number">4.3</span> Reading the data into R<a href="preparing-your-tree-and-data-for-pcms-in-r.html#reading-the-data-into-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The data are in a comma-delimited text file called <code>frog-eyes.csv</code>. Load these data as follows.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the data</span></span>
<span id="cb103-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb103-2" aria-hidden="true" tabindex="-1"></a>frogdata <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/frog-eyes.csv&quot;</span>)</span></code></pre></div>
<p>Check everything loaded correctly:</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the data</span></span>
<span id="cb104-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(frogdata)</span></code></pre></div>
<pre><code>## Rows: 215
## Columns: 11
## $ Binomial         &lt;chr&gt; &quot;Allophryne_ruthveni&quot;, &quot;Eupsophus_roseus&quot;, &quot;Alytes_obstetricans&quot;, &quot;…
## $ Family           &lt;chr&gt; &quot;Allophrynidae&quot;, &quot;Alsodidae&quot;, &quot;Alytidae&quot;, &quot;Alytidae&quot;, &quot;Aromobatidae…
## $ Genus            &lt;chr&gt; &quot;Allophryne&quot;, &quot;Eupsophus&quot;, &quot;Alytes&quot;, &quot;Discoglossus&quot;, &quot;Allobates&quot;, &quot;…
## $ tiplabel         &lt;chr&gt; &quot;Allophryne_ruthveni_Allophrynidae&quot;, &quot;Eupsophus_roseus_Alsodidae&quot;, …
## $ Adult_habitat    &lt;chr&gt; &quot;Scansorial&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwellin…
## $ Life_history     &lt;chr&gt; &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;…
## $ Sex_dichromatism &lt;chr&gt; &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent…
## $ SVL              &lt;dbl&gt; 23.76667, 38.37500, 37.46667, 62.64000, 25.60000, 27.03333, 38.4000…
## $ mass             &lt;dbl&gt; 1.0000000, 7.5000000, 6.6666667, 24.4000000, 2.1000000, 2.4666667, …
## $ rootmass         &lt;dbl&gt; 0.9917748, 1.9273451, 1.8690601, 2.8885465, 1.2549697, 1.3507613, 1…
## $ eyesize          &lt;dbl&gt; 3.200000, 5.362500, 6.366667, 7.550000, 4.116667, 3.783333, 4.75000…</code></pre>
<p>As you can see, the data contains 215 species, and the following 11 variables:</p>
<ul>
<li><code>Binomial</code> - the species binomial name.<br />
</li>
<li><code>Family</code> - the family the species belongs to.<br />
</li>
<li><code>Genus</code> - the genus the species belongs to.<br />
</li>
<li><code>tiplabel</code> - the name used for the species in the phylogeny.<br />
</li>
<li><code>Adult_habitat</code> - habitat of adults. Categories are: Ground-dwelling, Subfossorial, Scansorial (i.e. tree-dwelling), Semiaquatic, Aquatic, or Fossorial (i.e. burrowers).<br />
</li>
<li><code>Life_history</code> - whether the larvae are free-living (Free-living larvae) or not (No free-living larvae).<br />
</li>
<li><code>Sex_dichromatism</code> - whether different sexes are different colours (Present) or not (Absent).</li>
<li><code>SVL</code> - snout vent length (in mm). This is a common way to measure body size in amphibians.<br />
</li>
<li><code>mass</code> - body mass (in g).</li>
<li><code>rootmass</code> - cube root of the body mass.<br />
</li>
<li><code>eyesize</code> - eye size (in mm) for the species. This is an everage across left and right eyes from three individuals per species.</li>
</ul>
<div class="detail">
<p>
Note that in some comparative datasets (most of mine for instance)
the species names column (here <code>Binomial</code>) also contains the
names of the species in the tree so we would match up that column to the
species names in the tree (see below). In this dataset and others, the
authors have instead included the names of the tips of the tree as a
separate column (here <code>tiplabel</code>). This is often done when
the tips of the tree are not species names, but are instead some kind of
code. This is common where the tree contains multiple tips for one
species, or where the tip labels contain additional information (here
they include family names). A final approach used by some authors is to
have the species names as the row names of the dataset. Any approach is
fine, but make sure you know which column contains the names that should
match up with the tree.
</p>
</div>
</div>
<div id="matching-your-data-to-your-phylogeny" class="section level2 hasAnchor" number="4.4">
<h2><span class="header-section-number">4.4</span> Matching your data to your phylogeny<a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-your-data-to-your-phylogeny" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now we have the tree and the data in R, we need to match the two up if we want to perform any kind of PCM analyses. Below are some common issues you might encounter, and how to fix them.</p>
<div id="species-names-with-spaces" class="section level3 hasAnchor" number="4.4.1">
<h3><span class="header-section-number">4.4.1</span> Species names with spaces<a href="preparing-your-tree-and-data-for-pcms-in-r.html#species-names-with-spaces" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Species names in phylogenies are generally written as <em>Genus_species</em> (the gap between the genus name and species name is replaced by an underscore <code>_</code>). If the species names in the data are written as <em>Genus species</em> with a space, then you will have to replace the spaces with <code>_</code> so that they match up with the species names in the tree. You can do this as follows using <code>str_replace</code>.</p>
<p>We don’t need to do this in our frog data, but if we did we could use code like this:</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace spaces with underscores in species names</span></span>
<span id="cb106-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb106-2" aria-hidden="true" tabindex="-1"></a>frogdata <span class="ot">&lt;-</span> </span>
<span id="cb106-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb106-3" aria-hidden="true" tabindex="-1"></a>  frogdata <span class="sc">%&gt;%</span></span>
<span id="cb106-4"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb106-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Binomial =</span> <span class="fu">str_replace</span>(<span class="at">string =</span> Binomial, <span class="at">pattern =</span> <span class="st">&quot; &quot;</span>, <span class="at">replacement =</span> <span class="st">&quot;_&quot;</span>))</span>
<span id="cb106-5"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb106-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb106-6"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb106-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Check it worked</span></span>
<span id="cb106-7"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb106-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(frogdata<span class="sc">$</span>Binomial)</span></code></pre></div>
<div class="warning">
<p>
Beware trailing spaces! If you’re trying to match up species names
and are having some trouble, check what happened when you replaced
spaces with underscores (<code>_</code>). Sometimes we accidentally
leave extra spaces when typing, and R will convert these to underscores
too which can cause problems. For example, <code>Genus_species</code>
can become <code>Genus_species_</code> if you accidentally left a
trailing space. Check all the names in your dataset after replacing the
spaces to make sure this is not a problem.
</p>
</div>
</div>
<div id="mismatches-between-species-in-your-data-and-phylogeny" class="section level3 hasAnchor" number="4.4.2">
<h3><span class="header-section-number">4.4.2</span> Mismatches between species in your data and phylogeny<a href="preparing-your-tree-and-data-for-pcms-in-r.html#mismatches-between-species-in-your-data-and-phylogeny" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Often you will have data for species which are not in your phylogeny and/or species in your phylogeny which are not in your data. Many functions in R can deal with this and will match the species for you, others will produce an error telling you the tree and data do not match (e.g. most <code>ape</code> functions).</p>
<p>Even in functions that can cope with this, it’s useful to match up the species <em>before</em> your analyses. This can help you identify things like spelling mistakes or variations in the taxonomy of the tree and the data.</p>
<div class="info">
<p>
If you have even slightly misspelled a species name in the tree or
the data it will automatically be dropped from the analyses. It is
therefore <strong>very important</strong> to check this before running
an analysis, especially one with lots of taxa.
</p>
</div>
<p>We can use the <code>geiger</code> function <code>name.check</code> to find out which names do not match. Remember that the species names that match up with the tree from <code>frogdata</code> are in the variable called <code>tiplabel</code>.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the names match in the data and the tree</span></span>
<span id="cb107-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb107-2" aria-hidden="true" tabindex="-1"></a>check <span class="ot">&lt;-</span> <span class="fu">name.check</span>(<span class="at">phy =</span> frogtree, <span class="at">data =</span> frogdata, </span>
<span id="cb107-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb107-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">data.names =</span> frogdata<span class="sc">$</span>tiplabel)</span></code></pre></div>
<p>The output of <code>check</code> has two parts, <code>tree_not_data</code> for species in the tree but not in the dataset, and <code>data_not_tree</code> for species in the dataset but not in the tree. You need to look at both of these in turn.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at species in the tree but not the data</span></span>
<span id="cb108-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb108-2" aria-hidden="true" tabindex="-1"></a>check<span class="sc">$</span>tree_not_data</span></code></pre></div>
<pre><code>## [1] &quot;Incilius_nebulifer_Bufonidae&quot;            &quot;Leptobrachella_bidoupensis_Megophryidae&quot;
## [3] &quot;Microhyla_fissipes_Microhylidae&quot;         &quot;Microhyla_marmorata_Microhylidae&quot;</code></pre>
<p>There are four species in the tree and not the data. We were expecting this here, so no worries.</p>
<p>For your analyses you should <em>always</em> check this list carefully. If I were running this analysis for the first time I’d want to check that these species really weren’t in my data. Maybe they are misspelled in the data? If so correct this now. Maybe the species name has changed? If so change this now. It doesn’t really matter whether you make the change in the tree or the data, but make sure anything that <em>should</em> match up, does. I’d usually fix issues in the dataset in Excel or another spreadsheet program, unless it’s a blatant typo in the tree.</p>
<p>Next check the species in the data but not the tree.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at species in the data but not the tree</span></span>
<span id="cb110-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb110-2" aria-hidden="true" tabindex="-1"></a>check<span class="sc">$</span>data_not_tree</span></code></pre></div>
<pre><code>## [1] &quot;Gastrophryne_carolinensis_Microhylidae&quot; &quot;Leptobrachella_dringi_Megophryidae&quot;    
## [3] &quot;Megophrys_gerti_Megophryidae&quot;           &quot;Microhyla_pulverata_Microhylidae&quot;      
## [5] &quot;Oreobates_quixensis_Strabomantidae&quot;</code></pre>
<p>This gives us five species in the data but not in the tree. As above, make sure to correct any errors before moving to the next step. Make sure anything that <em>should</em> match up, does match up.</p>
<p>Here we know that these species are missing from our data, so we don’t need to worry.</p>
</div>
<div id="matching-the-tree-and-the-data" class="section level3 hasAnchor" number="4.4.3">
<h3><span class="header-section-number">4.4.3</span> Matching the tree and the data<a href="preparing-your-tree-and-data-for-pcms-in-r.html#matching-the-tree-and-the-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Finally, once we know which species do not match up, we need to remove species missing from the data from the tree, and remove species missing from the tree from the data.</p>
<p>We can remove species missing from the tree easily using <code>drop.tip</code>.
You need to list the species which you do <strong>not</strong> want to select and then drop them from the tree instead of selecting the species you do want.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb112-1" aria-hidden="true" tabindex="-1"></a>mytree <span class="ot">&lt;-</span> <span class="fu">drop.tip</span>(frogtree, check<span class="sc">$</span>tree_not_data)</span></code></pre></div>
<p>To remove species from the data which are not in the tree you can use <code>match</code> and <code>subset</code> as follows:</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb113-1" aria-hidden="true" tabindex="-1"></a>matches <span class="ot">&lt;-</span> <span class="fu">match</span>(frogdata<span class="sc">$</span>tiplabel, check<span class="sc">$</span>data_not_tree, <span class="at">nomatch =</span> <span class="dv">0</span>)</span>
<span id="cb113-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb113-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">subset</span>(frogdata, matches <span class="sc">==</span> <span class="dv">0</span>)</span></code></pre></div>
<p><code>==</code> means <em>equals</em>.
So this line of code selects species which do appear in the <code>data_not_tree</code> list of species, i.e. their value from <code>matches</code> is 0.</p>
<p>Always check this has worked as expected by checking the data and the phylogeny.
In the first instance you can just use <code>str</code> to make sure you have the expected number of species in each:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(mytree)</span></code></pre></div>
<pre><code>## List of 4
##  $ edge       : int [1:418, 1:2] 211 212 213 214 215 216 217 218 219 220 ...
##  $ edge.length: num [1:418] 0.166 0.114 0.102 0.4 0.133 ...
##  $ Nnode      : int 209
##  $ tip.label  : chr [1:210] &quot;Ascaphus_truei_Ascaphidae&quot; &quot;Leiopelma_hochstetteri_Leiopelmatidae&quot; &quot;Alytes_obstetricans_Alytidae&quot; &quot;Discoglossus_pictus_Alytidae&quot; ...
##  - attr(*, &quot;class&quot;)= chr &quot;phylo&quot;
##  - attr(*, &quot;order&quot;)= chr &quot;cladewise&quot;</code></pre>
<p>The matched phylogeny has 210 species in it (instead of the original 214 species in because the four that weren’t in the data have been removed).</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(mydata)</span></code></pre></div>
<pre><code>## Rows: 210
## Columns: 11
## $ Binomial         &lt;chr&gt; &quot;Allophryne_ruthveni&quot;, &quot;Eupsophus_roseus&quot;, &quot;Alytes_obstetricans&quot;, &quot;…
## $ Family           &lt;chr&gt; &quot;Allophrynidae&quot;, &quot;Alsodidae&quot;, &quot;Alytidae&quot;, &quot;Alytidae&quot;, &quot;Aromobatidae…
## $ Genus            &lt;chr&gt; &quot;Allophryne&quot;, &quot;Eupsophus&quot;, &quot;Alytes&quot;, &quot;Discoglossus&quot;, &quot;Allobates&quot;, &quot;…
## $ tiplabel         &lt;chr&gt; &quot;Allophryne_ruthveni_Allophrynidae&quot;, &quot;Eupsophus_roseus_Alsodidae&quot;, …
## $ Adult_habitat    &lt;chr&gt; &quot;Scansorial&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwellin…
## $ Life_history     &lt;chr&gt; &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;…
## $ Sex_dichromatism &lt;chr&gt; &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent…
## $ SVL              &lt;dbl&gt; 23.76667, 38.37500, 37.46667, 62.64000, 25.60000, 27.03333, 38.4000…
## $ mass             &lt;dbl&gt; 1.0000000, 7.5000000, 6.6666667, 24.4000000, 2.1000000, 2.4666667, …
## $ rootmass         &lt;dbl&gt; 0.9917748, 1.9273451, 1.8690601, 2.8885465, 1.2549697, 1.3507613, 1…
## $ eyesize          &lt;dbl&gt; 3.200000, 5.362500, 6.366667, 7.550000, 4.116667, 3.783333, 4.75000…</code></pre>
<p>We now have 210 species in the dataset too, great!</p>
<p>One final check to make is to check what type of dataset R thinks <code>mydata</code> is. We can do that using the function <code>class</code></p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mydata)</span></code></pre></div>
<pre><code>## [1] &quot;tbl_df&quot;     &quot;tbl&quot;        &quot;data.frame&quot;</code></pre>
<p>Interestingly <code>mydata</code> is a <code>tbl_df</code>, <code>tbl</code> and <code>data.frame</code>. This is a data type created by functions in the <code>tidyverse</code> called a tibble. Tibbles have all sorts of useful properties (and a cute name!) but they’re also a pain as many functions don’t recognise that they contain a data frame and will not work on tibbles. We can fix this very easily though, just force R to make <code>mydata</code> into a data frame as follows:</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to a data frame</span></span>
<span id="cb120-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb120-2" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(mydata)</span>
<span id="cb120-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb120-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Check this is now a data frame</span></span>
<span id="cb120-4"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb120-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mydata)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<p>Finally, we might want to output these cleaned and tidied data and tree to our folder so rather than doing this every time we start an analysis, we can just use these tidy versions. To do this we can use:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Write the cleaned data to a new file</span></span>
<span id="cb122-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb122-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(mydata, <span class="at">path =</span> <span class="st">&quot;data/clean-frog-data.csv&quot;</span>)</span>
<span id="cb122-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb122-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb122-4"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb122-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Write the cleaned tree to a new file</span></span>
<span id="cb122-5"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb122-5" aria-hidden="true" tabindex="-1"></a><span class="fu">write.nexus</span>(mytree, <span class="at">file =</span> <span class="st">&quot;data/clean-frog-tree.nex&quot;</span>)</span></code></pre></div>
<p>Note, however, that you will have to repeat this preparation process if you add species or data to your tree or dataset at a later date.</p>
</div>
<div id="nas-in-your-data" class="section level3 hasAnchor" number="4.4.4">
<h3><span class="header-section-number">4.4.4</span> NAs in your data<a href="preparing-your-tree-and-data-for-pcms-in-r.html#nas-in-your-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>When running analyses we can only use species that have a value for each of the variables in the model. In comparative analyses it’s common to have missing data, recorded as NA (not applicable) in our .csv files and in R.
It can therefore be useful to subset your data so that you have a complete set of variables for certain analyses. For example let’s see how many “NAs” there are in each of our variables (don’t worry too much about this code if it seems confusing to you!):</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb123-1" aria-hidden="true" tabindex="-1"></a>mydata <span class="sc">%&gt;%</span></span>
<span id="cb123-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb123-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">everything</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb123-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb123-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise_all</span>(<span class="fu">list</span>(<span class="sc">~</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(.))))</span></code></pre></div>
<pre><code>##   Binomial Family Genus tiplabel Adult_habitat Life_history Sex_dichromatism SVL mass
## 1        0      0     0        0             0           22               36   0    0
##   rootmass eyesize
## 1        0       0</code></pre>
<p>22 species don’t have a value for <code>Life_history</code>, and 36 species don’t have a value for <code>Sex_dichromatism</code>. Many PCMs will just ignore NA values, but if you needed to remove these you could subset them out as follows, leaving only the 163 species with complete data.</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset out the species with NA values for some variables</span></span>
<span id="cb125-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb125-2" aria-hidden="true" tabindex="-1"></a>frog_noNA <span class="ot">&lt;-</span> <span class="fu">filter</span>(mydata, </span>
<span id="cb125-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb125-3" aria-hidden="true" tabindex="-1"></a>                    <span class="sc">!</span><span class="fu">is.na</span>(Life_history) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(Sex_dichromatism))</span>
<span id="cb125-4"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb125-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb125-5"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb125-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the data</span></span>
<span id="cb125-6"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb125-6" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(frog_noNA)</span></code></pre></div>
<pre><code>## Rows: 163
## Columns: 11
## $ Binomial         &lt;chr&gt; &quot;Allophryne_ruthveni&quot;, &quot;Eupsophus_roseus&quot;, &quot;Alytes_obstetricans&quot;, &quot;…
## $ Family           &lt;chr&gt; &quot;Allophrynidae&quot;, &quot;Alsodidae&quot;, &quot;Alytidae&quot;, &quot;Alytidae&quot;, &quot;Aromobatidae…
## $ Genus            &lt;chr&gt; &quot;Allophryne&quot;, &quot;Eupsophus&quot;, &quot;Alytes&quot;, &quot;Discoglossus&quot;, &quot;Allobates&quot;, &quot;…
## $ tiplabel         &lt;chr&gt; &quot;Allophryne_ruthveni_Allophrynidae&quot;, &quot;Eupsophus_roseus_Alsodidae&quot;, …
## $ Adult_habitat    &lt;chr&gt; &quot;Scansorial&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwelling&quot;, &quot;Ground-dwellin…
## $ Life_history     &lt;chr&gt; &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;Free-living larvae&quot;, &quot;…
## $ Sex_dichromatism &lt;chr&gt; &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent&quot;, &quot;Absent…
## $ SVL              &lt;dbl&gt; 23.76667, 38.37500, 37.46667, 62.64000, 25.60000, 27.03333, 38.4000…
## $ mass             &lt;dbl&gt; 1.0000000, 7.5000000, 6.6666667, 24.4000000, 2.1000000, 2.4666667, …
## $ rootmass         &lt;dbl&gt; 0.9917748, 1.9273451, 1.8690601, 2.8885465, 1.2549697, 1.3507613, 1…
## $ eyesize          &lt;dbl&gt; 3.200000, 5.362500, 6.366667, 7.550000, 4.116667, 3.783333, 4.75000…</code></pre>
</div>
</div>
<div id="quick-template-code" class="section level2 hasAnchor" number="4.5">
<h2><span class="header-section-number">4.5</span> Quick template code<a href="preparing-your-tree-and-data-for-pcms-in-r.html#quick-template-code" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To help you do this with your own data, I’ve condensed the above into one script below so you don’t forget any of the steps.</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load packages</span></span>
<span id="cb127-2"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb127-3"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(geiger)</span>
<span id="cb127-4"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb127-5"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-6"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the tree</span></span>
<span id="cb127-7"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-7" aria-hidden="true" tabindex="-1"></a>frogtree <span class="ot">&lt;-</span> <span class="fu">read.nexus</span>(<span class="st">&quot;data/frog-tree.nex&quot;</span>)</span>
<span id="cb127-8"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the tree summary</span></span>
<span id="cb127-9"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-9" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(frogtree)</span>
<span id="cb127-10"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-11"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the tree as a circular/fan phylogeny with small labels</span></span>
<span id="cb127-12"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(frogtree, <span class="at">cex =</span> <span class="fl">0.2</span>, <span class="at">typ =</span> <span class="st">&quot;fan&quot;</span>, <span class="at">no.margin =</span> <span class="cn">TRUE</span>)</span>
<span id="cb127-13"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-14"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is binary</span></span>
<span id="cb127-15"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-15" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE</span></span>
<span id="cb127-16"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-16" aria-hidden="true" tabindex="-1"></a><span class="fu">is.binary</span>(frogtree) </span>
<span id="cb127-17"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-18"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is rooted</span></span>
<span id="cb127-19"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-19" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE</span></span>
<span id="cb127-20"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-20" aria-hidden="true" tabindex="-1"></a><span class="fu">is.rooted</span>(frogtree)</span>
<span id="cb127-21"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-22"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the tree is ultrametric</span></span>
<span id="cb127-23"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-23" aria-hidden="true" tabindex="-1"></a><span class="co"># We want this to be TRUE </span></span>
<span id="cb127-24"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-24" aria-hidden="true" tabindex="-1"></a><span class="fu">is.ultrametric</span>(frogtree)</span>
<span id="cb127-25"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-26"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the data</span></span>
<span id="cb127-27"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-27" aria-hidden="true" tabindex="-1"></a>frogdata <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/frog-eyes.csv&quot;</span>)</span>
<span id="cb127-28"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-29"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the data</span></span>
<span id="cb127-30"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-30" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(frogdata)</span>
<span id="cb127-31"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-32"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Check whether the names match in the data and the tree</span></span>
<span id="cb127-33"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-33" aria-hidden="true" tabindex="-1"></a>check <span class="ot">&lt;-</span> <span class="fu">name.check</span>(<span class="at">phy =</span> frogtree, <span class="at">data =</span> frogdata, </span>
<span id="cb127-34"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-34" aria-hidden="true" tabindex="-1"></a>                    <span class="at">data.names =</span> frogdata<span class="sc">$</span>tiplabel)</span>
<span id="cb127-35"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at check</span></span>
<span id="cb127-36"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-36" aria-hidden="true" tabindex="-1"></a>check</span>
<span id="cb127-37"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-38"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-38" aria-hidden="true" tabindex="-1"></a><span class="do">### Correct any typos/taxonomic errors in the tree or data </span><span class="al">###</span></span>
<span id="cb127-39"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-39" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb127-40"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-40" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove species missing from the data</span></span>
<span id="cb127-41"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-41" aria-hidden="true" tabindex="-1"></a>mytree <span class="ot">&lt;-</span> <span class="fu">drop.tip</span>(frogtree, check<span class="sc">$</span>tree_not_data)</span>
<span id="cb127-42"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-43"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-43" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove species missing from the tree</span></span>
<span id="cb127-44"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-44" aria-hidden="true" tabindex="-1"></a>matches <span class="ot">&lt;-</span> <span class="fu">match</span>(frogdata<span class="sc">$</span>tiplabel, check<span class="sc">$</span>data_not_tree, <span class="at">nomatch =</span> <span class="dv">0</span>)</span>
<span id="cb127-45"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-45" aria-hidden="true" tabindex="-1"></a>mydata <span class="ot">&lt;-</span> <span class="fu">subset</span>(frogdata, matches <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb127-46"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-47"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-47" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the tree summary</span></span>
<span id="cb127-48"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-48" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(mytree)</span>
<span id="cb127-49"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-50"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-50" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at the data</span></span>
<span id="cb127-51"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-51" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(mydata)</span>
<span id="cb127-52"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-53"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-53" aria-hidden="true" tabindex="-1"></a><span class="do">## OPTIONAL</span></span>
<span id="cb127-54"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Make any required subsets of the tree/data</span></span>
<span id="cb127-55"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-56"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-56" aria-hidden="true" tabindex="-1"></a><span class="do">## OPTIONAL</span></span>
<span id="cb127-57"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Write the cleaned data to a new file</span></span>
<span id="cb127-58"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-58" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(mydata, <span class="at">path =</span> <span class="st">&quot;data/clean-frog-data.csv&quot;</span>)</span>
<span id="cb127-59"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-60"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-60" aria-hidden="true" tabindex="-1"></a><span class="co"># Write the cleaned tree to a new file</span></span>
<span id="cb127-61"><a href="preparing-your-tree-and-data-for-pcms-in-r.html#cb127-61" aria-hidden="true" tabindex="-1"></a><span class="fu">write.nexus</span>(mytree, <span class="at">file =</span> <span class="st">&quot;data/clean-frog-tree.nex&quot;</span>)</span></code></pre></div>
</div>
<div id="summary-3" class="section level2 hasAnchor" number="4.6">
<h2><span class="header-section-number">4.6</span> Summary<a href="preparing-your-tree-and-data-for-pcms-in-r.html#summary-3" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>You should now know how to prepare your tree and dataset for a PCM analysis in R.</p>
</div>
<div id="practical-exercise-2" class="section level2 hasAnchor" number="4.7">
<h2><span class="header-section-number">4.7</span> Practical exercise<a href="preparing-your-tree-and-data-for-pcms-in-r.html#practical-exercise-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the data folder there is another tree (<code>primate-tree.nex</code>) and dataset (<code>primate-data.csv</code>) for investigating the evolution of primate life-history variables. These data come from the PanTHERIA database <span class="citation">(<a href="#ref-jones2009pantheria" role="doc-biblioref">Jones et al. 2009</a>)</span> and 10kTrees <span class="citation">(<a href="#ref-arnold201010ktrees" role="doc-biblioref">Arnold, Matthews, and Nunn 2010</a>)</span>.</p>
<p>Read in the tree and data then prepare them for a PCM analysis.</p>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-arnold201010ktrees" class="csl-entry">
Arnold, Christian, Luke J Matthews, and Charles L Nunn. 2010. <span>“The 10kTrees Website: A New Online Resource for Primate Phylogeny.”</span> <em>Evolutionary Anthropology: Issues, News, and Reviews</em> 19 (3): 114–18.
</div>
<div id="ref-feng2017phylogenomics" class="csl-entry">
Feng, Yan-Jie, David C Blackburn, Dan Liang, David M Hillis, David B Wake, David C Cannatella, and Peng Zhang. 2017. <span>“Phylogenomics Reveals Rapid, Simultaneous Diversification of Three Major Clades of Gondwanan Frogs at the Cretaceous–Paleogene Boundary.”</span> <em>Proceedings of the National Academy of Sciences</em> 114 (29): E5864–70.
</div>
<div id="ref-jones2009pantheria" class="csl-entry">
Jones, Kate E, Jon Bielby, Marcel Cardillo, Susanne A Fritz, Justin O’Dell, C David L Orme, Kamran Safi, et al. 2009. <span>“PanTHERIA: A Species-Level Database of Life History, Ecology, and Geography of Extant and Recently Extinct Mammals: Ecological Archives E090-184.”</span> <em>Ecology</em> 90 (9): 2648–48.
</div>
<div id="ref-thomas2020eye" class="csl-entry">
Thomas, Kate N, David J Gower, Rayna C Bell, Matthew K Fujita, Ryan K Schott, and Jeffrey W Streicher. 2020. <span>“Eye Size and Investment in Frogs and Toads Correlate with Adult Habitat, Activity Pattern and Breeding Ecology.”</span> <em>Proceedings of the Royal Society B</em> 287 (1935): 20201393.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="understanding-plotting-and-manipulating-phylogenies-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="phylogenetic-signal-in-r.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/04-Preparation.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["pcm-primer.pdf", "pcm-primer.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
